{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf70ba23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>jQuery(function() {if (jQuery(\"body.notebook_app\").length == 0) { jQuery(\".input_area\").toggle(); jQuery(\".prompt\").toggle();}});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import IPython.core.display as di # Example: di.display_html('<h3>%s:</h3>' % str, raw=True)\n",
    "di.display_html('<script>jQuery(function() {if (jQuery(\"body.notebook_app\").length == 0) { jQuery(\".input_area\").toggle(); jQuery(\".prompt\").toggle();}});</script>', raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b74def7",
   "metadata": {},
   "source": [
    "# Biophysical model\n",
    "\n",
    "The goal of `multidms` is to jointly infer mutational effects across multiple deep mutational scanning (DMS) experiments, including how much each mutation's effect differs between experiments.\n",
    "We refer to these differences as *shifts*.\n",
    "If the experiments were performed with different homologs of the same protein, a shift would indicate epistasis between the shifted mutation and the amino-acid mutations that separate the homologs.\n",
    "Or, if they were performed with the same wildtype protein under different selective conditions (e.g., selection for viral entry using different but related host receptors), shifts would indicate condition-specific effects.\n",
    "Ultimately, this model was designed to identify those shifts relative to userâ€™s chosen _reference condition_ with feature selection via a lasso ($L_1$) regularization \n",
    "<cite data-footcite=\"Hastie2015Lasso\">(Hastie et al. 2015)</cite> \n",
    "of the shift parameters described in the additive _latent phenotype_ section below.\n",
    "\n",
    "`multidms` is compatible with DMS data that have the following characteristics.\n",
    "First, the data must report a functional score (e.g., log enrichment ratio) for each variant from each DMS experiment, where a variant constitutes a unique gene sequence covering the entire mutagenized region.\n",
    "Thus, the deep-sequencing data must resolve complete haplotypes.\n",
    "Second, most mutations must be seen in multiple unique variants per experiment, as this number is the basis by which shifts are regularized (see below).\n",
    "This second requirement is often met in DMS libraries with multiple mutations per variant, as long as each mutation occurs in multiple genetic backgrounds, or in libraries with one mutation per variant, as long as each variant is uniquely barcoded and individual mutations are found in the background of multiple barcodes. Given input data from one or more experiments, `multidms` fits to all the data and provides:\n",
    "(1) mutational effects $\\beta$ across all experiments,\n",
    "(2) $\\Delta$, how effects are shifted between experiments, and \n",
    "(3) a model to predict the functional score of any given variant.\n",
    "\n",
    "**Note**: We suggest reading the \n",
    "[Otwinowski et al. 2018](https://www.pnas.org/doi/10.1073/pnas.1804015115)\n",
    "<cite data-footcite=\"Otwinowski2018GlobalEpistasis\"></cite> \n",
    "paper to understand the approach for modeling global epistasis before reading the rest of the documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66a63e37",
   "metadata": {},
   "source": [
    "## Joint model composition\n",
    "\n",
    "At a high level, the model is a composition of three\n",
    "functions which describe the expected biophysical interactions underlying a given phenotype;\n",
    "(1) $\\phi$, an additive model describing a variant's latent phenotype under a given condition,\n",
    "(2) $g$, a global epistasis model shared by all conditions to disentangle the effects of multiple mutations on the same variant, and\n",
    "(3) $t$, an _optional_ final output activation function which can account for variant functional scores which have been clipped at some lower bound _prior_ to using `multidms`. \n",
    "\n",
    "Concretely, the predicted phenotype for a given variant $v$ under condition $d$ is given by \n",
    "\n",
    "$$\n",
    "\\hat{y}_{v, d} = t_{\\gamma_{d}}(g_{\\theta}(\\phi_{d, \\beta, \\Delta}(v))\n",
    "$$\n",
    "\n",
    "Where \n",
    "$\\gamma$, $\\theta$, $\\beta$, and $\\Delta$\n",
    "are _free_ parameters inferred from experimental observations during the fitting process. We describe the individual components and their associated parameters in more detail below.\n",
    "\n",
    "The motivation behind defining an abstract model in terms of its components are (1) modularity for method testing and development, and (2) the ability to provide multiple options for model components before jit-compilation and fitting of the specified model architecture. While there is only a single option for the latent prediction, $\\phi$, we offer a few options _post-latent_ modeling, $g$ and $t$, that encompass the needs of differing research goals and experimental techniques. Generally speaking, the package defaults for these components should be sufficient for most purposes, and in this case feel free to ignore the `multidms.biophysical` module all-together as this functionality is hidden unless explicitly specified during the instantiation of a `Model` object. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3941b412",
   "metadata": {},
   "source": [
    "For more about how `multidms` model components are defined and composed, as well as how the models work with the [JAX](https://github.com/google/jax) infrastructure to fit parameters, we suggest you take a look at the [multidms package biophysical docs](https://matsengrp.github.io/multidms/multidms.biophysical.html#biophysical-model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b9a42e3",
   "metadata": {},
   "source": [
    "## Latent phenotype\n",
    "\n",
    "For each mutation $m$, the model defines a single mutation effect parameter, $\\beta_{m}$ shared by all conditions.\n",
    "Additionally for each non-reference experiment $d$ (the letter $d$ is used as a mnemonic for DMS), the model defines the set of mutational shift parameters, $\\Delta_{d,m}$, that quantifies the shift a given mutation's effect relative to some reference condition. \n",
    "Each mutation, for each non-reference condition, is associated with an independent shift parameter.\n",
    "For example, consider three total experimental conditions, $d \\in \\{1, 2, 3\\}$, \n",
    "where $d = 1$ is the _reference_ condition.\n",
    "The model then defines two sets of (non-reference) _shift_ parameters $\\Delta_{2, m}$, $\\Delta_{3, m}$ that may be fit to non-zero values. \n",
    "\n",
    "Concretely, the latent phenotype of any variant, $v$, from the experimental condition, $d$,\n",
    "is computed like so:\n",
    "\n",
    "$$\n",
    "\\phi_d(v) = \\beta_0 + \\alpha_d + \\sum_{m \\in v} (\\beta_{m} + \\Delta_{d, m})\n",
    "$$\n",
    "\n",
    "Where:\n",
    "\n",
    "* $\\beta_0$ is a wildtype offset parameter applied to the latent prediction of all experiments.\n",
    "* $\\alpha_d$ is a set of bias parameters, optionally applied to the latent phenotype of all _non-reference_ experiments. These parameters may capture the effect of the _bundle_ of non-identical mutations which may exist between experiments.\n",
    "* $\\beta_{m}$ is the latent phenotypic effect of mutation $m$ shared by all experiments (See the note below), \n",
    "* $\\Delta_{d, m}$ is the shift of the effect of mutation $m$ in condition $d$.\n",
    "* $v$ is the set of all mutations relative to the reference wild type sequence including all potential non-identical wildtype mutations that separate condition $d$ from the reference condition.\n",
    "\n",
    "If a _non-reference_ experiment has a different wildtype sequence from the _reference_ (e.g., the wildtype sequences are homologs), then the `multidms.Data` object will encode non-reference genotypes relative to the reference wild type. \n",
    "Thus, the summation term includes all mutations at non-identical sites that convert between the two sequences.\n",
    "Likewise, If a mutation occurs at a non-identical site, then the mutation is encoded relative to the reference.\n",
    "For example, consider a protein where site $30$ is a Y in the non-reference experiment's wildtype sequence and an A in the reference experiment's wildtype sequence.\n",
    "If a variant from the non-reference experiment had a Y30G mutation, then the mutation would be defined as A30G in the summation term.\n",
    "This does not assume that Y30G has the same effect as A30G.\n",
    "It merely follows the strategy to define all sequences relative to the reference, which is practical because it ensures that each experiment informs the exact same set of $\\beta_m$ and $\\Delta_{d,m}$ parameters.\n",
    "If a variant from a non-reference experiment had a reversion mutation at a non-identical site (e.g., Y30A), then the mutation would not be included in the summation term for that variant since the variant would have the reference amino-acid identity at that site.\n",
    "In comparison, A30Y would be included in the summation term of all variants from the non-reference experiment that lack a mutation at site 30.\n",
    "If mutations at non-identical sites are not sampled in the DMS libraries (e.g., Y30A is missing from the non-reference experiment and A30Y is missing from the reference experiment), then $\\alpha_d$ can be used to capture the combined effects of these missing mutations.\n",
    "Otherwise, if all such mutations are sampled, $\\alpha_d$ can be locked at zero.\n",
    "\n",
    "One aspect of this approach which may be important to note, is that none of the experiment wildtype predictions are guerenteed to be zero-centered due to the shift, and offset parameters. To correct for this, `multidms` offers the ability (in predictive functions) to report both functional scores, as well as the functional score _effect_ values which are reported as the difference between any given variant functional score and the wildtype such that $\\hat{y}_{v, d} = \\hat{y}_{v, d} - \\hat{y}_{\\text{WT}, d}$. The API documentation specifies how to achieve the desired behavior."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6416322",
   "metadata": {},
   "source": [
    "## Global epistasis\n",
    "\n",
    "Latent phenotypes as described above give rise to functional scores according to a global-epistasis function.\n",
    "If this function is non-linear, then the model allows mutations to non-additively effect functional scores, helping to account for global epistasis. This type of model is useful for inferring effects of individual mutations in variants with more than one mutation. Below, we decribe the available options to model global-epistasis in the `multidms` infrastructure.\n",
    "\n",
    "**Note**: when analyzing DMS libraries in which variants have a maximum of one mutation, then it will not be possible for the model to learn the shape of global-epistasis, in which case we provide an `identity` global-epistasis function described below, which assumes no global epistasis, but still allows the user to take advantage of the rest of the `multidms` approach."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd8613ea",
   "metadata": {},
   "source": [
    "### Sigmoidal model (default):\n",
    "\n",
    "By default, the global-epistasis function here assumes a sigmoidal relationship between\n",
    "a protein's latent property and its functional score measured in the experiment\n",
    "(e.g., log enrichment score). Using free parameters, the sigmoid\n",
    "can flexibly conform to an optimal shape informed by the data. \n",
    "Note that this function is independent from the\n",
    "experimental condition from which a variant is observed.\n",
    "\n",
    "Given latent phenotype, $\\phi_d(v) = z$, let\n",
    "\n",
    "$$\n",
    "g(z) =  \\frac{\\theta_{\\text{scale}}}{1 + e^{-z}} + \\theta_{\\text{bias}}\n",
    "$$\n",
    "\n",
    "Where $\\theta_{\\text{scale}}$ and $\\theta_{\\text{bias}}$ are free parameters defining the range and lower bound of the sigmoid, respectively.\n",
    "\n",
    "Below is an interactive plot showing the effect of the sigmoidal global epistasis as a function of an adjustable $\\theta_{\\text{scale}}$, and $\\theta_{\\text{bias}}$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d3e310c",
   "metadata": {
    "hideCode": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-751fe3f1ca74427391e995ac96a6ec5b\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-751fe3f1ca74427391e995ac96a6ec5b\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-751fe3f1ca74427391e995ac96a6ec5b\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.6.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.6.1\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-ebf3a22bc86bf79e10d2059712505f43\"}, \"mark\": {\"type\": \"line\"}, \"encoding\": {\"x\": {\"field\": \"latent\", \"scale\": {\"domain\": [-5, 5]}, \"title\": \"latent phenotype\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"phenotype\", \"scale\": {\"domain\": [-5, 5]}, \"title\": \"predicted phenotype\", \"type\": \"quantitative\"}}, \"params\": [{\"name\": \"theta_scale\", \"bind\": {\"input\": \"range\", \"max\": 10, \"min\": 0.1}, \"value\": 5}, {\"name\": \"theta_bias\", \"bind\": {\"input\": \"range\", \"max\": 5, \"min\": -10}, \"value\": 0}], \"transform\": [{\"calculate\": \"(((1 / (1 + exp((-1 * datum['latent'])))) * theta_scale) + theta_bias)\", \"as\": \"phenotype\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.6.1.json\", \"datasets\": {\"data-ebf3a22bc86bf79e10d2059712505f43\": [{\"latent\": -5.0}, {\"latent\": -4.795918367346939}, {\"latent\": -4.591836734693878}, {\"latent\": -4.387755102040816}, {\"latent\": -4.183673469387755}, {\"latent\": -3.979591836734694}, {\"latent\": -3.7755102040816326}, {\"latent\": -3.571428571428571}, {\"latent\": -3.36734693877551}, {\"latent\": -3.163265306122449}, {\"latent\": -2.9591836734693877}, {\"latent\": -2.7551020408163263}, {\"latent\": -2.5510204081632653}, {\"latent\": -2.3469387755102042}, {\"latent\": -2.142857142857143}, {\"latent\": -1.9387755102040813}, {\"latent\": -1.7346938775510203}, {\"latent\": -1.5306122448979593}, {\"latent\": -1.3265306122448979}, {\"latent\": -1.1224489795918364}, {\"latent\": -0.9183673469387754}, {\"latent\": -0.7142857142857144}, {\"latent\": -0.5102040816326525}, {\"latent\": -0.3061224489795915}, {\"latent\": -0.1020408163265305}, {\"latent\": 0.1020408163265305}, {\"latent\": 0.3061224489795915}, {\"latent\": 0.5102040816326534}, {\"latent\": 0.7142857142857144}, {\"latent\": 0.9183673469387754}, {\"latent\": 1.1224489795918373}, {\"latent\": 1.3265306122448983}, {\"latent\": 1.5306122448979593}, {\"latent\": 1.7346938775510203}, {\"latent\": 1.9387755102040813}, {\"latent\": 2.1428571428571432}, {\"latent\": 2.3469387755102042}, {\"latent\": 2.5510204081632653}, {\"latent\": 2.755102040816327}, {\"latent\": 2.959183673469388}, {\"latent\": 3.163265306122449}, {\"latent\": 3.36734693877551}, {\"latent\": 3.571428571428571}, {\"latent\": 3.775510204081632}, {\"latent\": 3.979591836734695}, {\"latent\": 4.183673469387756}, {\"latent\": 4.387755102040817}, {\"latent\": 4.591836734693878}, {\"latent\": 4.795918367346939}, {\"latent\": 5.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"latent\": numpy.linspace(-5, 5, 50)})\n",
    "\n",
    "slider_s = alt.binding_range(min=0.1, max=10)\n",
    "var_s = alt.param(bind=slider_s, value=5, name=\"theta_scale\")\n",
    "\n",
    "slider_b = alt.binding_range(min=-10, max=5)\n",
    "var_b = alt.param(bind=slider_b, value=0, name=\"theta_bias\")\n",
    "\n",
    "(\n",
    "    alt.Chart(df)\n",
    "    .transform_calculate(\n",
    "        phenotype=(1 / (1 + alt.expr.exp(-1*alt.datum['latent'])))\n",
    "        * var_s\n",
    "        + var_b\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"latent\", title=\"latent phenotype\", scale=alt.Scale(domain=[-5, 5])),\n",
    "        y=alt.Y(\"phenotype:Q\", title=\"predicted phenotype\", scale=alt.Scale(domain=[-5, 5]))\n",
    "    )\n",
    "    .mark_line()\n",
    "    .add_params(var_s, var_b)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e6885d",
   "metadata": {},
   "source": [
    "### Softplus model\n",
    "\n",
    "This function is a log-transformed version of a sigmoid function from above, with  $\\theta_\\text{scale}$ and $\\theta_\\text{bias}$ parameters serving similar roles.\n",
    "The shape of this function mimics the Hill equation from the above example if the y-axis is instead the log of the fraction of protein molecules that are folded.\n",
    "Such a function could be more appropriate for modeling functional scores in log space.\n",
    "You may note that this function has no natural lower bound, and thus it is recommended to use this model with\n",
    "a user defined lower bound which is described in the _Optional truncation of predicted functional scores_ section below.\n",
    "\n",
    "Given latent phenotype, $\\phi_d(v) = z$, let\n",
    "\n",
    "$$\n",
    "g(z) =  -\\theta_\\text{scale}\\log\\left(1+e^{-z}\\right) + \\theta_\\text{bias}\n",
    "$$\n",
    "\n",
    "**Note** In the `multidms` API, This behavior is accomplished by setting the `output_activation` parameter in the constructor for `Model` to be a pointer to the function `multidms.biophysical.softplus_activation`\n",
    "\n",
    "Below is an interactive plot showing the effect of the modified \n",
    "softplus function as a function of an adjustable $\\theta_\\text{scale}$ scaling parameter, and lower bound, $\\theta_\\text{bias}$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51286a0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-ac092d0f062a4e1b931f8a04c69a2da6\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-ac092d0f062a4e1b931f8a04c69a2da6\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-ac092d0f062a4e1b931f8a04c69a2da6\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.6.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.6.1\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-ebf3a22bc86bf79e10d2059712505f43\"}, \"mark\": {\"type\": \"line\"}, \"encoding\": {\"x\": {\"field\": \"latent\", \"scale\": {\"domain\": [-5, 5]}, \"title\": \"global epistasis prediction (z')\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"phenotype\", \"scale\": {\"domain\": [-5, 5]}, \"title\": \"predicted phenotype\", \"type\": \"quantitative\"}}, \"params\": [{\"name\": \"lambda_sp\", \"bind\": {\"input\": \"range\", \"max\": 5, \"min\": 1}, \"value\": 1}, {\"name\": \"lower_bound\", \"bind\": {\"input\": \"range\", \"max\": 5, \"min\": 0}, \"value\": 2}], \"transform\": [{\"calculate\": \"((log((1 + exp((-1 * datum['latent'])))) * (-1 * lambda_sp)) + lower_bound)\", \"as\": \"phenotype\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.6.1.json\", \"datasets\": {\"data-ebf3a22bc86bf79e10d2059712505f43\": [{\"latent\": -5.0}, {\"latent\": -4.795918367346939}, {\"latent\": -4.591836734693878}, {\"latent\": -4.387755102040816}, {\"latent\": -4.183673469387755}, {\"latent\": -3.979591836734694}, {\"latent\": -3.7755102040816326}, {\"latent\": -3.571428571428571}, {\"latent\": -3.36734693877551}, {\"latent\": -3.163265306122449}, {\"latent\": -2.9591836734693877}, {\"latent\": -2.7551020408163263}, {\"latent\": -2.5510204081632653}, {\"latent\": -2.3469387755102042}, {\"latent\": -2.142857142857143}, {\"latent\": -1.9387755102040813}, {\"latent\": -1.7346938775510203}, {\"latent\": -1.5306122448979593}, {\"latent\": -1.3265306122448979}, {\"latent\": -1.1224489795918364}, {\"latent\": -0.9183673469387754}, {\"latent\": -0.7142857142857144}, {\"latent\": -0.5102040816326525}, {\"latent\": -0.3061224489795915}, {\"latent\": -0.1020408163265305}, {\"latent\": 0.1020408163265305}, {\"latent\": 0.3061224489795915}, {\"latent\": 0.5102040816326534}, {\"latent\": 0.7142857142857144}, {\"latent\": 0.9183673469387754}, {\"latent\": 1.1224489795918373}, {\"latent\": 1.3265306122448983}, {\"latent\": 1.5306122448979593}, {\"latent\": 1.7346938775510203}, {\"latent\": 1.9387755102040813}, {\"latent\": 2.1428571428571432}, {\"latent\": 2.3469387755102042}, {\"latent\": 2.5510204081632653}, {\"latent\": 2.755102040816327}, {\"latent\": 2.959183673469388}, {\"latent\": 3.163265306122449}, {\"latent\": 3.36734693877551}, {\"latent\": 3.571428571428571}, {\"latent\": 3.775510204081632}, {\"latent\": 3.979591836734695}, {\"latent\": 4.183673469387756}, {\"latent\": 4.387755102040817}, {\"latent\": 4.591836734693878}, {\"latent\": 4.795918367346939}, {\"latent\": 5.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"latent\": numpy.linspace(-5, 5, 50)})\n",
    "\n",
    "slider_lsp = alt.binding_range(min=1, max=5)\n",
    "alpha_scale = alt.param(bind=slider_lsp, value=1, name=\"lambda_sp\")\n",
    "\n",
    "slider_lb = alt.binding_range(min=0, max=5)\n",
    "alpha_bias = alt.param(bind=slider_lb, value=2, name=\"lower_bound\")\n",
    "\n",
    "(\n",
    "    alt.Chart(df)\n",
    "    .transform_calculate(\n",
    "        phenotype=alt.expr.log(1 + alt.expr.exp(-1*alt.datum['latent']))\n",
    "        * (-1 * alpha_scale)\n",
    "        + alpha_bias\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"latent\", title=\"global epistasis prediction (z')\", scale=alt.Scale(domain=[-5, 5])),\n",
    "        y=alt.Y(\"phenotype:Q\", title=\"predicted phenotype\", scale=alt.Scale(domain=[-5, 5]))\n",
    "    )\n",
    "    .mark_line()\n",
    "    .add_params(alpha_scale, alpha_bias)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8127046f",
   "metadata": {},
   "source": [
    "### Single-layer neural network model:\n",
    "\n",
    "If you prefer a less constrained shape for global epistasis, we also offer the ability to learn the shape of global epistasis using a single-layer neural network, sometimes referred to as a [multi-layer perceptron](https://en.wikipedia.org/wiki/Multilayer_perceptron). This is similar to what was presented by [Zhou et. al. 2022](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9522415/)\n",
    "\n",
    "For this option, the user defines a number of units in the singular hidden layer of the model. For each hidden unit, we introduce three parameters (two weights and a bias) to be inferred. All weights are clipped at zero to maintain assumptions of monotonicity in the resulting epistasis function shape.  The network applies a sigmoid activation to each internal unit before a final transformation and addition of a constant gives us our predicted functional score. \n",
    "\n",
    "Given latent phenotype, $\\phi_d(v) = z$, let\n",
    "\n",
    "$$\n",
    "g(z) = b^{o}+ \\sum_{i}^{n} \\frac{w^{o}_{i}}{1 + e^{w^{l}_{i}*z + b^{l}_{i}}}\n",
    "$$\n",
    "\n",
    "Where: \n",
    "\n",
    "* $n$ is the number of units in the hidden layer.\n",
    "* $w^{l}_{i}$ and $w^{o}_{i}$ are free parameters representing latent and output tranformations, respectively, associated with unit $i$ in the hidden layer of the network. \n",
    "* $b^{l}_{i}$ is a free parameter, as an added bias term to unit $i$.\n",
    "* $b^{o}$ is a constant, singular free parameter.\n",
    "\n",
    "**Note** This is an advanced feature and we advise against its use unless the other options are not sufficiently parameterized for particularly complex experimental conditions.\n",
    "\n",
    "**Note** In the `multidms` API, This behavior is accomplished by setting the `output_activation` parameter in the constructor for `Model` to be a pointer to the function `multidms.biophysical.single_layer_nn`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6cab98f",
   "metadata": {},
   "source": [
    "### Identity (no epistasis):\n",
    "\n",
    "Given latent phenotype, $\\phi_d(v) = z$, let\n",
    "\n",
    "$$\n",
    "g(z) = z\n",
    "$$\n",
    "\n",
    "In this functional form, there is no global epistasis: latent phenotypes are identical to functional scores.\n",
    "We recommend this functional form if none of the variants in the DMS experiment have more than one mutation, since multi-mutant variants are needed for the model to accurately infer a non-linear global-epistasis function.\n",
    "It can also be used as a baseline to determine if a non-linear global-epistasis function leads to better model fit.\n",
    "\n",
    "**Note** In the `multidms` API, This is accomplished by setting the `output_activation` parameter in the constructor for `Model` to be a pointer to the function definition `multidms.biophysical.identity_activation`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99477fac",
   "metadata": {},
   "source": [
    "## Optional normalization of observed functional scores\n",
    "\n",
    "A common way to compute functional scores is with log enrichment ratios, where all scores from a given condition are normalized so that the wildtype sequence from that condition has a value of zero.\n",
    "If wildtype sequences differ between conditions, then log enrichment ratios may not be directly comparable between conditions, as they are normalized to different reference points.\n",
    "This breaks the assumption of our joint-modeling scheme: that all functional scores are directly comparable.\n",
    "\n",
    "Ideally one or more of the same sequences would be included in the DMS library of each condition, so that all scores could be normalized to the same sequence.\n",
    "However, if there are no common sequences, then it may be possible to computationally estimate how to normalize scores so that they are more directly comparable.\n",
    "\n",
    "To this end, the model includes an additional parameter $\\gamma_d$ for each non-reference condition that allows functional scores from that condition to be normalized as follows:\n",
    "\n",
    "$$\n",
    "y_{v,d}^{\\text{norm}} = y_{v,d} + \\gamma_d\n",
    "$$\n",
    "\n",
    "where $\\gamma_d$ for the reference condition is locked at zero.\n",
    "There is a theoretical basis for adding $\\gamma_d$ to $y_{v,d}$ if functional scores are log enrichment ratios.\n",
    "As mentioned above, log enrichment ratios are normalized so that the wildtype sequence from a given experiment has a value of zero, according to the formula:\n",
    "\n",
    "$$\n",
    "y_{v,d} = \\log(E_{v,d}) - \\log(E_{\\text{wt},d})\n",
    "$$\n",
    "\n",
    "Thus, adding $\\gamma_d$ to $y_{v,d}$ is akin to renormalizing the log enrichment ratios so that a different sequence has a functional score of zero.\n",
    "In theory, for each non-reference condition, there is a $\\gamma_d$ that normalizes functional scores to be relative to the wildtype sequence of the reference condition.\n",
    "If these values are not experimentally measured, the model allows $\\gamma_d$ parameters to be fit during optimization, which assumes that the correct $\\gamma_d$ values will give the best model fit.\n",
    "Alternatively, $\\gamma_d$ values can be locked at zero if the user does not wish to implement this optional feature."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8f8e0a9",
   "metadata": {},
   "source": [
    "## Optional truncation of predicted functional scores\n",
    "\n",
    "By default, the output of the model is equal to the output of the global epistasis model chosen $\\hat{y}_d(v) = g(\\phi_d(v))$. However, we provide infrastructure to clip the predicted functional scores at some lower bound for users who may clip their input data. Here, we'll describe the motivation and approach for this.\n",
    "\n",
    "The commonly reported fold-change metric for functional scores described above falls victim to \n",
    "a limit of detection problem for deleterious mutations: once a mutation is sufficiently deleterious, further differences (eg, between -3 and -5) become largely meaningless because it is already basically nonfunctional.\n",
    "Thus, it's common for researchers to simply _truncate_ (i.e. _clip_) the functional at some lower bound $l$, where observations below are assumed largely to be outlier.\n",
    "\n",
    "When fitting the model to these data, it may be desirable to truncate _predicted_ functional scores in a similar way.\n",
    "This is especially relevant when allowing $\\gamma_d$ values to be non-zero.\n",
    "For instance, say the user has truncated observed functional scores at a lower bound of $-3$ across all conditions.\n",
    "If $\\gamma_d$ is fit to $-1.0$ for one of the non-reference conditions, then the new floor of normalized functional scores for that condition would be $-4$, while the floor for the reference condition would still be $-3$.\n",
    "In this case, the global-epistasis function could find itself in a pickle: if it allowed predictions to go below $-3$, it could help model the floor of points in the non-reference condition, but hurt with modeling the floor of points in the reference condition.\n",
    "However, if it was able to truncate predicted scores for a given condition at the floor for that condition, then this tension would be relieved: predicted scores for the reference condition could be truncated at $-3$, while (non-truncated) predicted scores for the non-reference condition could go as low as $-4$.\n",
    "\n",
    "To this end, the software package provides an option to truncate predicted functional scores which should be used if (and only if) the user has clipped the functional scores in their data to some lower bound.\n",
    "To enable this, the mathematical model passes predicted functional scores through an activation function, $t$.\n",
    "\n",
    "In essence, this is a modified _softplus_ activation, ($\\text{softplus}(x)=\\log(1 + e^{x})$) with a _lower bound_ at $l + \\gamma_{h}$, as well as a _ramping_ coefficient, $\\lambda_{\\text{sp}}$. \n",
    "\n",
    "Concretely, if we let $z' = g(\\phi_d(v))$, then the predicted functional score of our model is given by:\n",
    "\n",
    "$$\n",
    "t(z') = \\lambda_{sp}\\log(1 + e^{\\frac{z' - l}{\\lambda_{sp}}}) + l\n",
    "$$\n",
    "\n",
    "Functionally speaking, this truncates scores below a lower bound, while leaving scores above (mostly) unaltered. There is a small range of input values where the function smoothly transitions between a flat regime (where data is truncated) and a linear regime (where data is not truncated). \n",
    "\n",
    "**Note** We recommend leaving the $\\lambda_{sp}$ parameter at It's default value of $0.1$. this ensures a sharp transition between regimes similar to a [ReLU](https://en.wikipedia.org/wiki/Rectifier_(neural_networks)) function, but retain the differentiable property for gradient based optimization. However, the option is there in case you find the model will not converge.\n",
    "\n",
    "Below is an interactive plot showing the effect of the modified \n",
    "softplus activation as a function of an adjustable $\\lambda_{sp}$ scaling parameter, and lower bound, $l$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e02d0f67",
   "metadata": {
    "hideCode": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-7a419dff89f64d09b5456acd6411811f\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-7a419dff89f64d09b5456acd6411811f\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-7a419dff89f64d09b5456acd6411811f\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.6.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.6.1\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-cabdca2257cf20e376d04c62a66e94ee\"}, \"mark\": {\"type\": \"line\"}, \"encoding\": {\"x\": {\"field\": \"latent\", \"scale\": {\"domain\": [-10, 5]}, \"title\": \"global epistasis prediction (z')\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"phenotype\", \"scale\": {\"domain\": [-10, 5]}, \"title\": \"predicted phenotype\", \"type\": \"quantitative\"}}, \"params\": [{\"name\": \"lambda_sp\", \"bind\": {\"input\": \"range\", \"max\": 2, \"min\": 0.1}, \"value\": 1}, {\"name\": \"lower_bound\", \"bind\": {\"input\": \"range\", \"max\": 0, \"min\": -10}, \"value\": -3.5}], \"transform\": [{\"calculate\": \"((log((1 + exp(((datum['latent'] - lower_bound) / lambda_sp)))) * lambda_sp) + lower_bound)\", \"as\": \"phenotype\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.6.1.json\", \"datasets\": {\"data-cabdca2257cf20e376d04c62a66e94ee\": [{\"latent\": -10.0}, {\"latent\": -9.848484848484848}, {\"latent\": -9.696969696969697}, {\"latent\": -9.545454545454545}, {\"latent\": -9.393939393939394}, {\"latent\": -9.242424242424242}, {\"latent\": -9.09090909090909}, {\"latent\": -8.93939393939394}, {\"latent\": -8.787878787878787}, {\"latent\": -8.636363636363637}, {\"latent\": -8.484848484848484}, {\"latent\": -8.333333333333334}, {\"latent\": -8.181818181818182}, {\"latent\": -8.030303030303031}, {\"latent\": -7.878787878787879}, {\"latent\": -7.727272727272727}, {\"latent\": -7.575757575757576}, {\"latent\": -7.424242424242424}, {\"latent\": -7.2727272727272725}, {\"latent\": -7.121212121212121}, {\"latent\": -6.96969696969697}, {\"latent\": -6.818181818181818}, {\"latent\": -6.666666666666666}, {\"latent\": -6.515151515151516}, {\"latent\": -6.363636363636363}, {\"latent\": -6.212121212121212}, {\"latent\": -6.0606060606060606}, {\"latent\": -5.909090909090909}, {\"latent\": -5.757575757575758}, {\"latent\": -5.6060606060606055}, {\"latent\": -5.454545454545454}, {\"latent\": -5.303030303030303}, {\"latent\": -5.151515151515151}, {\"latent\": -5.0}, {\"latent\": -4.848484848484849}, {\"latent\": -4.696969696969697}, {\"latent\": -4.545454545454545}, {\"latent\": -4.393939393939394}, {\"latent\": -4.242424242424242}, {\"latent\": -4.090909090909091}, {\"latent\": -3.9393939393939394}, {\"latent\": -3.787878787878788}, {\"latent\": -3.636363636363636}, {\"latent\": -3.4848484848484844}, {\"latent\": -3.333333333333333}, {\"latent\": -3.1818181818181817}, {\"latent\": -3.0303030303030303}, {\"latent\": -2.878787878787879}, {\"latent\": -2.7272727272727266}, {\"latent\": -2.5757575757575752}, {\"latent\": -2.424242424242424}, {\"latent\": -2.2727272727272725}, {\"latent\": -2.121212121212121}, {\"latent\": -1.9696969696969688}, {\"latent\": -1.8181818181818183}, {\"latent\": -1.666666666666666}, {\"latent\": -1.5151515151515156}, {\"latent\": -1.3636363636363633}, {\"latent\": -1.212121212121211}, {\"latent\": -1.0606060606060606}, {\"latent\": -0.9090909090909083}, {\"latent\": -0.7575757575757578}, {\"latent\": -0.6060606060606055}, {\"latent\": -0.45454545454545503}, {\"latent\": -0.30303030303030276}, {\"latent\": -0.1515151515151505}, {\"latent\": 0.0}, {\"latent\": 0.15151515151515227}, {\"latent\": 0.30303030303030276}, {\"latent\": 0.45454545454545503}, {\"latent\": 0.6060606060606055}, {\"latent\": 0.7575757575757578}, {\"latent\": 0.9090909090909101}, {\"latent\": 1.0606060606060606}, {\"latent\": 1.2121212121212128}, {\"latent\": 1.3636363636363633}, {\"latent\": 1.5151515151515156}, {\"latent\": 1.6666666666666679}, {\"latent\": 1.8181818181818183}, {\"latent\": 1.9696969696969706}, {\"latent\": 2.121212121212121}, {\"latent\": 2.2727272727272734}, {\"latent\": 2.424242424242424}, {\"latent\": 2.575757575757576}, {\"latent\": 2.7272727272727284}, {\"latent\": 2.878787878787879}, {\"latent\": 3.030303030303031}, {\"latent\": 3.1818181818181817}, {\"latent\": 3.333333333333334}, {\"latent\": 3.4848484848484844}, {\"latent\": 3.6363636363636367}, {\"latent\": 3.787878787878789}, {\"latent\": 3.9393939393939394}, {\"latent\": 4.090909090909092}, {\"latent\": 4.242424242424242}, {\"latent\": 4.3939393939393945}, {\"latent\": 4.545454545454547}, {\"latent\": 4.696969696969697}, {\"latent\": 4.8484848484848495}, {\"latent\": 5.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Below is an interactive plot showing the effect of the modified \n",
    "# softplus activation as a function of an adjustable $\\lambda_{sp}$ scaling parameter, and lower bound, $l$:\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"latent\": numpy.linspace(-10, 5, 100)})\n",
    "\n",
    "slider_lsp = alt.binding_range(min=0.1, max=2)\n",
    "var_lambda_sp = alt.param(bind=slider_lsp, value=1, name=\"lambda_sp\")\n",
    "\n",
    "slider_lb = alt.binding_range(min=-10, max=0)\n",
    "var_lower_bound = alt.param(bind=slider_lb, value=-3.5, name=\"lower_bound\")\n",
    "\n",
    "(\n",
    "    alt.Chart(df)\n",
    "    .transform_calculate(\n",
    "        phenotype=alt.expr.log(1 + alt.expr.exp((alt.datum['latent']-var_lower_bound)/var_lambda_sp))\n",
    "        * var_lambda_sp\n",
    "        + var_lower_bound\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"latent\", title=\"global epistasis prediction (z')\", scale=alt.Scale(domain=[-10, 5])),\n",
    "        y=alt.Y(\"phenotype:Q\", title=\"predicted phenotype\", scale=alt.Scale(domain=[-10, 5]))\n",
    "    )\n",
    "    .mark_line()\n",
    "    .add_params(var_lambda_sp, var_lower_bound)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9401c976",
   "metadata": {},
   "source": [
    "## Fitting procedure\n",
    "\n",
    "The `multidms` software package has a framework for fitting free parameters in the model given input DMS data.\n",
    "The mathematical model is coded in [Python](https://www.python.org/) using the \n",
    "[JAX library](https://github.com/google/jax) \n",
    "<cite data-footcite=\"jax2018github\"></cite>\n",
    "allowing for autograd and XLA compilation for high-performance optimization.\n",
    "We approach our non-smooth optimization problem via proximal gradient descent in order to satisfy several simultaneous constraints such as L1 regularization, optionally locking parameters, and clipping parameters at pre-specified lower bounds.\n",
    "Specifically, we use [JAXopt](https://jaxopt.github.io/stable/index.html) to optimize parameters via full-batch proximal gradient descent using the [JAXopt.ProximalGradient](https://jaxopt.github.io/stable/_autosummary/jaxopt.ProximalGradient.html) function.\n",
    "\n",
    "We now define notation and introduce our model more formally to describe the fitting procedure without ambiguity.\n",
    "Let $M\\in\\mathbb{N}$ denote the number of distinct mutations, and represent a given variant $v\\subset\\mathcal{M}\\equiv\\{1,\\dots,M\\}$ as an index set of the mutations it contains ($v$ is in the set of subsets of the $M$ mutations, i.e. $v\\in\\mathcal{V}\\equiv2^\\mathcal{M}$, where $2^\\mathcal{M}$ denotes the power set of $\\mathcal{M}$).\n",
    "\n",
    "Consider a variant $v$ as an indicator (one-hot) vector $x_v\\in\\{0,1\\}^M$ where $\\left[x_v\\right]_i=1$ if $i\\in v$ and $\\left[x_v\\right]_i=0$ otherwise.\n",
    "Let $D\\in\\mathbb{N}$ be the number of experiments and write $\\mathbf{1}$ for the $D$-vector of ones (the letter $D$ is used as a mnemonic for DMS).\n",
    "We introduce an additive _latent phenotype_ model jointly for $D$ experiments via a family of affine maps $\\phi_{(\\beta_0,\\beta,\\Delta)}:\\{0, 1\\}^M\\to\\mathbb{R}^D$ defined by\n",
    "\n",
    "$$\n",
    "\\phi_{(\\beta_0,\\beta,\\Delta)}(x) = \\beta_0\\mathbf{1} + (\\mathbf{1}\\beta^\\intercal + \\Delta) x, \\quad x\\in\\{0, 1\\}^M,\n",
    "$$\n",
    "\n",
    "where the family is parameterized by intercept $\\beta_0\\in\\mathbb{R}$ and mutational effects $\\beta\\in\\mathbb{R}^M$ that are shared by all $D$ output dimensions, and shift matrix $\\Delta\\in\\mathbb{R}^{D\\times M}$.\n",
    "We require that the first row of $\\Delta$ is the zero $M$-vector, so that the reference experiment (indexed 1 WLOG) has no shifts, and $\\beta$ is then interpreted as the vector of mutational effects in the reference experiment, with the intercept $\\beta_0$ representing the latent phenotype of the wildtype sequence in the reference experiment.\n",
    "\n",
    "Next, we introduce a _global-epistasis function_ via a family of strictly monotone maps $g_\\theta:\\mathbb{R}\\to\\mathbb{R}$ that we use to take latent phenotypes to predicted functional scores.\n",
    "This family is parameterized by $\\theta\\in\\mathbb{R}^r$ for some $r\\in\\mathbb{N}$.\n",
    "By default, `multidms` will use the sigmoid function\n",
    "\n",
    "$$\n",
    "g_\\theta(z) = \\theta_0 + \\frac{\\theta_1}{1+e^{-z}}, \\quad z\\in\\mathbb{R},\n",
    "$$\n",
    "\n",
    "with $r=2$ parameters, which allows us to adapt the output range of the global-epistasis function (the interval $(\\theta_0, \\theta_0 + \\theta_1)$) to the range of our functional score data, but is otherwise a fixed link function (imposing a _gauge_ on our latent phenotype model parameters).\n",
    "We finally compute the predicted functional score in experiment $d\\in\\{1,\\dots,D\\}$ of a variant $v\\in\\mathcal{V}$ with one-hot encoding $x_v\\in\\{0, 1\\}^M$ as\n",
    "\n",
    "$$\n",
    "\\hat{y}_d(x_v) = g_\\theta\\left(\\left[\\phi_{(\\beta_0,\\beta,\\Delta)}(x_v)\\right]_d\\right).\n",
    "$$\n",
    "\n",
    "Our data consists of sets of one-hot encoded variants and their associated functional scores from each of $D$ experiments.\n",
    "Denote these as $\\mathcal{D}_d \\subset \\{0, 1\\}^M\\times\\mathbb{R}$ for $d=1,\\dots,D$.\n",
    "We minimize an objective of the form:\n",
    "\n",
    "$$\n",
    "f(\\beta_0,\\beta,\\Delta,\\theta) = \\sum_{d=1}^D\\sum_{(x, y)\\in\\mathcal{D}_d}\\!\\!\\ell(y,\\hat{y}_d(x)) + \\lambda\\|\\Delta\\|_{1,1},\n",
    "$$\n",
    "\n",
    "where $\\ell:\\mathbb{R}\\times\\mathbb{R}\\to\\mathbb{R}$ is a Huber loss function measuring the difference between a predicted and an observed functional score, $\\lambda\\in\\mathbb{R}$ is the lasso penalty weight, and $\\|\\cdot\\|_{1,1}$ denotes the entrywise $L_1$ norm (not to be confused with the matrix 1-norm $\\|\\cdot\\|_1$).\n",
    "Note that the parameters $(\\beta_0,\\beta,\\Delta,\\theta)$ appear in the loss function via the predicted functional score equation, but are suppressed in the objective for notational compactness.\n",
    "Note also that, by taking $\\lambda=0$, the loss term of the objective becomes separable over the $D$ experiments, so marginal inference is recovered as a special case.\n",
    "\n",
    "For a general global epistasis function $g_\\theta$, the objective is in general non-convex.\n",
    "However, with the simple sigmoid function, it is bi-convex in $(\\beta_0, \\beta, \\Delta)$ and $\\theta$.\n",
    "This can be seen by noting that, for fixed $\\theta$, the prediction model takes the form of a generalized linear model with a sigmoid link function, and for fixed $(\\beta_0, \\beta, \\Delta)$, the model parameterized by $\\theta$ is a linear regression problem.\n",
    "We optimize the objective using the Nesterov-accelerated proximal gradient method with backtracking line search\n",
    "<cite data-footcite=\"beck2009fast\"></cite>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e76efc",
   "metadata": {},
   "source": [
    "## References"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
