{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7f0224-78d0-4cfc-ac06-48c1d1b4dfe2",
   "metadata": {},
   "source": [
    "# multidms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc224f4-d5ca-49a6-ae17-d3f8c7fac997",
   "metadata": {},
   "source": [
    "## Overview of model\n",
    "\n",
    "The `multidms` model applies to a case where you have DMS datasets for two or more homologs and are interested in identifying shifts in mutational effects between homologs.\n",
    "To do so, the model defines one homolog as a reference homolog.\n",
    "For each mutation, the model fits one parameter that quantifies the effect of the mutation in the reference homolog.\n",
    "For each non-reference homolog, it also fits a shift parameter that quantifies the shift in the mutation's effect in the non-reference homolog relative to the reference.\n",
    "Shift parameters can be regularized, encouraging most of them to be close to zero.\n",
    "This regularization step is a useful way to eliminate the effects of experimental noise, and is most useful in cases where you expect most mutations to have the same effects between homologs, such as for homologs that are close relatives.\n",
    "\n",
    "The model uses a global-epistasis function to disentangle the effects of multiple mutations on the same variant.\n",
    "To do so, it assumes that mutational effects additively influence a latent biophysical property the protein (e.g., $\\Delta G$ of folding).\n",
    "The mutational-effect parameters described above operate at this latent level.\n",
    "The global-epistasis function then assumes a sigmoidal relationship between a protein's latent property and its functional score measured in the experiment (e.g., log enrichment score).\n",
    "Ultimately, mutaitonal parameters, as well as ones controlling the shape of the sigmoid, are all jointly fit to maximize agreement between predicted and observed functional scores acorss all variants of all homologs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd6382d-47a3-49d1-ac4f-a909f7ebf8a8",
   "metadata": {},
   "source": [
    "## Detailed description of the model\n",
    "\n",
    "For each variant $v$ from homolog $h$, we use a global-epistasis function $g$ to convert a latent phenotype $\\phi$ to a functional score $f$:\n",
    "\n",
    "$$f(v,h) = g_{\\alpha}(\\phi(v,h))$$\n",
    "\n",
    "where $g$ is a sigmoid and $\\alpha$ is a set of parameters encoding the sigmoid.\n",
    "\n",
    "The latent phenotype is computed in the following way:\n",
    "\n",
    "$$\\phi(v,h) = c_h + \\sum_{m \\in v} (x_m + s_{m,h})$$\n",
    "\n",
    "where:\n",
    "* $c_h$ is the wildtype latent phenotype for homolog $h$\n",
    "* $x_m$ is the latent phenotypic effect of mutation $m$. See details below.\n",
    "* $s_{m,h}$ is the shift of the effect of mutation $m$ in homolog $h$. These parameters are fixed to zero for the reference homolog. For non-reference homologs, they are defined in the same way as $x_m$ parameters.\n",
    "\n",
    "The $x_m$ variable is defined such that mutations are always relative to the reference homolog.\n",
    "This way, each homolog informs the exact same parameters, even at sites that differ in wildtype amino acid.\n",
    "For instance, if the wildtype amino acid at site 30 is an A in the reference homolog, but is a Y in a second homolog, then a Y30G mutation in the second homolog is defined as the sum of the following two mutations relative to A: A30Y (negated) and A30G.\n",
    "This approach assumes that mutational effects can be negated, such that A30Y has the opposite effect as Y30A.\n",
    "It also assumes that mutational effects are additive, such that the effect of Y30G is the sum of the effects of Y30A and A30G.\n",
    "\n",
    "The below expression defines this approach more explicitly for an arbitrary site. It uses the notation $x_{\\mathtt{X},n,\\mathtt{Z}}$ where $\\mathtt{X}$ and $\\mathtt{Z}$ are amino acids.\n",
    "For a site $n$, if $aa_{\\text{wt}}$ is the site's wildtype amino acid in a non-reference homolog, $aa_{\\text{mut}}$ is a mutant amino acid in a variant of that homolog, and $aa_{\\text{ref}}$ is the site's wildtype amino acid in the reference homolog, then $x_m$ is:\n",
    "\n",
    "$$x_{aa_{\\text{wt}},n,aa_{\\text{mut}}} = \\begin{cases}\n",
    "      x_{aa_{\\text{ref}},n,aa_{\\text{mut}}} & \\text{if } aa_{\\text{wt}} = aa_{\\text{ref}}\\\\\n",
    "      -x_{aa_{\\text{ref}},n,aa_{\\text{wt}}} & \\text{if } aa_{\\text{mut}} = aa_{\\text{ref}}\\\\\n",
    "      - x_{aa_{\\text{ref}},n,aa_{\\text{wt}}} + x_{aa_{\\text{ref}},n,aa_{\\text{mut}}} & \\text{otherwise}\\\\\n",
    "\\end{cases}$$\n",
    "\n",
    "In this way, we can express the whole model with reference only to coefficients that differ from the reference homolog.\n",
    "\n",
    "Ultimately, we fit parameters using a loss function with one term that scores differences between predicted and observed values and another that uses L1 regularization to penalize non-zero $s_{m,h}$ values:\n",
    "\n",
    "$$ L_{\\text{total}} = \\sum_{h} \\left[\\sum_{v} L_{\\text{fit}}(y_{v,h}, f(v,h)) + \\lambda \\sum_{m} |s_{m,h}|\\right]$$\n",
    "\n",
    "where:\n",
    "* $L_{\\text{total}}$ is the total loss function\n",
    "* $L_{\\text{fit}}$ is a loss function that penalizes differences in predicted vs. observed functional scores\n",
    "* $y_{v,h}$ is the experimentally measured functional score of variant $v$ from homolog $h$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e943741e-4fd1-47d8-b330-ec1713f76a42",
   "metadata": {},
   "source": [
    "## Model using matrix algebra\n",
    "\n",
    "We compute a vector or predicted latent phenotypes $P_{h}$ as:\n",
    "\n",
    "$$P_{h} = C_h + (W_h \\cdot (X + S_h))$$\n",
    "\n",
    "where:\n",
    "* $X$ is a vector of all $x_m$ values\n",
    "* $S_h$ is a matrix of all $s_{m,h}$ values\n",
    "* $W_h$ is a sparse matrix, where rows are variants, columns are mutations (all defined relative to the reference homolog), and values are weights of 0, 1, or -1. These weights are used to compute the phenotype of each variant given the mutations present.\n",
    "* $C_h$ is a vector of the homolog's $c_h$ value, repeated $n$ times, where $n$ is the number of variants.\n",
    "\n",
    "In the matrix algebra, the sum of $X_m$ and $S_{m,h}$ gives a vector of mutational effects, with one entry per mutation.\n",
    "Multiplying the matrix $W_h$ by this vector gives a new vector with one entry per variant, where values are the sum of mutational effects, weighted by the variant-specific weights in $W_h$.\n",
    "Adding $C_h$ values to this vector will give a vector of predicted latent phenotypes for each variant.\n",
    "\n",
    "Next, the global-epistasis function can be used to convert a vector of predicted latent phenotypes to a vector of predicted functional scores.\n",
    "\n",
    "$$F_{h,pred} = g_{\\alpha}(P_h)$$\n",
    "\n",
    "Finally, this vector could be fed into a loss function and compared with a vector of observed functional scores."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b14e961",
   "metadata": {},
   "source": [
    "## Binarymap augmentation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dd2195a",
   "metadata": {},
   "source": [
    "Some toy code for making a \"HomologBinaryMap\" Class. \n",
    "\n",
    "No need to re-invent a binary encoding for aa sequence data, we can simply augment the `BinaryClass`\n",
    "Object after converting the mutations to be with respect to the reference sequence as stated above defining $x_{aa_{\\text{wt}},n,aa_{\\text{mut}}}$. As seen above, some of the binary columns (substitutions) that are encoding the effect of a homolog substitution must be negated. These are sites that distinguish the wildtype sequences of the reference and homolog.\n",
    "\n",
    "This could eventually be a proper implimentation of a `BinaryMap`-inhereted class that overides the `__init__()` and any other relevent functions. It would inherit all the same parameters, add the homolog wildtype sequence (`hom_wtseq`, below), and make the reference wildtype sequence (`wtseq`, in `BinaryMap`) manditory. \n",
    "\n",
    "Note: we're still not entirely sure how to handle gapped alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3a529bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [1,1,1,2,2,2,2,2],\n",
    "    'variant' : ['M1E', 'G2R', 'G2P', 'M1E', 'P2R', 'P2G', 'M1E P2G', 'M1E P2R'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, -5, 0.4, 2.7, 0.5],\n",
    "}\n",
    "test_df = pd.DataFrame(test_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6fba9914",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import binarymap as bmap\n",
    "\n",
    "\n",
    "def mut_wrt_ref(wtseq:str, aa_subs:list[str]):\n",
    "    \"\"\"\n",
    "    Convert a list of given variants for a non reference homolog\n",
    "    to be with respect to the reference sequence\n",
    "    such that a given mutation, {X_aawt,n,aamut} is equal to:\n",
    "\n",
    "    1. X_{aawt,n,aamut}                             if aawt == aaref\n",
    "    2. -1 * X_{aaref,n,aawt}                        if aamut == aaref\n",
    "    3. (-1 * X_{aaref,n,aawt}) + X_{aawt,n,aamut}   else\n",
    "    \n",
    "    aawt is the wildtype aa of the homolog\n",
    "    aamut is the substitution aa of the homolog\n",
    "    aaref is the wildttype aa of the reference sequence\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    wtseq : str\n",
    "        amino acid sequence for the reference homolog.\n",
    "    \n",
    "    aa_subs : list[str]\n",
    "        a list a <wt><site><sub> formatted substitutions on a \n",
    "        non-reference homolog.\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    list:\n",
    "        non reference homolog substitutions converted to be with respect \n",
    "        the reference homolog. The -1 coefficients from the above\n",
    "        expressions are not included.\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    def convert_single_mut(mutation:str):\n",
    "        \"\"\"function for converting single mutations\"\"\"\n",
    "\n",
    "        # Parse the aawt, site, and aamut\n",
    "        pattern = r'(?P<aawt>\\w)(?P<site>\\d+)(?P<aamut>\\w)'\n",
    "        match = re.search(pattern, mutation)\n",
    "        assert match != None, mutation\n",
    "        aawt = match.group('aawt')\n",
    "        site = match.group('site')\n",
    "        aamut = match.group('aamut')\n",
    "\n",
    "        # Get the wildtype amino acid of the reference homolog\n",
    "        ref_aa = wtseq[int(site) - 1]\n",
    "\n",
    "        ret = []\n",
    "        \n",
    "        # If the site wt amino acid is homologous, no conversion needed\n",
    "        if ref_aa == aawt:\n",
    "            ret.append(mutation)\n",
    "            \n",
    "        # If the substitution is a back-mutation to the ref,\n",
    "        # return the reference substitution to the non-ref \n",
    "        # wildtype sequence, this will be negated\n",
    "        elif aamut == ref_aa:\n",
    "            ret.append(f\"{ref_aa}{site}{aawt}\")\n",
    "            \n",
    "        # Otherwise, the site is non homologous, and\n",
    "        # there's a new mutation. The ref sub to non-red\n",
    "        # wildtype will be negated and added to the effect\n",
    "        # of the ref wt aa mutating to the homolog mutation\n",
    "        else:\n",
    "            ret.append(f\"{ref_aa}{site}{aawt}\")\n",
    "            ret.append(f\"{ref_aa}{site}{aamut}\")\n",
    "            \n",
    "        return ret\n",
    "\n",
    "    \n",
    "    def convert_string_muts(mutations:str):\n",
    "        \"\"\"function for converting multiple mutations\"\"\"\n",
    "        \n",
    "        muts = [m for mut in mutations.split() for m in convert_single_mut(mut)]\n",
    "        return \" \".join(muts)\n",
    "\n",
    "    \n",
    "    return [convert_string_muts(muts) for muts in aa_subs]\n",
    "\n",
    "\n",
    "def homolog_bmap(wtseq, hom_wtseq, func_score_df, *args, **kwargs):\n",
    "    \"\"\"\n",
    "    Initialize a `HomologBinaryMap` which essentially converts the mutations\n",
    "    of the homolog to be wrt the reference sequence, the negates the necessary\n",
    "    columns on the binary encoding.\n",
    "    \n",
    "    extra arguments are passed to the constructor of a regular BinaryMap.\n",
    "    For the model fitting, it is suggested to scan all the homolog converted\n",
    "    mutations and pass those to the allowed_subs so that all homologs share\n",
    "    the 1'th dimension.\n",
    "    \n",
    "    Note: We're not handeling gaps or insertions, yet. For the current code,\n",
    "    we expect all homologs are alignable without gaps.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    wtseq : str\n",
    "        The wildtype sequence of the reference homolog.\n",
    "    \n",
    "    hom_wtseq : str\n",
    "        The wildtype sequence of the non-reference homolog you're encoding\n",
    "    \n",
    "    func_score_df : pandas.DataFrame\n",
    "        This should be in the same format as described in BinaryMap\n",
    "        The aa substitution converstions are handled here so \n",
    "        substitution_col should be with respect the the \n",
    "        non-reference homolog\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    binarymap.BinaryMap\n",
    "        The non-reference homolog binary map\n",
    "    \"\"\"\n",
    "\n",
    "    sub_col = \"aa_substitutions\" \n",
    "    if \"substitutions_col\" in kwargs:\n",
    "        sub_col = kwargs[\"substitutions_col\"]\n",
    "\n",
    "    h_func_scores = func_score_df.copy()\n",
    "    h_func_scores[sub_col] = mut_wrt_ref(\n",
    "        wtseq,\n",
    "        func_score_df[sub_col]\n",
    "    )\n",
    "\n",
    "    hbmap = bmap.BinaryMap(h_func_scores, *args, **kwargs)\n",
    "\n",
    "    # get all subs in variants wrt reference sequence computed above\n",
    "    subs_in_variants = {\n",
    "        s for subs in h_func_scores[sub_col]\n",
    "        for s in subs.split()\n",
    "    }\n",
    "\n",
    "    # all possible mutations that would need to be negated\n",
    "    all_non_homologous_muts = {\n",
    "        f\"{ref_aa}{i+1}{hom_aa}\" \n",
    "        for i, (ref_aa, hom_aa) in enumerate(zip(wtseq, hom_wtseq))\n",
    "        if ref_aa != hom_aa\n",
    "    }\n",
    "\n",
    "    # negate those that exist in our dataset and are non-homologous\n",
    "    to_negate = subs_in_variants.intersection(all_non_homologous_muts)\n",
    "\n",
    "    idx_to_negate = [hbmap.sub_to_i(s) for s in to_negate]\n",
    "    hbmap.binary_variants[:, idx_to_negate] *= -1\n",
    "    return hbmap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc207477",
   "metadata": {},
   "source": [
    "## Test Case"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7222d340",
   "metadata": {},
   "source": [
    "With some imaginary variants from two imaginary homologs, (`ref` and `hom`), we'll test the code above for a sanity check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1dd9def7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_wt_seq = \"MG\"\n",
    "hom_wt_seq = \"MP\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "380d116f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2R</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   homolog  variant  log2E\n",
       "0        1      M1E    2.0\n",
       "1        1      G2R   -7.0\n",
       "2        1      G2P   -0.5\n",
       "3        2      M1E    2.3\n",
       "4        2      P2R   -5.0\n",
       "5        2      P2G    0.4\n",
       "6        2  M1E P2G    2.7\n",
       "7        2  M1E P2R    0.5"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [1,1,1,2,2,2,2,2],\n",
    "    'variant' : ['M1E', 'G2R', 'G2P', 'M1E', 'P2R', 'P2G', 'M1E P2G', 'M1E P2R'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, -5, 0.4, 2.7, 0.5],\n",
    "}\n",
    "test_df = pd.DataFrame(test_dict)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60bdac10",
   "metadata": {},
   "source": [
    "Let's test it out on the df created earlier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e174a755",
   "metadata": {},
   "source": [
    "The first thing our `HomologBinarymap` will do after constructing a regular binarymap, is to conver the mutations in the homolog to be wrt to the wildtype. This step is carried out in the `homolog_bmap` function, but the below cell shows an example of the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16c7b5c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>G2R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>G2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>G2P G2R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "      <td>G2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "      <td>M1E G2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2R</td>\n",
       "      <td>0.5</td>\n",
       "      <td>M1E G2P G2R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   homolog  variant  log2E  var_wrt_ref\n",
       "0        1      M1E    2.0          M1E\n",
       "1        1      G2R   -7.0          G2R\n",
       "2        1      G2P   -0.5          G2P\n",
       "3        2      M1E    2.3          M1E\n",
       "4        2      P2R   -5.0      G2P G2R\n",
       "5        2      P2G    0.4          G2P\n",
       "6        2  M1E P2G    2.7      M1E G2P\n",
       "7        2  M1E P2R    0.5  M1E G2P G2R"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df_wrt_ref = test_df.assign(var_wrt_ref=mut_wrt_ref(ref_wt_seq, test_df.variant.values))\n",
    "test_df_wrt_ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5da7b5cd",
   "metadata": {},
   "source": [
    "Each of the homologs will get their own binary encoding, but they need to all include the same possible substitutions. This way they will have the same shape when informing the $X$ parameters in out model. So, we extract all the allowed subs from all homologs that will be included in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "623e8503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'G2P', 'G2R', 'M1E'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allowed_subs = {\n",
    "    s for subs in test_df_wrt_ref.var_wrt_ref\n",
    "    for s in subs.split()\n",
    "}\n",
    "allowed_subs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9b55f553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2R</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   homolog  variant  log2E\n",
       "0        1      M1E    2.0\n",
       "1        1      G2R   -7.0\n",
       "2        1      G2P   -0.5\n",
       "3        2      M1E    2.3\n",
       "4        2      P2R   -5.0\n",
       "5        2      P2G    0.4\n",
       "6        2  M1E P2G    2.7\n",
       "7        2  M1E P2R    0.5"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07e9939d",
   "metadata": {},
   "source": [
    "Using the set of common substitutions from above, the below cell makes a regular `BinaryMap` object for the _reference_ homolog."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15af8eb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 0 0]\n",
      " [0 0 1]\n",
      " [0 1 0]]\n",
      "{'G2R': 2, 'G2P': 1, 'M1E': 0}\n"
     ]
    }
   ],
   "source": [
    "ref_bmap = bmap.BinaryMap(\n",
    "    test_df.query('homolog == 1'),\n",
    "    substitutions_col=\"variant\",\n",
    "    allowed_subs=allowed_subs\n",
    ")\n",
    "print(ref_bmap.binary_variants.toarray())\n",
    "print({mut:ref_bmap.sub_to_i(mut) for mut in allowed_subs})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0541535e",
   "metadata": {},
   "source": [
    "Using the same list of substitutions, the next cell makes a homolog-`BinaryMap` for the _non-reference_ homolog, which includes which has done the conversions/negations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "12c3012a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 1  0  0]\n",
      " [ 0 -1  1]\n",
      " [ 0 -1  0]\n",
      " [ 1 -1  0]\n",
      " [ 1 -1  1]]\n",
      "{'G2R': 2, 'G2P': 1, 'M1E': 0}\n"
     ]
    }
   ],
   "source": [
    "hom_bmap = homolog_bmap(\n",
    "    ref_wt_seq,\n",
    "    hom_wt_seq,\n",
    "    test_df.query('homolog == 2'),\n",
    "    substitutions_col=\"variant\",\n",
    "    allowed_subs=allowed_subs\n",
    ")\n",
    "print(hom_bmap.binary_variants.toarray())\n",
    "print({mut:hom_bmap.sub_to_i(mut) for mut in allowed_subs})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b9f63a",
   "metadata": {},
   "source": [
    "Okay, so basically in order to allow for multiple mutations at the same site in the same variant, I had to remove a check in binarymap source code that raised and error for such a case. Currently, hese are lines 525 and 526 in the binarymap.py file. I could submit a PR that makes this check optional maybe?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b70bfb4",
   "metadata": {},
   "source": [
    "## JAX Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95b48012",
   "metadata": {},
   "source": [
    "## Simulation Fit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98cee907",
   "metadata": {},
   "source": [
    "## Empirical Data?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
