{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7f0224-78d0-4cfc-ac06-48c1d1b4dfe2",
   "metadata": {},
   "source": [
    "# multidms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc224f4-d5ca-49a6-ae17-d3f8c7fac997",
   "metadata": {},
   "source": [
    "## Overview of model\n",
    "\n",
    "The `multidms` model applies to a case where you have DMS datasets for two or more homologs and are interested in identifying shifts in mutational effects between homologs.\n",
    "To do so, the model defines one homolog as a reference homolog.\n",
    "For each mutation, the model fits one parameter that quantifies the effect of the mutation in the reference homolog.\n",
    "For each non-reference homolog, it also fits a shift parameter that quantifies the shift in the mutation's effect in the non-reference homolog relative to the reference.\n",
    "Shift parameters can be regularized, encouraging most of them to be close to zero.\n",
    "This regularization step is a useful way to eliminate the effects of experimental noise, and is most useful in cases where you expect most mutations to have the same effects between homologs, such as for homologs that are close relatives.\n",
    "\n",
    "The model uses a global-epistasis function to disentangle the effects of multiple mutations on the same variant.\n",
    "To do so, it assumes that mutational effects additively influence a latent biophysical property the protein (e.g., $\\Delta G$ of folding).\n",
    "The mutational-effect parameters described above operate at this latent level.\n",
    "The global-epistasis function then assumes a sigmoidal relationship between a protein's latent property and its functional score measured in the experiment (e.g., log enrichment score).\n",
    "Ultimately, mutational parameters, as well as ones controlling the shape of the sigmoid, are all jointly fit to maximize agreement between predicted and observed functional scores acorss all variants of all homologs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd6382d-47a3-49d1-ac4f-a909f7ebf8a8",
   "metadata": {},
   "source": [
    "## Detailed description of the model\n",
    "\n",
    "For each variant $v$ from homolog $h$, we use a global-epistasis function $g$ to convert a latent phenotype $\\phi$ to a functional score $f$:\n",
    "\n",
    "$$f(v,h) = g_{\\alpha}(\\phi(v,h))$$\n",
    "\n",
    "where $g$ is a sigmoid and $\\alpha$ is a set of parameters encoding the sigmoid.\n",
    "\n",
    "The latent phenotype is computed in the following way:\n",
    "\n",
    "$$\\phi(v,h) = c + \\sum_{m \\in v} (x_m + s_{m,h})$$\n",
    "\n",
    "where:\n",
    "* $c$ is the wildtype latent phenotype for the reference homolog.\n",
    "* $x_m$ is the latent phenotypic effect of mutation $m$. See details below.\n",
    "* $s_{m,h}$ is the shift of the effect of mutation $m$ in homolog $h$. These parameters are fixed to zero for the reference homolog. For non-reference homologs, they are defined in the same way as $x_m$ parameters.\n",
    "* $v$ is the set of all mutations relative to the reference wildtype sequence (including all mutations that separate homolog $h$ from the reference homolog).\n",
    "\n",
    "The $x_m$ variable is defined such that mutations are always relative to the reference homolog.\n",
    "For example, if the wildtype amino acid at site 30 is an A in the reference homolog, and a G in a non-reference homolog, then a Y30G mutation in the non-reference homolog is recorded as an A30G mutation relative to the reference.\n",
    "This way, each homolog informs the exact same parameters, even at sites that differ in wildtype amino acid.\n",
    "These are encoded in a `BinaryMap` object, where all sites that are non-identical to the reference are 1's.\n",
    "\n",
    "Ultimately, we fit parameters using a loss function with one term that scores differences between predicted and observed values and another that uses L1 regularization to penalize non-zero $s_{m,h}$ values:\n",
    "\n",
    "$$ L_{\\text{total}} = \\sum_{h} \\left[\\sum_{v} L_{\\text{fit}}(y_{v,h}, f(v,h)) + \\lambda \\sum_{m} |s_{m,h}|\\right]$$\n",
    "\n",
    "where:\n",
    "* $L_{\\text{total}}$ is the total loss function.\n",
    "* $L_{\\text{fit}}$ is a loss function that penalizes differences in predicted vs. observed functional scores.\n",
    "* $y_{v,h}$ is the experimentally measured functional score of variant $v$ from homolog $h$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e943741e-4fd1-47d8-b330-ec1713f76a42",
   "metadata": {},
   "source": [
    "## Model using matrix algebra\n",
    "\n",
    "We compute a vector or predicted latent phenotypes $P_{h}$ as:\n",
    "\n",
    "$$P_{h} = c + (X_h \\cdot (β + S_h))$$\n",
    "\n",
    "where:\n",
    "* $β$ is a vector of all $β_m$ values.\n",
    "* $S_h$ is a matrix of all $s_{m,h}$ values.\n",
    "* $X_h$ is a sparse matrix, where rows are variants, columns are mutations (all defined relative to the reference homolog), and values are weights of 0's and 1's. These weights are used to compute the phenotype of each variant given the mutations present.\n",
    "* $c$ is the same as above.\n",
    "\n",
    "In the matrix algebra, the sum of $β_m$ and $S_{m,h}$ gives a vector of mutational effects, with one entry per mutation.\n",
    "Multiplying the matrix $X_h$ by this vector gives a new vector with one entry per variant, where values are the sum of mutational effects, weighted by the variant-specific weights in $X_h$.\n",
    "Adding the $c$ value to this vector will give a vector of predicted latent phenotypes for each variant.\n",
    "\n",
    "Next, the global-epistasis function can be used to convert a vector of predicted latent phenotypes to a vector of predicted functional scores.\n",
    "\n",
    "$$F_{h,pred} = g_{\\alpha}(P_h)$$\n",
    "\n",
    "Finally, this vector could be fed into a loss function and compared with a vector of observed functional scores."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809b2cd8",
   "metadata": {},
   "source": [
    "## Import `Python` modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e3816f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import json\n",
    "import binarymap as bmap\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import jax\n",
    "jax.config.update(\"jax_enable_x64\", True)\n",
    "import jax.numpy as jnp\n",
    "from jax.experimental import sparse\n",
    "import jaxopt\n",
    "import numpy as onp\n",
    "from scipy.stats import pearsonr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b14e961",
   "metadata": {},
   "source": [
    "## Strategy for converting mutations to be relative to the reference homolog"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dd2195a",
   "metadata": {},
   "source": [
    "As described above, our strategy involves converting all mutations in all homologs to be relative to the amino-acid sequence of the reference homolog. The below function performs this step.\n",
    "\n",
    "Note 1: we will probably need some special-purpose code to handle gaps. This isn't done yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1cbc3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_homolog_modeling_data(\n",
    "    func_score_df:pd.DataFrame, \n",
    "    homologs:dict,\n",
    "    homolog_name_col: str,\n",
    "    substitution_col: str,\n",
    "    func_score_col: str\n",
    "):\n",
    "    \"\"\"\n",
    "    Takes a dataframe for making a `BinaryMap` object, and adds\n",
    "    a column where each entry is a list of mutations in a variant\n",
    "    relative to the amino-acid sequence of the reference homolog.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    func_score_df : pandas.DataFrame\n",
    "        This should be in the same format as described in BinaryMap.\n",
    "        \n",
    "    homologs : dict\n",
    "        A dictionary containing all possible target homolog \n",
    "        names (keys) and sequences (values).\n",
    "    \n",
    "    homolog_name_col : str\n",
    "        The name of the column in func_score_df that identifies the\n",
    "        homolog for a given variant. We require that the\n",
    "        reference homolog variants are labeled as 'reference'\n",
    "        in this column.\n",
    "    \n",
    "    substitution_col : str \n",
    "        The name of the column in func_score_df that\n",
    "        lists mutations in each variant relative to the wildtype\n",
    "        amino-acid sequence of the homolog in which they occur.\n",
    "        \n",
    "    func_score_col : str\n",
    "        Column in func_scores_df giving functional score for each variant.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "        \n",
    "    pd.DataFrame :\n",
    "        A dataframe with a new column 'var_wrt_ref' that contains\n",
    "        all variant mutations with respect to the wildtype.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    def mutations_wrt_ref(mutations, hom_wtseq):\n",
    "        \"\"\"\n",
    "        Takes a list of mutations for a given variant relative\n",
    "        to its background homolog and returns a list of all\n",
    "        mutations that separate the variant from the reference\n",
    "        homolog.\n",
    "        \"\"\"\n",
    "        \n",
    "        # Compute the full amino-acid sequence of the\n",
    "        # given variant\n",
    "        mutated_homolog = list(hom_wtseq)\n",
    "        for mutation in mutations.split():\n",
    "\n",
    "            # TODO: Do we need to change the regex to allow\n",
    "            # for gap '-' and stop '*' characters?\n",
    "            pattern = r'(?P<aawt>\\w)(?P<site>\\d+)(?P<aamut>[\\w\\*])'\n",
    "            match = re.search(pattern, mutation)\n",
    "            assert match != None, mutation\n",
    "            aawt = match.group('aawt')\n",
    "            site = match.group('site')\n",
    "            aamut = match.group('aamut')\n",
    "            mutated_homolog[int(site)-1] = aamut\n",
    "            \n",
    "        hom_var_seq = ''.join(mutated_homolog)\n",
    "        \n",
    "        # Make a list of all mutations that separate the variant\n",
    "        # from the reference homolog\n",
    "        ref_muts = [\n",
    "            f\"{aaref}{i+1}{aavar}\" \n",
    "            for i, (aaref, aavar) in enumerate(zip(homologs[\"reference\"], hom_var_seq))\n",
    "            if aaref != aavar\n",
    "        ]\n",
    "        \n",
    "        return \" \".join(ref_muts)\n",
    "\n",
    "    # Duplicate the substitutions_col, then convert the respective functional scores\n",
    "    func_score_df = func_score_df.assign(var_wrt_ref = func_score_df[substitution_col].values)\n",
    "    for hom_name, hom_seq in homologs.items():\n",
    "        if hom_name == \"reference\": continue\n",
    "        hom_df = func_score_df.query(f\"{homolog_name_col} == '{hom_name}'\")\n",
    "        hom_var_wrt_ref = [\n",
    "            mutations_wrt_ref(muts, homologs[hom_name]) \n",
    "            for muts in hom_df[substitution_col]\n",
    "        ]\n",
    "        func_score_df.loc[hom_df.index.values, \"var_wrt_ref\"] = hom_var_wrt_ref   \n",
    "    \n",
    "    # Get list of all allowed substitutions that we will tune beta parameters for\n",
    "    allowed_subs = {\n",
    "        s for subs in func_score_df.var_wrt_ref\n",
    "        for s in subs.split()\n",
    "    }\n",
    "    \n",
    "    # Make BinaryMap representations for each homolog\n",
    "    X, y = {}, {}\n",
    "    for homolog, homolog_func_score_df in func_score_df.groupby(\"homolog\"):\n",
    "        ref_bmap = bmap.BinaryMap(\n",
    "            homolog_func_score_df,\n",
    "            substitutions_col=\"var_wrt_ref\",\n",
    "            allowed_subs=allowed_subs,\n",
    "            func_score_col = None\n",
    "        )\n",
    "        \n",
    "        # convert binarymaps into sparse arrays for model input\n",
    "        X[homolog] = sparse.BCOO.fromdense(ref_bmap.binary_variants.toarray())\n",
    "        \n",
    "        # create jax array for functional score targets\n",
    "        y[homolog] = jnp.array(func_score_df[func_score_col].values)\n",
    "    \n",
    "    return (X, y), func_score_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7222d340",
   "metadata": {},
   "source": [
    "Next, we will test the above function with a small test case. Below, we define variants from two imaginary homologs: \"reference\" and \"2\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dd9def7",
   "metadata": {},
   "outputs": [],
   "source": [
    "homologs = {\n",
    "    \"reference\" : \"MG\",\n",
    "    \"2\" : \"MP\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "380d116f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>reference</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>reference</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>reference</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2R</td>\n",
       "      <td>-2.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     homolog  variant  log2E\n",
       "0  reference      M1E    2.0\n",
       "1  reference      G2R   -7.0\n",
       "2  reference      G2P   -0.5\n",
       "3          2      M1E    2.3\n",
       "4          2      P2R   -5.0\n",
       "5          2      P2G    0.4\n",
       "6          2  M1E P2G    2.7\n",
       "7          2  M1E P2R   -2.7"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [\"reference\",\"reference\",\"reference\",\"2\",\"2\",\"2\",\"2\",\"2\"],\n",
    "    'variant' : ['M1E', 'G2R', 'G2P', 'M1E', 'P2R', 'P2G', 'M1E P2G', 'M1E P2R'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, -5, 0.4, 2.7, -2.7],\n",
    "}\n",
    "test_df = pd.DataFrame(test_dict)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "14c72cf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modeling Input Data: {'2': BCOO(int8[5, 3], nse=6), 'reference': BCOO(int8[3, 3], nse=3)}\n",
      "Modeling Target Data: {'2': DeviceArray([ 2. , -7. , -0.5,  2.3, -5. ,  0.4,  2.7, -2.7], dtype=float64), 'reference': DeviceArray([ 2. , -7. , -0.5,  2.3, -5. ,  0.4,  2.7, -2.7], dtype=float64)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>reference</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>reference</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>G2R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>reference</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>G2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "      <td>M1E G2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>G2R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2R</td>\n",
       "      <td>-2.7</td>\n",
       "      <td>M1E G2R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     homolog  variant  log2E var_wrt_ref\n",
       "0  reference      M1E    2.0         M1E\n",
       "1  reference      G2R   -7.0         G2R\n",
       "2  reference      G2P   -0.5         G2P\n",
       "3          2      M1E    2.3     M1E G2P\n",
       "4          2      P2R   -5.0         G2R\n",
       "5          2      P2G    0.4            \n",
       "6          2  M1E P2G    2.7         M1E\n",
       "7          2  M1E P2R   -2.7     M1E G2R"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(X, y), updated_func_score_df = create_homolog_modeling_data(test_df, homologs, \"homolog\", \"variant\", \"log2E\")\n",
    "\n",
    "print(f\"Modeling Input Data: {X}\")\n",
    "print(f\"Modeling Target Data: {y}\")\n",
    "\n",
    "updated_func_score_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b70bfb4",
   "metadata": {},
   "source": [
    "## Implementing the full model with JAX"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce084f",
   "metadata": {},
   "source": [
    "Below, we implement the $\\phi$ and $g$ functions from above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f62a6fa1",
   "metadata": {},
   "source": [
    "Next, we initialize values of free parameters for downstream optimization by JAX."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eef82a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def initialize_model_params(homologs: dict, n_beta_shift_params: int, n_perceptron_units=1):\n",
    "    \"\"\"\n",
    "    initialize a set of starting parameters for the JAX model.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    homologs : dict\n",
    "        A dictionary containing all possible target homolog \n",
    "        names (keys) and sequences (values).\n",
    "    \n",
    "    n_beta_shift_params: int\n",
    "        The number of beta and shift parameters \n",
    "        (for each homolog) to initialize.\n",
    "        \n",
    "    n_perceptron_units : int\n",
    "        The number of connections that are tuned\n",
    "        for the shape of global epistasis before\n",
    "        being fed into a sigmoid.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    \n",
    "    dict :\n",
    "        all relevant parameters to be tuned with the JAX model.\n",
    "    \"\"\"\n",
    "    \n",
    "    params = {}\n",
    "    seed = 0\n",
    "    key = jax.random.PRNGKey(seed)\n",
    "\n",
    "    # initialize beta parameters from normal distribution.\n",
    "    params[\"β\"] = jax.random.normal(shape=(n_beta_shift_params,), key=key)\n",
    "\n",
    "    # initialize shift parameters\n",
    "    for homolog in homologs.keys():\n",
    "        # We expect most shift parameters to be close to zero\n",
    "        params[f\"S_{homolog}\"] = jnp.zeros(shape=(n_beta_shift_params,))\n",
    "\n",
    "    # Single bias param\n",
    "    params[\"C_ref\"] = jnp.zeros(shape=(1, ))\n",
    "    \n",
    "    # 'stretch' in x direction\n",
    "    # 'shift' in the x direction\n",
    "    # 'stretch' in the y direction\n",
    "    # 'shift' in the y direction\n",
    "    key, *subkeys = jax.random.split(key, num=5)\n",
    "    params[\"α\"]=dict(\n",
    "        sig_stretch_x = jax.random.normal(shape=(n_perceptron_units,), key=subkeys[0]),\n",
    "        sig_shift_x = jax.random.normal(shape=(1,), key=subkeys[1]),\n",
    "        sig_stretch_y = jax.random.normal(shape=(n_perceptron_units,), key=subkeys[2]),\n",
    "        sig_shift_y = jax.random.normal(shape=(1,), key=subkeys[3]),\n",
    "    )\n",
    "\n",
    "    return params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "82aca890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'β': DeviceArray([ 0.18784401, -1.28334229,  0.6494182 ], dtype=float64),\n",
       " 'S_reference': DeviceArray([0., 0., 0.], dtype=float64),\n",
       " 'S_2': DeviceArray([0., 0., 0.], dtype=float64),\n",
       " 'C_ref': DeviceArray([0.], dtype=float64),\n",
       " 'α': {'sig_stretch_x': DeviceArray([-1.31768732], dtype=float64),\n",
       "  'sig_shift_x': DeviceArray([0.44991238], dtype=float64),\n",
       "  'sig_stretch_y': DeviceArray([0.47178911], dtype=float64),\n",
       "  'sig_shift_y': DeviceArray([-0.94836575], dtype=float64)}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of columns in the sparse arrays are the number of beta/shift params to fit\n",
    "params = initialize_model_params(homologs, n_beta_shift_params = X[\"reference\"].shape[1])\n",
    "params"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78ba62df",
   "metadata": {},
   "source": [
    "Define an objective function to optimize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0405cbeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO centering wildtype -\n",
    "# TODO static args for reference -\n",
    "# TODO regularization -\n",
    "\n",
    "@jax.jit\n",
    "def ϕ(params:dict, X_h:jnp.array):\n",
    "    \"\"\"Model for predicting latent space\"\"\"\n",
    "    \n",
    "    return (X_h @ (params[\"β\"] + params[f\"S\"])) + params[\"C_ref\"]\n",
    "\n",
    "\n",
    "@jax.jit\n",
    "def g(α:dict, z_h:jnp.array):\n",
    "    \"\"\"Model for global epistasis as 'flexible' sigmoid.\"\"\"\n",
    "     \n",
    "    activations = jax.nn.sigmoid(α[\"sig_stretch_x\"] * z_h[:, None] + α[\"sig_shift_x\"])\n",
    "#     wt_activations = jax.nn.sigmoid(α[\"sig_stretch_x\"] * α[\"C_ref\"] + α[\"sig_shift_x\"])\n",
    "#     activations -= wt_activations\n",
    "    return (α[\"sig_stretch_y\"] * activations) + α[\"sig_shift_y\"] # sig shift y needs to be 0 for the reference\n",
    "\n",
    "\n",
    "@jax.jit\n",
    "def cost_smooth(params, data, δ=1):\n",
    "    \"\"\"Cost (Objective) function as a sum of huber loss across all homologs\"\"\"\n",
    "\n",
    "    X, y = data\n",
    "    loss = 0   \n",
    "    \n",
    "    # Sum the huber loss across all homologs\n",
    "    for homolog, X_h in X.items():\n",
    "        \n",
    "        # Fix the shift parameters for reference to 0\n",
    "        # Static arguments for reference params?\n",
    "        S_h = jnp.where(\n",
    "            homolog==\"reference\", \n",
    "            jnp.zeros(len(params['β'])), \n",
    "            params[f\"S_{homolog}\"]\n",
    "        )\n",
    "        \n",
    "        # Subset the params being passed into latent prediction, ϕ\n",
    "        h_params = {\"β\":params[\"β\"], \"S\":S_h, \"C_ref\":params[\"C_ref\"]}\n",
    "        \n",
    "        z_h = ϕ(h_params, X_h)\n",
    "        \n",
    "        # Pass the latent predictions through GE model prediction\n",
    "        # all GE specific parameters are stored in α\n",
    "        y_h_predicted = g(params[\"α\"], z_h)\n",
    "        \n",
    "        # compute loss at current parameter state.\n",
    "        loss += jaxopt.loss.huber_loss(y[homolog], y_h_predicted, δ).mean()\n",
    "\n",
    "    return loss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "585f4dae",
   "metadata": {},
   "source": [
    "Calculate initial loss with random parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "48d9f474",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DeviceArray(4.82309878, dtype=float64)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cost_smooth(params, (X, y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8b717b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "tol = 1e-6\n",
    "maxiter = 1000\n",
    "solver = jaxopt.GradientDescent(cost_smooth, tol=tol, maxiter=maxiter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16c38e7a",
   "metadata": {},
   "source": [
    "Optimize the entire model at once with gradient decent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ec41022",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C_ref': DeviceArray([-0.41645006], dtype=float64),\n",
       " 'S_2': DeviceArray([-0.08995514,  0.06204292, -0.22570208], dtype=float64),\n",
       " 'S_reference': DeviceArray([0., 0., 0.], dtype=float64),\n",
       " 'α': {'sig_shift_x': DeviceArray([0.79438156], dtype=float64),\n",
       "  'sig_shift_y': DeviceArray([-0.05026768], dtype=float64),\n",
       "  'sig_stretch_x': DeviceArray([-1.02919693], dtype=float64),\n",
       "  'sig_stretch_y': DeviceArray([0.00032971], dtype=float64)},\n",
       " 'β': DeviceArray([ 0.02419704, -1.12096377,  0.19726752], dtype=float64)}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "params, state = solver.run(params, data=(X, y))\n",
    "params"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95b48012",
   "metadata": {},
   "source": [
    "## Simulation Fit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3e0150b",
   "metadata": {},
   "source": [
    "* TODO J: Try removing for loop in cost function (only compute cost on homolog)\n",
    "\n",
    "* TODO H: Get will's notebook running in our environment (positive control for env)\n",
    "\n",
    "* TODO J: jnp where culprit:\n",
    "    * Try removing the shift parameters & update Phi accordingly\n",
    "    * static args for reference\n",
    "\n",
    "* TODO H: Run the Ab-CGGnaive_DMS with our model\n",
    "* TODO H: Smaller simulated sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "736a592b",
   "metadata": {},
   "outputs": [],
   "source": [
    "simulated_dataset = pd.read_csv(\"results/simulated_dataset_v1.csv\")\n",
    "simulated_dataset_lib1 = simulated_dataset.query(\"library == 'lib_1'\").copy()\n",
    "simulated_dataset_lib1.aa_substitutions.fillna(\"\", inplace=True)\n",
    "simulated_dataset_lib1 = simulated_dataset_lib1.sample(n=15000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e1b1df9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'reference': 'RSVILRAYTNSRVKRVILCNNDLPIRNIRLMMILHNSDASFSTPVGLRSG',\n",
       " 'H2': 'RVVILRAYTNSRVKRIKLCNNDRPIRNIRTMMIEHNSDAKFHTPYGLDSG'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homologs = json.load(open(\"results/homolog_aa_seqs.json\", \"r\"))\n",
    "homologs[\"reference\"] = homologs['1']\n",
    "homologs[\"H2\"] = homologs['2']\n",
    "del homologs['1']\n",
    "del homologs['2']\n",
    "homologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a82edcf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modeling Input Data: {'H2': BCOO(int8[7533, 1000], nse=86632), 'reference': BCOO(int8[7467, 1000], nse=14200)}\n",
      "Modeling Target Data: {'H2': DeviceArray([0.97637767, 1.        , 1.00481545, ..., 0.06785799,\n",
      "             1.        , 0.00103619], dtype=float64), 'reference': DeviceArray([0.97637767, 1.        , 1.00481545, ..., 0.06785799,\n",
      "             1.        , 0.00103619], dtype=float64)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>variant_call_support</th>\n",
       "      <th>codon_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "      <th>latent_phenotype</th>\n",
       "      <th>observed_phenotype</th>\n",
       "      <th>observed_enrichment</th>\n",
       "      <th>homolog</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13743</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>GATATCGCTACTGACA</td>\n",
       "      <td>1</td>\n",
       "      <td>CGG12TTG</td>\n",
       "      <td>R12L</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3.469912</td>\n",
       "      <td>-0.034489</td>\n",
       "      <td>0.976378</td>\n",
       "      <td>reference</td>\n",
       "      <td>R12L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64881</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>GCGAGGGGAGAAGTGG</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.612939</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V V16I I17K L23R L30T L34E S40K S42H V45Y R48D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9091</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>CCTGAACACTTAGCCT</td>\n",
       "      <td>1</td>\n",
       "      <td>ATC28CAA CAC35ATC</td>\n",
       "      <td>I28Q H35I</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6.261448</td>\n",
       "      <td>0.006931</td>\n",
       "      <td>1.004815</td>\n",
       "      <td>reference</td>\n",
       "      <td>I28Q H35I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59455</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>CGACTACCAAATAATA</td>\n",
       "      <td>3</td>\n",
       "      <td>CGT15GAA</td>\n",
       "      <td>R15E</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5.614693</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>1.000006</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V R15E V16I I17K L23R L30T L34E S40K S42H V4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72388</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>TGCCCATTGTTACCAC</td>\n",
       "      <td>1</td>\n",
       "      <td>CGT15GCA AAG17TTT</td>\n",
       "      <td>R15A K17F</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.917020</td>\n",
       "      <td>-2.948603</td>\n",
       "      <td>0.129533</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V R15A V16I I17F L23R L30T L34E S40K S42H V4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13769</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>GATCACTTGGATACTG</td>\n",
       "      <td>1</td>\n",
       "      <td>TGC19ATG AAC21ACA AGT40CAC</td>\n",
       "      <td>C19M N21T S40H</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>-2.593791</td>\n",
       "      <td>-3.817305</td>\n",
       "      <td>0.070938</td>\n",
       "      <td>reference</td>\n",
       "      <td>C19M N21T S40H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64131</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>GCAAGGGAACCCATGC</td>\n",
       "      <td>3</td>\n",
       "      <td>AAA14GTA ATG31CAG</td>\n",
       "      <td>K14V M31Q</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4.753101</td>\n",
       "      <td>-0.007126</td>\n",
       "      <td>0.995073</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V K14V V16I I17K L23R L30T M31Q L34E S40K S4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61828</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>CTGCGCTTTAGTGTTC</td>\n",
       "      <td>3</td>\n",
       "      <td>AGA1GCC TAC8CAG</td>\n",
       "      <td>R1A Y8Q</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-2.638819</td>\n",
       "      <td>-3.881337</td>\n",
       "      <td>0.067858</td>\n",
       "      <td>H2</td>\n",
       "      <td>R1A S2V Y8Q V16I I17K L23R L30T L34E S40K S42H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57316</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>CAGTGCTTCGTAGCCA</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5.612939</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V V16I I17K L23R L30T L34E S40K S42H V45Y R48D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74256</th>\n",
       "      <td>lib_1</td>\n",
       "      <td>TTGAAACATGATGATC</td>\n",
       "      <td>2</td>\n",
       "      <td>TAC8TGC TTA18CTA ATC28GCC GAA34GCT</td>\n",
       "      <td>Y8C I28A E34A</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>-10.229315</td>\n",
       "      <td>-9.914495</td>\n",
       "      <td>0.001036</td>\n",
       "      <td>H2</td>\n",
       "      <td>S2V Y8C V16I I17K L23R I28A L30T L34A S40K S42...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15000 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      library           barcode  variant_call_support  \\\n",
       "13743   lib_1  GATATCGCTACTGACA                     1   \n",
       "64881   lib_1  GCGAGGGGAGAAGTGG                     1   \n",
       "9091    lib_1  CCTGAACACTTAGCCT                     1   \n",
       "59455   lib_1  CGACTACCAAATAATA                     3   \n",
       "72388   lib_1  TGCCCATTGTTACCAC                     1   \n",
       "...       ...               ...                   ...   \n",
       "13769   lib_1  GATCACTTGGATACTG                     1   \n",
       "64131   lib_1  GCAAGGGAACCCATGC                     3   \n",
       "61828   lib_1  CTGCGCTTTAGTGTTC                     3   \n",
       "57316   lib_1  CAGTGCTTCGTAGCCA                     2   \n",
       "74256   lib_1  TTGAAACATGATGATC                     2   \n",
       "\n",
       "                      codon_substitutions aa_substitutions  \\\n",
       "13743                            CGG12TTG             R12L   \n",
       "64881                                 NaN                    \n",
       "9091                    ATC28CAA CAC35ATC        I28Q H35I   \n",
       "59455                            CGT15GAA             R15E   \n",
       "72388                   CGT15GCA AAG17TTT        R15A K17F   \n",
       "...                                   ...              ...   \n",
       "13769          TGC19ATG AAC21ACA AGT40CAC   C19M N21T S40H   \n",
       "64131                   AAA14GTA ATG31CAG        K14V M31Q   \n",
       "61828                     AGA1GCC TAC8CAG          R1A Y8Q   \n",
       "57316                                 NaN                    \n",
       "74256  TAC8TGC TTA18CTA ATC28GCC GAA34GCT    Y8C I28A E34A   \n",
       "\n",
       "       n_codon_substitutions  n_aa_substitutions  latent_phenotype  \\\n",
       "13743                      1                   1          3.469912   \n",
       "64881                      0                   0          5.612939   \n",
       "9091                       2                   2          6.261448   \n",
       "59455                      1                   1          5.614693   \n",
       "72388                      2                   2         -1.917020   \n",
       "...                      ...                 ...               ...   \n",
       "13769                      3                   3         -2.593791   \n",
       "64131                      2                   2          4.753101   \n",
       "61828                      2                   2         -2.638819   \n",
       "57316                      0                   0          5.612939   \n",
       "74256                      4                   3        -10.229315   \n",
       "\n",
       "       observed_phenotype  observed_enrichment    homolog  \\\n",
       "13743           -0.034489             0.976378  reference   \n",
       "64881            0.000000             1.000000         H2   \n",
       "9091             0.006931             1.004815  reference   \n",
       "59455            0.000009             1.000006         H2   \n",
       "72388           -2.948603             0.129533         H2   \n",
       "...                   ...                  ...        ...   \n",
       "13769           -3.817305             0.070938  reference   \n",
       "64131           -0.007126             0.995073         H2   \n",
       "61828           -3.881337             0.067858         H2   \n",
       "57316            0.000000             1.000000         H2   \n",
       "74256           -9.914495             0.001036         H2   \n",
       "\n",
       "                                             var_wrt_ref  \n",
       "13743                                               R12L  \n",
       "64881   S2V V16I I17K L23R L30T L34E S40K S42H V45Y R48D  \n",
       "9091                                           I28Q H35I  \n",
       "59455  S2V R15E V16I I17K L23R L30T L34E S40K S42H V4...  \n",
       "72388  S2V R15A V16I I17F L23R L30T L34E S40K S42H V4...  \n",
       "...                                                  ...  \n",
       "13769                                     C19M N21T S40H  \n",
       "64131  S2V K14V V16I I17K L23R L30T M31Q L34E S40K S4...  \n",
       "61828  R1A S2V Y8Q V16I I17K L23R L30T L34E S40K S42H...  \n",
       "57316   S2V V16I I17K L23R L30T L34E S40K S42H V45Y R48D  \n",
       "74256  S2V Y8C V16I I17K L23R I28A L30T L34A S40K S42...  \n",
       "\n",
       "[15000 rows x 12 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(X, y), updated_sim_func_score_df = create_homolog_modeling_data(\n",
    "                            simulated_dataset_lib1, \n",
    "                            homologs, \n",
    "                            \"homolog\", \n",
    "                            \"aa_substitutions\",\n",
    "                            \"observed_enrichment\"\n",
    "                        )\n",
    "print(f\"Modeling Input Data: {X}\")\n",
    "print(f\"Modeling Target Data: {y}\")\n",
    "\n",
    "updated_sim_func_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "149364d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of columns in the sparse arrays are the number of beta/shift params to fit\n",
    "params = initialize_model_params(homologs, n_beta_shift_params = X[\"reference\"].shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9ae2667d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parameter β has shape: (1000,)\n",
      "Parameter S_reference has shape: (1000,)\n",
      "Parameter S_H2 has shape: (1000,)\n",
      "Parameter C_ref has shape: (1,)\n",
      "Parameter sig_stretch_x has shape: (1,)\n",
      "Parameter sig_shift_x has shape: (1,)\n",
      "Parameter sig_stretch_y has shape: (1,)\n",
      "Parameter sig_shift_y has shape: (1,)\n"
     ]
    }
   ],
   "source": [
    "for key, value in params.items():\n",
    "    if key == \"α\":\n",
    "        for key_a, value_a in value.items():\n",
    "            print(f\"Parameter {key_a} has shape: {value_a.shape}\")        \n",
    "    else:\n",
    "        print(f\"Parameter {key} has shape: {value.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1c029eaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DeviceArray(1.24799972, dtype=float64)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cost_smooth(params, (X, y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3d603a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "tol = 1e-6\n",
    "maxiter = 10\n",
    "solver = jaxopt.GradientDescent(cost_smooth, tol=tol, maxiter=maxiter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4406d759",
   "metadata": {},
   "outputs": [],
   "source": [
    "params, state = solver.run(params, data=(X, y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ce499789",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DeviceArray(0.20697753, dtype=float64)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cost_smooth(params, (X, y))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98cee907",
   "metadata": {},
   "source": [
    "## Empirical Data (TODO)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
