{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7f0224-78d0-4cfc-ac06-48c1d1b4dfe2",
   "metadata": {},
   "source": [
    "# multidms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc224f4-d5ca-49a6-ae17-d3f8c7fac997",
   "metadata": {},
   "source": [
    "## Overview of model\n",
    "\n",
    "The `multidms` model applies to a case where you have DMS datasets for two or more homologs and are interested in identifying shifts in mutational effects between homologs.\n",
    "To do so, the model defines one homolog as a reference homolog.\n",
    "For each mutation, the model fits one parameter that quantifies the effect of the mutation in the reference homolog.\n",
    "For each non-reference homolog, it also fits a shift parameter that quantifies the shift in the mutation's effect in the non-reference homolog relative to the reference.\n",
    "Shift parameters can be regularized, encouraging most of them to be close to zero.\n",
    "This regularization step is a useful way to eliminate the effects of experimental noise, and is most useful in cases where you expect most mutations to have the same effects between homologs, such as for homologs that are close relatives.\n",
    "\n",
    "The model uses a global-epistasis function to disentangle the effects of multiple mutations on the same variant.\n",
    "To do so, it assumes that mutational effects additively influence a latent biophysical property the protein (e.g., $\\Delta G$ of folding).\n",
    "The mutational-effect parameters described above operate at this latent level.\n",
    "The global-epistasis function then assumes a sigmoidal relationship between a protein's latent property and its functional score measured in the experiment (e.g., log enrichment score).\n",
    "Ultimately, mutaitonal parameters, as well as ones controlling the shape of the sigmoid, are all jointly fit to maximize agreement between predicted and observed functional scores acorss all variants of all homologs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd6382d-47a3-49d1-ac4f-a909f7ebf8a8",
   "metadata": {},
   "source": [
    "## Detailed description of the model\n",
    "\n",
    "For each variant $v$ from homolog $h$, we use a global-epistasis function $g$ to convert a latent phenotype $\\phi$ to a functional score $f$:\n",
    "\n",
    "$$f(v,h) = g_{\\alpha}(\\phi(v,h))$$\n",
    "\n",
    "where $g$ is a sigmoid and $\\alpha$ is a set of parameters encoding the sigmoid.\n",
    "\n",
    "The latent phenotype is computed in the following way:\n",
    "\n",
    "$$\\phi(v,h) = c_h + \\sum_{m \\in v} (x_m + s_{m,h})$$\n",
    "\n",
    "where:\n",
    "* $c_h$ is the wildtype latent phenotype for homolog $h$\n",
    "* $x_m$ is the latent phenotypic effect of mutation $m$. See details below.\n",
    "* $s_{m,h}$ is the shift of the effect of mutation $m$ in homolog $h$. These parameters are fixed to zero for the reference homolog. For non-reference homologs, they are defined in the same way as $x_m$ parameters.\n",
    "\n",
    "The $x_m$ variable is defined such that mutations are always relative to the reference homolog.\n",
    "This way, each homolog informs the exact same parameters, even at sites that differ in wildtype amino acid.\n",
    "For instance, if the wildtype amino acid at site 30 is an A in the reference homolog, but is a Y in a second homolog, then a Y30G mutation in the second homolog is defined as the sum of the following two mutations relative to A: A30Y (negated) and A30G.\n",
    "This approach assumes that mutational effects can be negated, such that A30Y has the opposite effect as Y30A.\n",
    "It also assumes that mutational effects are additive, such that the effect of Y30G is the sum of the effects of Y30A and A30G.\n",
    "\n",
    "The below expression defines this approach more explicitly for an arbitrary site. It uses the notation $x_{\\mathtt{X},n,\\mathtt{Z}}$ where $\\mathtt{X}$ and $\\mathtt{Z}$ are amino acids.\n",
    "For a site $n$, if $aa_{\\text{wt}}$ is the site's wildtype amino acid in a non-reference homolog, $aa_{\\text{mut}}$ is a mutant amino acid in a variant of that homolog, and $aa_{\\text{ref}}$ is the site's wildtype amino acid in the reference homolog, then $x_m$ is:\n",
    "\n",
    "$$x_{aa_{\\text{wt}},n,aa_{\\text{mut}}} = \\begin{cases}\n",
    "      x_{aa_{\\text{ref}},n,aa_{\\text{mut}}} & \\text{if } aa_{\\text{wt}} = aa_{\\text{ref}}\\\\\n",
    "      -x_{aa_{\\text{ref}},n,aa_{\\text{wt}}} & \\text{if } aa_{\\text{mut}} = aa_{\\text{ref}}\\\\\n",
    "      - x_{aa_{\\text{ref}},n,aa_{\\text{wt}}} + x_{aa_{\\text{ref}},n,aa_{\\text{mut}}} & \\text{otherwise}\\\\\n",
    "\\end{cases}$$\n",
    "\n",
    "In this way, we can express the whole model with reference only to coefficients that differ from the reference homolog.\n",
    "\n",
    "Ultimately, we fit parameters using a loss function with one term that scores differences between predicted and observed values and another that uses L1 regularization to penalize non-zero $s_{m,h}$ values:\n",
    "\n",
    "$$ L_{\\text{total}} = \\sum_{h} \\left[\\sum_{v} L_{\\text{fit}}(y_{v,h}, f(v,h)) + \\lambda \\sum_{m} |s_{m,h}|\\right]$$\n",
    "\n",
    "where:\n",
    "* $L_{\\text{total}}$ is the total loss function\n",
    "* $L_{\\text{fit}}$ is a loss function that penalizes differences in predicted vs. observed functional scores\n",
    "* $y_{v,h}$ is the experimentally measured functional score of variant $v$ from homolog $h$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e943741e-4fd1-47d8-b330-ec1713f76a42",
   "metadata": {},
   "source": [
    "## Model using matrix algebra\n",
    "\n",
    "We compute a vector or predicted latent phenotypes $P_{h}$ as:\n",
    "\n",
    "$$P_{h} = C_h + (W_h \\cdot (X + S_h))$$\n",
    "\n",
    "where:\n",
    "* $X$ is a vector of all $x_m$ values\n",
    "* $S_h$ is a matrix of all $s_{m,h}$ values\n",
    "* $W_h$ is a sparse matrix, where rows are variants, columns are mutations (all defined relative to the reference homolog), and values are weights of 0, 1, or -1. These weights are used to compute the phenotype of each variant given the mutations present.\n",
    "* $C_h$ is a vector of the homolog's $c_h$ value, repeated $n$ times, where $n$ is the number of variants.\n",
    "\n",
    "In the matrix algebra, the sum of $X_m$ and $S_{m,h}$ gives a vector of mutational effects, with one entry per mutation.\n",
    "Multiplying the matrix $W_h$ by this vector gives a new vector with one entry per variant, where values are the sum of mutational effects, weighted by the variant-specific weights in $W_h$.\n",
    "Adding $C_h$ values to this vector will give a vector of predicted latent phenotypes for each variant.\n",
    "\n",
    "Next, the global-epistasis function can be used to convert a vector of predicted latent phenotypes to a vector of predicted functional scores.\n",
    "\n",
    "$$F_{h,pred} = g_{\\alpha}(P_h)$$\n",
    "\n",
    "Finally, this vector could be fed into a loss function and compared with a vector of observed functional scores.\n",
    "\n",
    "Quesitons\n",
    "* how tell JAX to fix $s_m,h$ parameters at zero for reference homolog?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74cba1d7-1efb-47cd-a01a-effa499142e6",
   "metadata": {},
   "source": [
    "## Import Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3e5b4c44-179d-4211-abec-24450f3c9254",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import re\n",
    "import jax\n",
    "from jax.experimental import sparse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a1492e-4097-47f9-8605-6197006d7932",
   "metadata": {},
   "source": [
    "Make a test case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9da79fa5-896a-4e91-9e27-591b411072c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   homolog  variant  log2E\n",
       "0        1      M1E    2.0\n",
       "1        1      G2R   -7.0\n",
       "2        1      G2P   -0.5\n",
       "3        2      M1E    2.3\n",
       "4        2      P2R   -5.0\n",
       "5        2      P2G    0.4\n",
       "6        2  M1E P2G    2.7"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [1,1,1,2,2,2,2],\n",
    "    'variant' : ['M1E', 'G2R', 'G2P', 'M1E', 'P2R', 'P2G', 'M1E P2G'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, -5, 0.4, 2.7],\n",
    "}\n",
    "test_df = pandas.DataFrame(test_dict)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b404fb9d-b806-4479-aa76-6b86f3c296fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_weight_matrix(variants, ref_aas):\n",
    "    \"\"\"\n",
    "    Make a sparse matrix of weights associated with particular\n",
    "    variants (rows) and mutations (columns).\n",
    "    \n",
    "    Args:\n",
    "    `variants`: a list of variants, with string values of `XnY`,\n",
    "        where `X` and `Y` are amino acids and `n` is the site number\n",
    "    `ref_aas`: a dictionary with sites as keys and the wildtype\n",
    "        amino acid of the reference homolog as values\n",
    "    \n",
    "    Retruns: A tupple of: i) the sparse matrix and ii) a list of\n",
    "        all unique mutations observed across all variants\n",
    "    \"\"\"\n",
    "    \n",
    "    # Loop over each variant and record mutations, as well as\n",
    "    # weights and indices for making the sparse matrix\n",
    "    all_mutations = []\n",
    "    weights = []\n",
    "    indices = []\n",
    "    for (i, variant) in enumerate(variants):\n",
    "        mutations = set(variant.split())\n",
    "        for mutation in mutations:\n",
    "\n",
    "            # Parse the wt_aa, site, and mut_aa\n",
    "            pattern = r'(?P<wt_aa>\\w)(?P<site>\\d+)(?P<mut_aa>\\w)'\n",
    "            match = re.search(pattern, mutation)\n",
    "            assert match != None, mutation\n",
    "            wt_aa = match.group('wt_aa')\n",
    "            site = match.group('site')\n",
    "            mut_aa = match.group('mut_aa')\n",
    "            \n",
    "            # Get the wildtype amino acid of the reference homolog\n",
    "            ref_aa = ref_aas[site]\n",
    "\n",
    "            # If the wildtype amino acid matches the reference, then\n",
    "            # assign a weight of 1 to the mutation and record the\n",
    "            # mutation's indices in a sparse matrix\n",
    "            if wt_aa == ref_aa:\n",
    "                if mutation not in all_mutations:\n",
    "                    all_mutations.append(mutation)\n",
    "                j = all_mutations.index(mutation)\n",
    "                weights.append(1)\n",
    "                indices.append([i, j])\n",
    "\n",
    "            # Otherwise, decompose the mutation into 1-2 mutations\n",
    "            # relative to the reference sequence, depending on the\n",
    "            # context\n",
    "            else:\n",
    "                mut_to_ref = ref_aa + site + wt_aa\n",
    "                if mut_to_ref not in all_mutations:\n",
    "                    all_mutations.append(mut_to_ref)\n",
    "                j = all_mutations.index(mut_to_ref)\n",
    "                weights.append(-1)\n",
    "                indices.append([i, j])\n",
    "                if ref_aa != mut_aa:\n",
    "                    mut_from_ref = ref_aa + site + mut_aa\n",
    "                    if mut_from_ref not in all_mutations:\n",
    "                        all_mutations.append(mut_from_ref)\n",
    "                    j = all_mutations.index(mut_from_ref)\n",
    "                    weights.append(1)\n",
    "                    indices.append([i, j])\n",
    "\n",
    "    # Make a sparse matrix from the above weights and indices\n",
    "    #assert len(indices) == len(set(indices))\n",
    "    Wm = jax.experimental.sparse.BCOO(\n",
    "        (weights, indices),\n",
    "        shape=(len(variants), len(all_mutations))\n",
    "    )\n",
    "    \n",
    "    return (Wm, all_mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b9b389c9-12bb-4b12-90d8-fa9e1bc5394d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['M1E', 'G2R', 'G2P']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "DeviceArray([[ 1,  0,  0],\n",
       "             [ 0,  1,  0],\n",
       "             [ 0,  0,  1],\n",
       "             [ 1,  0,  0],\n",
       "             [ 0,  1, -1],\n",
       "             [ 0,  0, -1],\n",
       "             [ 1,  0, -1]], dtype=int32)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = list(test_df['variant'])\n",
    "ref_aas = {'1':'M', '2':'G'}\n",
    "#all_mutations = ['M1E', 'G2R', 'G2P']\n",
    "(Wm, all_mutations) = make_weight_matrix(variants, ref_aas)\n",
    "print(all_mutations)\n",
    "Wm.todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cef9eb66-a88d-4019-828f-e00f2d707542",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G2R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G2P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P2R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P2G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P2G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   homolog  variant  log2E\n",
       "0        1      M1E    2.0\n",
       "1        1      G2R   -7.0\n",
       "2        1      G2P   -0.5\n",
       "3        2      M1E    2.3\n",
       "4        2      P2R   -5.0\n",
       "5        2      P2G    0.4\n",
       "6        2  M1E P2G    2.7"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39741a09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b9deaf3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd7bff41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fce04e52",
   "metadata": {},
   "source": [
    "Below cell is work in progress and may be outdated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2500cd8f-19b7-4bac-ab88-2707a8f25630",
   "metadata": {},
   "source": [
    "Implementing the above equations in code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b16d955-44c8-465a-83fa-835a9b1577d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a dataframe where rows are variants and one of the\n",
    "# columns lists the mutations in each variant\n",
    "\n",
    "# Make a list of all possible amino-acid mutations relative\n",
    "# to the wildtype amino acid of the reference homolog\n",
    "assert len(all_mutations) == len(set(all_mutations))\n",
    "\n",
    "# Make weight matrix\n",
    "Wm = make_wt_matrix(df, ref_aas, all_mutations)\n",
    "\n",
    "# `Ch`: float of wildtype latent phenotype of homolog\n",
    "# `Wh`: sparse matrix, where rows are variants, columns are mutations, and values are weights\n",
    "# `Xm`: vector of x_m coefficients for all mutations across all homologs (one vector for all homologs)\n",
    "# `Smh`: vector of s_m,h coefficients for homolog h and mutations m (one vector per homolog)\n",
    "\n",
    "# sum (Xh + Shm) for all mutations m in each variant, using the Wh matrix to identify which\n",
    "# values to add for each variant\n",
    "# then add Ch to each value in the resulting vector\n",
    "# this gives back a vector of the predicted latent phenotype of each variant\n",
    "Lh_pred = Ch + (Wh @ (Xm + Smh))\n",
    "\n",
    "# Use the global-epistasis function to compute a predicted functional score for each variant\n",
    "# alpha contains parameters related to optimizing the sigmoid\n",
    "Fh_pred = g(Lh, alpha)\n",
    "\n",
    "# Compute the Huber loss for homolog h\n",
    "huber_loss_h = jaxopt.loss.huber_loss(Fh_exp, Fh_pred, δ).mean()\n",
    "\n",
    "# Compute the L1 loss for smh params for homolog h\n",
    "L1_loss_h = L1(Smh)\n",
    "\n",
    "# Compute total loss\n",
    "total_loss_h = huber_loss_h + L1_loss_h\n",
    "total_loss += total_loss_h \n",
    "\n",
    "# Optimize params Ch, Xm, Smh, and alpha using the above loss function\n",
    "\n",
    "\n",
    "# To do\n",
    "# look up how to optimize parameters of a sigmoid in JAX"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
