<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>multidms package &#8212; multidms 0.4.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=420b4293" />
    <script src="_static/documentation_options.js?v=6c02275b"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="biophysical" href="multidms.biophysical.html" />
    <link rel="prev" title="multidms usage examples" href="fit_delta_BA1_example.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="module-multidms">
<span id="multidms-package"></span><h1>multidms package<a class="headerlink" href="#module-multidms" title="Link to this heading">¶</a></h1>
<section id="multidms">
<h2>multidms<a class="headerlink" href="#multidms" title="Link to this heading">¶</a></h2>
<p>multidms is a Python package for modeling deep mutational scanning data.
In particular, it is designed to model data from more than one experiment,
even if they don’t share the same wildtype amino acid sequence.
It uses joint modeling to inform parameters across all experiments,
while identifying experiment-specific mutation effects which differ.</p>
<p>Importing this package imports the following objects
into the package namespace:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="multidms.data.html#multidms.data.Data" title="multidms.data.Data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">Data</span></code></a></p></li>
<li><p><a class="reference internal" href="multidms.model.html#multidms.model.Model" title="multidms.model.Model"><code class="xref py py-mod docutils literal notranslate"><span class="pre">Model</span></code></a></p></li>
<li><p><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection" title="multidms.model_collection.ModelCollection"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ModelCollection</span></code></a></p></li>
</ul>
</div></blockquote>
<p>For a brief description about how the <a class="reference internal" href="multidms.model.html#multidms.model.Model" title="multidms.model.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">Model</span></code></a>
class works to compose, compile, and optimize the model parameters
- as well as detailed code code documentation for each of the
equations described in the
<a class="reference external" href="https://matsengrp.github.io/multidms/biophysical_model.html">biophysical docs</a> -
see:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="multidms.biophysical.html#module-multidms.biophysical" title="multidms.biophysical"><code class="xref py py-mod docutils literal notranslate"><span class="pre">biophysical</span></code></a></p></li>
</ul>
</div></blockquote>
<p><a class="reference internal" href="multidms.plot.html#module-multidms.plot" title="multidms.plot"><code class="xref py py-mod docutils literal notranslate"><span class="pre">plot</span></code></a> mostly contains code for interactive plotting
at the moment.</p>
<p>It also imports the following alphabets:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="xref py py-const docutils literal notranslate"><span class="pre">AAS</span></code></p></li>
<li><p><code class="xref py py-const docutils literal notranslate"><span class="pre">AAS_WITHSTOP</span></code></p></li>
<li><p><code class="xref py py-const docutils literal notranslate"><span class="pre">AAS_WITHGAP</span></code></p></li>
<li><p><code class="xref py py-const docutils literal notranslate"><span class="pre">AAS_WITHSTOP_WITHGAP</span></code></p></li>
</ul>
</div></blockquote>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="multidms.Data">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">multidms.</span></span><span class="sig-name descname"><span class="pre">Data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variants_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><span class="pre">DataFrame</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_identical_variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('#0072B2',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#009E73',</span> <span class="pre">'#17BECF',</span> <span class="pre">'#BCDB22')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter_suffixed_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assert_site_integrity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Data" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Prep and store one-hot encoding of
variant substitutions data.
Individual objects of this type can be shared
by multiple <a class="reference internal" href="#multidms.Model" title="multidms.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">multidms.Model</span></code></a> Objects
for efficiently fitting various models to the same data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can initialize a <a class="reference internal" href="#multidms.Data" title="multidms.Data"><code class="xref py py-class docutils literal notranslate"><span class="pre">Data</span></code></a> object with a <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a>
with a row for each variant sampled and annotations
provided in the required columns:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><cite>condition</cite> - Experimental condition from</dt><dd><p>which a sample measurement was obtained.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>aa_substitutions</cite> - Defines each variant</dt><dd><p><span class="math notranslate nohighlight">\(v\)</span> as a string of substitutions (e.g., <code class="docutils literal notranslate"><span class="pre">'M3A</span> <span class="pre">K5G'</span></code>).
Note that while conditions may have differing wild types
at a given site, the sites between conditions should reference
the same site when alignment is performed between
condition wild types.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>func_score</cite> - The functional score computed from experimental</dt><dd><p>measurements.</p>
</dd>
</dl>
</li>
</ol>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants_df</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> or None) – The variant level information from all experiments you
wish to analyze. Should have columns named <code class="docutils literal notranslate"><span class="pre">'condition'</span></code>,
<code class="docutils literal notranslate"><span class="pre">'aa_substitutions'</span></code>, and <code class="docutils literal notranslate"><span class="pre">'func_score'</span></code>.
See the class note for descriptions of each of the features.</p></li>
<li><p><strong>reference</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Name of the condition which annotates the reference.
variants. Note that for model fitting this class will convert all
amino acid substitutions for non-reference condition groups
to relative to the reference condition.
For example, if the wild type amino acid at site 30 is an
A in the reference condition, and a G in a non-reference condition,
then a Y30G mutation in the non-reference condition is recorded as an A30G
mutation relative to the reference. This way, each condition informs
the exact same parameters, even at sites that differ in wild type amino acid.
These are encoded in a <a class="reference external" href="https://jbloomlab.github.io/binarymap/binarymap.binarymap.html#binarymap.binarymap.BinaryMap" title="(in binarymap)"><code class="xref py py-class docutils literal notranslate"><span class="pre">binarymap.binarymap.BinaryMap</span></code></a> object for each
condition,
where all sites that are non-identical to the reference are 1’s.
For motivation, see the <cite>Model overview</cite> section in <a class="reference internal" href="#multidms.Model" title="multidms.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">multidms.Model</span></code></a>
class notes.</p></li>
<li><p><strong>alphabet</strong> (<em>array-like</em>) – Allowed characters in mutation strings.</p></li>
<li><p><strong>collapse_identical_variants</strong> (<em>{'mean'</em><em>, </em><em>'median'</em><em>, </em><em>False}</em>) – If identical variants in <code class="docutils literal notranslate"><span class="pre">variants_df</span></code> (same ‘aa_substitutions’),
exist within individual condition groups,
collapse them by taking mean or median of ‘func_score’, or
(if <cite>False</cite>) do not collapse at all. Collapsing will make fitting faster,
but <em>not</em> a good idea if you are doing bootstrapping.</p></li>
<li><p><strong>condition_colors</strong> (<em>array-like</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Maps each condition to the color used for plotting. Either a dict keyed
by each condition, or an array of colors that are sequentially assigned
to the conditions.</p></li>
<li><p><strong>letter_suffixed_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – True if sites are sequential and integer, False otherwise.</p></li>
<li><p><strong>assert_site_integrity</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, will assert that all sites in the data frame
have the same wild type amino acid, grouped by condition.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, will print progress bars.</p></li>
<li><p><strong>nb_workers</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Number of workers to use for parallel operations.
If None, will use all available CPUs.</p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>None</em>) – Name of the data object. If None, will be assigned
a unique name based upon the number of data objects
instantiated.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Simple example with two conditions (<code class="docutils literal notranslate"><span class="pre">'a'</span></code> and <code class="docutils literal notranslate"><span class="pre">'b'</span></code>)</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">multidms</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_data</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;condition&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span><span class="s2">&quot;a&quot;</span><span class="p">,</span><span class="s2">&quot;a&quot;</span><span class="p">,</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">,</span><span class="s2">&quot;b&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;aa_substitutions&#39;</span> <span class="p">:</span> <span class="p">[</span>
<span class="gp">... </span>        <span class="s1">&#39;M1E&#39;</span><span class="p">,</span> <span class="s1">&#39;G3R&#39;</span><span class="p">,</span> <span class="s1">&#39;G3P&#39;</span><span class="p">,</span> <span class="s1">&#39;M1W&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;P3R&#39;</span><span class="p">,</span> <span class="s1">&#39;P3G&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E P3G&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E P3R&#39;</span><span class="p">,</span> <span class="s1">&#39;P2T&#39;</span>
<span class="gp">... </span>    <span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;func_score&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">7</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">2.3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">2.7</span><span class="p">,</span> <span class="o">-</span><span class="mf">2.7</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">func_score_data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_df</span>  
<span class="go">condition aa_substitutions  func_score</span>
<span class="go">0         a              M1E         2.0</span>
<span class="go">1         a              G3R        -7.0</span>
<span class="go">2         a              G3P        -0.5</span>
<span class="go">3         a              M1W         2.3</span>
<span class="go">4         b              M1E         1.0</span>
<span class="go">5         b              P3R        -5.0</span>
<span class="go">6         b              P3G         0.4</span>
<span class="go">7         b          M1E P3G         2.7</span>
<span class="go">8         b          M1E P3R        -2.7</span>
<span class="go">9         b              P2T         0.3</span>
</pre></div>
</div>
<p>Instantiate a <code class="docutils literal notranslate"><span class="pre">Data</span></code> Object allowing for stop codon variants
and declaring condition <cite>“a”</cite> as the reference condition.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">Data</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">func_score_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">alphabet</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">AAS_WITHSTOP</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span>  
<span class="gp">...</span>
</pre></div>
</div>
<p>Note this may take some time due to the string
operations that must be performed when converting
amino acid substitutions to be with respect to a
reference wild type sequence.</p>
<p>After the object has finished being instantiated,
we now have access to a few ‘static’ properties
of our data. See individual property docstring
for more information.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">reference</span>
<span class="go">&#39;a&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">conditions</span>
<span class="go">(&#39;a&#39;, &#39;b&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1E&#39;, &#39;M1W&#39;, &#39;G3P&#39;, &#39;G3R&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">site_map</span>  
<span class="go">a  b</span>
<span class="go">1  M  M</span>
<span class="go">3  G  P</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">mutations_df</span>  
<span class="go">  mutation wts  sites muts  times_seen_a  times_seen_b</span>
<span class="go">0      M1E   M      1    E             1           3.0</span>
<span class="go">1      M1W   M      1    W             1           0.0</span>
<span class="go">2      G3P   G      3    P             1           1.0</span>
<span class="go">3      G3R   G      3    R             1           2.0</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">variants_df</span>  
<span class="go">  condition aa_substitutions  func_score var_wrt_ref</span>
<span class="go">0         a              M1E         2.0         M1E</span>
<span class="go">1         a              G3R        -7.0         G3R</span>
<span class="go">2         a              G3P        -0.5         G3P</span>
<span class="go">3         a              M1W         2.3         M1W</span>
<span class="go">4         b              M1E         1.0     G3P M1E</span>
<span class="go">5         b              P3R        -5.0         G3R</span>
<span class="go">6         b              P3G         0.4</span>
<span class="go">7         b          M1E P3G         2.7         M1E</span>
<span class="go">8         b          M1E P3R        -2.7     G3R M1E</span>
</pre></div>
</div>
<dl class="py attribute">
<dt class="sig sig-object py" id="multidms.Data.counter">
<span class="sig-name descname"><span class="pre">counter</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#multidms.Data.counter" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#multidms.Data.name" title="Link to this definition">¶</a></dt>
<dd><p>The name of the data object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.conditions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">conditions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><span class="pre">tuple</span></a></em><a class="headerlink" href="#multidms.Data.conditions" title="Link to this definition">¶</a></dt>
<dd><p>A tuple of all conditions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#multidms.Data.reference" title="Link to this definition">¶</a></dt>
<dd><p>The name of the reference condition.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><span class="pre">tuple</span></a></em><a class="headerlink" href="#multidms.Data.mutations" title="Link to this definition">¶</a></dt>
<dd><p>A tuple of all mutations in the order relative to their index into
the binarymap.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.mutations_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_df</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><span class="pre">DataFrame</span></a></em><a class="headerlink" href="#multidms.Data.mutations_df" title="Link to this definition">¶</a></dt>
<dd><p>A dataframe summarizing all single mutations</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.variants_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants_df</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><span class="pre">DataFrame</span></a></em><a class="headerlink" href="#multidms.Data.variants_df" title="Link to this definition">¶</a></dt>
<dd><p>A dataframe summarizing all variants in the training data.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.site_map">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">site_map</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><span class="pre">DataFrame</span></a></em><a class="headerlink" href="#multidms.Data.site_map" title="Link to this definition">¶</a></dt>
<dd><p>A dataframe indexed by site, with columns
for all conditions giving the wild type amino acid
at each site.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.non_identical_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">non_identical_mutations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Data.non_identical_mutations" title="Link to this definition">¶</a></dt>
<dd><p>A dictionary keyed by condition names with values
being a string of all mutations that differ from the
reference sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.non_identical_sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">non_identical_sites</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Data.non_identical_sites" title="Link to this definition">¶</a></dt>
<dd><p>A dictionary keyed by condition names with values
being a <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a> indexed by site,
with columns for the reference
and non-reference amino acid at each site that differs.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.reference_sequence_conditions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_sequence_conditions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.12)"><span class="pre">list</span></a></em><a class="headerlink" href="#multidms.Data.reference_sequence_conditions" title="Link to this definition">¶</a></dt>
<dd><p>A list of conditions that have the same wild type
sequence as the reference condition.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.training_data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">training_data</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Data.training_data" title="Link to this definition">¶</a></dt>
<dd><p>A dictionary with keys ‘X’ and ‘y’ for the training data.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.binarymaps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">binarymaps</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Data.binarymaps" title="Link to this definition">¶</a></dt>
<dd><p>A dictionary keyed by condition names with values
being a <code class="docutils literal notranslate"><span class="pre">BinaryMap</span></code> object for each condition.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.targets">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">targets</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Data.targets" title="Link to this definition">¶</a></dt>
<dd><p>The functional scores for each variant in the training data.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.mutparser">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutparser</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">MutationParser</span></em><a class="headerlink" href="#multidms.Data.mutparser" title="Link to this definition">¶</a></dt>
<dd><p>The mutation <code class="docutils literal notranslate"><span class="pre">polyclonal.utils.MutationParser</span></code> used
to parse mutations.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.parse_mut">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_mut</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">MutationParser</span></em><a class="headerlink" href="#multidms.Data.parse_mut" title="Link to this definition">¶</a></dt>
<dd><p>returns a function that splits a single amino acid substitutions
into wildtype, site, and mutation
using the mutation parser.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Data.parse_muts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_muts</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">partial</span></em><a class="headerlink" href="#multidms.Data.parse_muts" title="Link to this definition">¶</a></dt>
<dd><p>A function that splits amino acid substitutions
(a string of more than one) into wildtype, site, and mutation
using the mutation parser.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Data.convert_subs_wrt_ref_seq">
<span class="sig-name descname"><span class="pre">convert_subs_wrt_ref_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_subs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Data.convert_subs_wrt_ref_seq" title="Link to this definition">¶</a></dt>
<dd><p>Covert amino acid substitutions to be with respect to the reference sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>condition</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – The condition from which aa substitutions are relative to.</p></li>
<li><p><strong>aa_subs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – A string of amino acid substitutions, relative to the condition sequence,
to converted</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A string of amino acid substitutions relative to the reference sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Data.plot_times_seen_hist">
<span class="sig-name descname"><span class="pre">plot_times_seen_hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Data.plot_times_seen_hist" title="Link to this definition">¶</a></dt>
<dd><p>Plot a histogram of the number of times each mutation was seen.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Data.plot_func_score_boxplot">
<span class="sig-name descname"><span class="pre">plot_func_score_boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Data.plot_func_score_boxplot" title="Link to this definition">¶</a></dt>
<dd><p>Plot a boxplot of the functional scores for each condition.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="multidms.Model">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">multidms.</span></span><span class="sig-name descname"><span class="pre">Model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data:</span> <span class="pre">~multidms.data.Data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epistatic_model=&lt;function</span> <span class="pre">sigmoidal_global_epistasis&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_activation=&lt;function</span> <span class="pre">identity_activation&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional_shifts=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha_d=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma_corrected=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PRNGKey=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_hidden_units=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_beta_naught=0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_theta_scale=5.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_theta_bias=-5.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Represent one or more DMS experiments
to obtain tuned parameters that provide insight into
individual mutational effects and conditional shifts
of those effects on all non-reference conditions.
For more see the biophysical model documentation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<a class="reference internal" href="#multidms.Data" title="multidms.Data"><em>multidms.Data</em></a>) – A reference to the dataset which will define the parameters
of the model to be fit.</p></li>
<li><p><strong>epistatic_model</strong> (<em>&lt;class 'function'&gt;</em>) – A function which will transform the latent
effects of mutations into a functional score.
See the biophysical model documentation
for more.</p></li>
<li><p><strong>output_activation</strong> (<em>&lt;class 'function'&gt;</em>) – A function which will transform the output of the
global epistasis function. Defaults to the identity function
(no activation). See the biophysical model documentation</p></li>
<li><p><strong>conditional_shifts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If true (default) initialize and fit the shift
parameters for each non-reference condition.
See Model Description section for more.
Defaults to True.</p></li>
<li><p><strong>alpha_d</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True introduce a latent offset parameter
for each condition.
See the biophysical docs section for more.
Defaults to True.</p></li>
<li><p><strong>gamma_corrected</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If true (default), introduce the ‘gamma’ parameter
for each non-reference parameter to
account for differences between wild type
behavior relative to its variants. This
is essentially a bias added to the functional
scores during fitting.
See Model Description section for more.
Defaults to False.</p></li>
<li><p><strong>PRNGKey</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – The initial seed key for random parameters
assigned to Betas and any other randomly
initialized parameters.
for more.</p></li>
<li><p><strong>init_beta_naught</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Initialize the latent offset parameter
applied to all conditions.
See the biophysical docs section for more.</p></li>
<li><p><strong>init_theta_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Initialize the scaling parameter <span class="math notranslate nohighlight">\(\theta_{\text{scale}}\)</span> of
a two-parameter epistatic model (Sigmoid or Softplus).</p></li>
<li><p><strong>init_theta_bias</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Initialize the bias parameter <span class="math notranslate nohighlight">\(\theta_{\text{bias}}\)</span> of
a two parameter epistatic model (Sigmoid or Softplus).</p></li>
<li><p><strong>n_hidden_units</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em> or </em><em>None</em>) – If using <a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.nn_global_epistasis" title="multidms.biophysical.nn_global_epistasis"><code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.biophysical.nn_global_epistasis()</span></code></a>
as the epistatic model, this is the number of hidden units
used in the transform.</p></li>
<li><p><strong>lower_bound</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a><em> or </em><em>None</em>) – If using <a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.softplus_activation" title="multidms.biophysical.softplus_activation"><code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.biophysical.softplus_activation()</span></code></a>
as the output activation, this is the lower bound of the
softplus function.</p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>None</em>) – Name of the Model object. If None, will be assigned
a unique name based upon the number of data objects
instantiated.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<p>To create a <a class="reference internal" href="#multidms.Model" title="multidms.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">Model</span></code></a> object, all you need is
the respective <a class="reference internal" href="#multidms.Data" title="multidms.Data"><code class="xref py py-class docutils literal notranslate"><span class="pre">Data</span></code></a> object for parameter fitting.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">multidms</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tests.test_data</span> <span class="kn">import</span> <span class="n">data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">Model</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<p>Upon initialization, you will now have access to the underlying data
and parameters.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1E&#39;, &#39;M1W&#39;, &#39;G3P&#39;, &#39;G3R&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">conditions</span>
<span class="go">(&#39;a&#39;, &#39;b&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">reference</span>
<span class="go">&#39;a&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">condition_colors</span>
<span class="go">{&#39;a&#39;: &#39;#0072B2&#39;, &#39;b&#39;: &#39;#CC79A7&#39;}</span>
</pre></div>
</div>
<p>The mutations_df and variants_df may of course also be accessed.
First, we set pandas to display all rows and columns.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">mutations_df</span>  
<span class="go">  mutation wts  sites muts  times_seen_a  times_seen_b</span>
<span class="go">0      M1E   M      1    E             1           3.0</span>
<span class="go">1      M1W   M      1    W             1           0.0</span>
<span class="go">2      G3P   G      3    P             1           1.0</span>
<span class="go">3      G3R   G      3    R             1           2.0</span>
</pre></div>
</div>
<p>However, if accessed directly through the <a class="reference internal" href="#multidms.Model" title="multidms.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">Model</span></code></a> object, you will
get the same information, along with model/parameter specific
features included. These are automatically updated each time you
request the property.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">get_mutations_df</span><span class="p">()</span>  
<span class="go">              beta  shift_b  predicted_func_score_a  predicted_func_score_b  \</span>
<span class="go">mutation</span>
<span class="go">M1E       1.816086      0.0                1.800479                1.379661</span>
<span class="go">M1W      -0.754885      0.0               -0.901211               -1.322029</span>
<span class="go">G3P       0.339889      0.0                0.420818                0.000000</span>
<span class="go">G3R      -0.534835      0.0               -0.653051               -1.073869</span>

<span class="go">          times_seen_a  times_seen_b wts  sites muts</span>
<span class="go">mutation</span>
<span class="go">M1E                  1           3.0   M      1    E</span>
<span class="go">M1W                  1           0.0   M      1    W</span>
<span class="go">G3P                  1           1.0   G      3    P</span>
<span class="go">G3R                  1           2.0   G      3    R</span>
</pre></div>
</div>
<p>Notice the respective single mutation effects (<code class="docutils literal notranslate"><span class="pre">&quot;beta&quot;</span></code>), conditional shifts
(<code class="docutils literal notranslate"><span class="pre">shift_d</span></code>),
and predicted functional score (<code class="docutils literal notranslate"><span class="pre">F_d</span></code>) of each mutation in the model are now
easily accessible. Similarly, we can take a look at the variants_df for the model,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">get_variants_df</span><span class="p">()</span>  
<span class="go">  condition aa_substitutions  func_score var_wrt_ref  predicted_latent  \</span>
<span class="go">0         a              M1E         2.0         M1E          1.816086</span>
<span class="go">1         a              G3R        -7.0         G3R         -0.534835</span>
<span class="go">2         a              G3P        -0.5         G3P          0.339889</span>
<span class="go">3         a              M1W         2.3         M1W         -0.754885</span>
<span class="go">4         b              M1E         1.0     G3P M1E          1.816086</span>
<span class="go">5         b              P3R        -5.0         G3R         -0.874724</span>
<span class="go">6         b              P3G         0.4                     -0.339889</span>
<span class="go">7         b          M1E P3G         2.7         M1E          1.476197</span>
<span class="go">8         b          M1E P3R        -2.7     G3R M1E          0.941362</span>

<span class="go">   predicted_func_score</span>
<span class="go">0              1.800479</span>
<span class="go">1             -0.653051</span>
<span class="go">2              0.420818</span>
<span class="go">3             -0.901211</span>
<span class="go">4              1.560311</span>
<span class="go">5             -1.073869</span>
<span class="go">6             -0.420818</span>
<span class="go">7              1.379661</span>
<span class="go">8              0.992495</span>
</pre></div>
</div>
<p>We now have access to the predicted (and gamma corrected) functional scores
as predicted by the models current parameters.</p>
<p>So far, these parameters and predictions results from them have not been tuned
to the dataset. Let’s take a look at the loss on the training dataset
given our initialized parameters</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">loss</span>
<span class="go">Array(4.7124467, dtype=float32)</span>
</pre></div>
</div>
<p>Next, we fit the model with some chosen hyperparameters.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxiter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">lasso_shift</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">,</span> <span class="n">warn_unconverged</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">loss</span>
<span class="go">Array(6.0517805e-06, dtype=float32)</span>
</pre></div>
</div>
<p>The model tunes its parameters in place, and the subsequent call to retrieve
the loss reflects our models loss given its updated parameters.</p>
<p>TODO: add more examples, and explain the convergence criteria and warning.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="multidms.Model.counter">
<span class="sig-name descname"><span class="pre">counter</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#multidms.Model.counter" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#multidms.Model.name" title="Link to this definition">¶</a></dt>
<dd><p>The name of the data object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.params">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">params</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Model.params" title="Link to this definition">¶</a></dt>
<dd><p>All current model parameters in a dictionary.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.state">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">state</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><span class="pre">dict</span></a></em><a class="headerlink" href="#multidms.Model.state" title="Link to this definition">¶</a></dt>
<dd><p>The current state of the model.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.converged">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">converged</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></em><a class="headerlink" href="#multidms.Model.converged" title="Link to this definition">¶</a></dt>
<dd><p>Whether the model tolerance threshold was passed on last fit.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">data</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="multidms.data.html#multidms.data.Data" title="multidms.data.Data"><span class="pre">Data</span></a></em><a class="headerlink" href="#multidms.Model.data" title="Link to this definition">¶</a></dt>
<dd><p>multidms.Data Object this model references for fitting
its parameters.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.model_components">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">model_components</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">frozendict</span></em><a class="headerlink" href="#multidms.Model.model_components" title="Link to this definition">¶</a></dt>
<dd><p>A frozendict which hold the individual components of the model
as well as the objective and forward functions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.loss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">loss</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><span class="pre">float</span></a></em><a class="headerlink" href="#multidms.Model.loss" title="Link to this definition">¶</a></dt>
<dd><p>Compute model loss on all experimental training data
without ridge or lasso penalties included.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.conditional_loss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">conditional_loss</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><span class="pre">float</span></a></em><a class="headerlink" href="#multidms.Model.conditional_loss" title="Link to this definition">¶</a></dt>
<dd><p>Compute loss individually for each condition.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.variants_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants_df</span></span><a class="headerlink" href="#multidms.Model.variants_df" title="Link to this definition">¶</a></dt>
<dd><p>Kept for backwards compatibility but will be removed in future versions.
Please use <cite>get_variants_df</cite> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.get_variants_df">
<span class="sig-name descname"><span class="pre">get_variants_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">phenotype_as_effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.get_variants_df" title="Link to this definition">¶</a></dt>
<dd><p>Training data with model predictions for latent,
and functional score phenotypes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>phenotype_as_effect</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – if True, phenotypes (both latent, and func_score)
are calculated as the _difference_ between predicted
phenotype of a given variant and the respective experimental
wildtype prediction. Otherwise, report the unmodified
model prediction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the training data, <cite>self.data.variants_df</cite>,
with the phenotypes added. Phenotypes are predicted
based on the current state of the model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)">pandas.DataFrame</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.mutations_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_df</span></span><a class="headerlink" href="#multidms.Model.mutations_df" title="Link to this definition">¶</a></dt>
<dd><p>Kept for backwards compatibility but will be removed in future versions.
Please use <cite>get_mutations_df</cite> instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.get_mutations_df">
<span class="sig-name descname"><span class="pre">get_mutations_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">phenotype_as_effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.get_mutations_df" title="Link to this definition">¶</a></dt>
<dd><p>Mutation attributes and phenotypic effects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>phenotype_as_effect</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – if True, phenotypes (both latent, and func_score)
are calculated as the _difference_ between predicted
phenotype of a given variant and the respective experimental
wildtype prediction. Otherwise, report the unmodified
model prediction.</p></li>
<li><p><strong>times_seen_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Only report mutations that have been seen at least
this many times in each condition. Defaults to 0.</p></li>
<li><p><strong>return_split</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If True, return the split mutations as separate columns:
‘wts’, ‘sites’, and ‘muts’.
Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the mutations data, <cite>self.data.mutations_df</cite>,
with the mutations column set as the index, and columns
with the mutational attributes (e.g. betas, shifts) and
conditional phenotypes (e.g. func_scores) added.
Phenotypes are predicted
based on the current state of the model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)">pandas.DataFrame</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.get_df_loss">
<span class="sig-name descname"><span class="pre">get_df_loss</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error_if_unknown</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.get_df_loss" title="Link to this definition">¶</a></dt>
<dd><p>Get the loss of the model on a given data frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><em>pandas.DataFrame</em></a>) – Data frame containing variants. Requirements are the same as
those used to initialize the <cite>multidms.Data</cite> object - except
the indices must be unique.</p></li>
<li><p><strong>error_if_unknown</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If some of the substitutions in a variant are not present in
the model (not in <code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code>)
then by default we do not include those variants
in the loss calculation. If <cite>True</cite>, raise an error.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, print the number of valid and invalid variants.</p></li>
<li><p><strong>conditional</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, return the loss for each condition as a dictionary.
If False, return the total loss.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The loss of the model on the given data frame.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)">float</a> or <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.add_phenotypes_to_df">
<span class="sig-name descname"><span class="pre">add_phenotypes_to_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitutions_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa_substitutions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'condition'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">latent_phenotype_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'predicted_latent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">observed_phenotype_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'predicted_func_score'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">converted_substitutions_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'aa_subs_wrt_ref'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unknown_as_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_as_effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.add_phenotypes_to_df" title="Link to this definition">¶</a></dt>
<dd><p>Add predicted phenotypes to data frame of variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><em>pandas.DataFrame</em></a>) – Data frame containing variants. Requirements are the same as
those used to initialize the <cite>multidms.Data</cite> object - except
the indices must be unique.</p></li>
<li><p><strong>substitutions_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Column in <cite>df</cite> giving variants as substitution strings
with respect to a given variants condition.
These will be converted to be with respect to the reference sequence
prior to prediction. Defaults to ‘aa_substitutions’.</p></li>
<li><p><strong>condition_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Column in <cite>df</cite> giving the condition from which a variant was
observed. Values must exist in the self.data.conditions and
and error will be raised otherwise. Defaults to ‘condition’.</p></li>
<li><p><strong>latent_phenotype_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Column added to <cite>df</cite> containing predicted latent phenotypes.</p></li>
<li><p><strong>observed_phenotype_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Column added to <cite>df</cite> containing predicted observed phenotypes.</p></li>
<li><p><strong>converted_substitutions_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>None</em>) – Columns added to <cite>df</cite> containing converted substitution strings
for non-reference conditions if they do not share a wildtype seq.</p></li>
<li><p><strong>overwrite_cols</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If the specified latent or observed phenotype column already
exist in <cite>df</cite>, overwrite it? If <cite>False</cite>, raise an error.</p></li>
<li><p><strong>unknown_as_nan</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If some of the substitutions in a variant are not present in
the model (not in <code class="xref py py-attr docutils literal notranslate"><span class="pre">AbstractEpistasis.binarymap</span></code>) set the
phenotypes to <cite>nan</cite> (not a number)? If <cite>False</cite>, raise an error.</p></li>
<li><p><strong>phenotype_as_effect</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – if True, phenotypes (both latent, and func_score)
are calculated as the _difference_ between predicted
phenotype of a given variant and the respective experimental
wildtype prediction. Otherwise, report the unmodified
model prediction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of <cite>df</cite> with the phenotypes added. Phenotypes are predicted
based on the current state of the model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)">pandas.DataFrame</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.Model.wildtype_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wildtype_df</span></span><a class="headerlink" href="#multidms.Model.wildtype_df" title="Link to this definition">¶</a></dt>
<dd><p>Get a dataframe indexed by condition wildtype
containing the prediction features for each.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.mutation_site_summary_df">
<span class="sig-name descname"><span class="pre">mutation_site_summary_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">agg_func=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.mutation_site_summary_df" title="Link to this definition">¶</a></dt>
<dd><p>Get all single mutational attributes from self._data
updated with all model specific attributes, then aggregate
all numerical columns by “sites” using
<code class="docutils literal notranslate"><span class="pre">agg</span></code> function. The mean values are given by default.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.get_condition_params">
<span class="sig-name descname"><span class="pre">get_condition_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.get_condition_params" title="Link to this definition">¶</a></dt>
<dd><p>Get the relent parameters for a model prediction</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.phenotype_fromsubs">
<span class="sig-name descname"><span class="pre">phenotype_fromsubs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_subs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.phenotype_fromsubs" title="Link to this definition">¶</a></dt>
<dd><p>take a single string of subs which are
not already converted wrt reference, convert them and
then make a functional score prediction and return the result.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.latent_fromsubs">
<span class="sig-name descname"><span class="pre">latent_fromsubs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_subs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.latent_fromsubs" title="Link to this definition">¶</a></dt>
<dd><p>take a single string of subs which are
not already converted wrt reference, convert them and
then make a latent prediction and return the result.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.phenotype_frombinary">
<span class="sig-name descname"><span class="pre">phenotype_frombinary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.phenotype_frombinary" title="Link to this definition">¶</a></dt>
<dd><p>Condition specific functional score prediction
on X using the biophysical model
given current model parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>jnp.array</em>) – Binary encoded variants to make predictions on.</p></li>
<li><p><strong>condition</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Condition to make predictions for. If None, use the reference</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.latent_frombinary">
<span class="sig-name descname"><span class="pre">latent_frombinary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.latent_frombinary" title="Link to this definition">¶</a></dt>
<dd><p>Condition specific latent phenotype prediction
on X using the biophysical model
given current model parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>jnp.array</em>) – Binary encoded variants to make predictions on.</p></li>
<li><p><strong>condition</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – Condition to make predictions for. If None, use the reference</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.fit_reference_beta">
<span class="sig-name descname"><span class="pre">fit_reference_beta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.fit_reference_beta" title="Link to this definition">¶</a></dt>
<dd><p>Fit the Model beta’s to the reference data.</p>
<p>This is an experimental feature and is not recommended
for general use.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lasso_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">acceleration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lock_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn_unconverged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_bound_theta_ge_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infer'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.fit" title="Link to this definition">¶</a></dt>
<dd><p>Use jaxopt.ProximalGradiant to optimize the model’s free parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lasso_shift</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – L1 penalty on the shift parameters. Defaults to 1e-5.</p></li>
<li><p><strong>tol</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – Tolerance for the optimization. Defaults to 1e-6.</p></li>
<li><p><strong>maxiter</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Maximum number of iterations for the optimization. Defaults to 1000.</p></li>
<li><p><strong>acceleration</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, use FISTA acceleration. Defaults to True.</p></li>
<li><p><strong>lock_params</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Dictionary of parameters, and desired value to constrain
them at during optimization. By default, none of the parameters
besides the reference shift, and reference latent offset are locked.</p></li>
<li><p><strong>warn_unconverged</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – If True, raise a warning if the optimization does not converge.
convergence is defined by whether the model tolerance (‘’tol’’) threshold
was passed during the optimization process.
Defaults to True.</p></li>
<li><p><strong>upper_bound_theta_ge_scale</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a><em>, </em><em>None</em><em>, or </em><em>'infer'</em>) – The positive upper bound of the theta scale parameter -
negative values are not allowed.
Passing <code class="docutils literal notranslate"><span class="pre">None</span></code> allows the scale of the sigmoid to be unconstrained.
Passing the string literal ‘infer’ results in the
scale being set to double the range of the training data.
Defaults to ‘infer’.</p></li>
<li><p><strong>**kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Additional keyword arguments passed to the objective function.
These include hyperparameters like a ridge penalty on beta, shift, and gamma
as well as huber loss scaling.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_pred_accuracy">
<span class="sig-name descname"><span class="pre">plot_pred_accuracy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_corr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_pred_accuracy" title="Link to this definition">¶</a></dt>
<dd><p>Create a figure which visualizes the correlation
between model predicted functional score of all
variants in the training with ground truth measurements.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_epistasis">
<span class="sig-name descname"><span class="pre">plot_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_epistasis" title="Link to this definition">¶</a></dt>
<dd><p>Plot latent predictions against
gamma corrected ground truth measurements
of all samples in the training set.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_param_hist">
<span class="sig-name descname"><span class="pre">plot_param_hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_param_hist" title="Link to this definition">¶</a></dt>
<dd><p>Plot the histogram of a parameter.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_param_heatmap">
<span class="sig-name descname"><span class="pre">plot_param_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_param_heatmap" title="Link to this definition">¶</a></dt>
<dd><p>Plot the heatmap of a parameters
associated with specific sites and substitutions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_shifts_by_site">
<span class="sig-name descname"><span class="pre">plot_shifts_by_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold=3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_func=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_shifts_by_site" title="Link to this definition">¶</a></dt>
<dd><p>Summarize shift parameter values by associated sites and conditions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.plot_fit_param_comp_scatter">
<span class="sig-name descname"><span class="pre">plot_fit_param_comp_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'beta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'beta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">4]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_agg_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.plot_fit_param_comp_scatter" title="Link to this definition">¶</a></dt>
<dd><p>Plot a scatter plot of the parameter values of two models</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.Model.mut_param_heatmap">
<span class="sig-name descname"><span class="pre">mut_param_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mut_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_as_effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">line_and_heat_kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.Model.mut_param_heatmap" title="Link to this definition">¶</a></dt>
<dd><p>Wrapper method for visualizing the shift plot.
see <cite>multidms.plot.mut_shift_plot()</cite> for more</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="multidms.ModelCollection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">multidms.</span></span><span class="sig-name descname"><span class="pre">ModelCollection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fit_models</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A class for the comparison and visualization of multiple
<cite>multidms.Model</cite> fits. The respective collection of
training datasets for each fit must
share the same reference sequence and conditions. Additionally,
the inferred site maps must agree upon condition wildtypes
for all shared sites.</p>
<p>The utility function <cite>multidms.model_collection.fit_models</cite> is used to fit
the collection of models, and the resulting dataframe is passed to the
constructor of this class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>fit_models</strong> (<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a>) – A dataframe containing the fit attributes and pickled model objects
as returned by <cite>multidms.model_collection.fit_models</cite>.</p>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="multidms.ModelCollection.site_map_union">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">site_map_union</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><span class="pre">DataFrame</span></a></em><a class="headerlink" href="#multidms.ModelCollection.site_map_union" title="Link to this definition">¶</a></dt>
<dd><p>The union of all site maps of all datasets used for fitting.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.ModelCollection.conditions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">conditions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.12)"><span class="pre">list</span></a></em><a class="headerlink" href="#multidms.ModelCollection.conditions" title="Link to this definition">¶</a></dt>
<dd><p>The conditions (shared by each fitting dataset) used for fitting.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.ModelCollection.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#multidms.ModelCollection.reference" title="Link to this definition">¶</a></dt>
<dd><p>The reference conditions (shared by each fitting dataset) used for fitting.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.ModelCollection.shared_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shared_mutations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><span class="pre">tuple</span></a></em><a class="headerlink" href="#multidms.ModelCollection.shared_mutations" title="Link to this definition">¶</a></dt>
<dd><p>The mutations shared by each fitting dataset.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="multidms.ModelCollection.all_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">all_mutations</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><span class="pre">tuple</span></a></em><a class="headerlink" href="#multidms.ModelCollection.all_mutations" title="Link to this definition">¶</a></dt>
<dd><p>The mutations shared by each fitting dataset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.split_apply_combine_muts">
<span class="sig-name descname"><span class="pre">split_apply_combine_muts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('dataset_name',</span> <span class="pre">'scale_coeff_lasso_shift')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggregate_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inner_merge_dataset_muts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.split_apply_combine_muts" title="Link to this definition">¶</a></dt>
<dd><p>wrapper to split-apply-combine the set of mutational dataframes
harbored by each of the fits in the collection.</p>
<p>Here, we group the collection of fits using attributes
(columns in <code class="xref py py-attr docutils literal notranslate"><span class="pre">ModelCollection.fit_models</span></code>) specified using the
<code class="docutils literal notranslate"><span class="pre">groupby</span></code> parameter.
Each of the individual fits within a groups may then be filtered
via <code class="docutils literal notranslate"><span class="pre">**kwargs</span></code>, and aggregated via <code class="docutils literal notranslate"><span class="pre">aggregate_func</span></code>, before
the function stacks all the groups back together in a
tall style dataframe. The resulting dataframe will have a multiindex
with the mutation and the groupby attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>groupby</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)"><em>tuple</em></a><em> of </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>None</em><em>, </em><em>optional</em>) – The attributes to group the fits by. If None, then group by all
attributes except for the model, data, and step_loss attributes.
The default is (“dataset_name”, “scale_coeff_lasso_shift”).</p></li>
<li><p><strong>aggregate_func</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>callable</em><em>, </em><em>optional</em>) – The function to aggregate the mutational dataframes within each group.
The default is “mean”.</p></li>
<li><p><strong>inner_merge_dataset_muts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to toss mutations which are _not_ shared across all datasets
before aggregation of group mutation parameter values.
The default is True.</p></li>
<li><p><strong>query</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – The pandas query to apply to the <cite>ModelCollection.fit_models</cite>
dataframe before splitting. The default is None.</p></li>
<li><p><strong>**kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Keyword arguments to pass to the <a class="reference internal" href="#multidms.Model.get_mutations_df" title="multidms.Model.get_mutations_df"><code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.Model.get_mutations_df()</span></code></a>
method (“phenotype_as_effect”, and “times_seen_threshold”) see the
method docstring for details.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dataframe containing the aggregated mutational parameter values</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.2.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.add_validation_loss">
<span class="sig-name descname"><span class="pre">add_validation_loss</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.add_validation_loss" title="Link to this definition">¶</a></dt>
<dd><p>Add validation loss to the fit collection dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<em>pd.DataFrame</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>pd.DataFrame</em><em>)</em>) – The testing dataframe to compute validation loss with respect to,
must have columns “aa_substitutitions”, “condition”, and “func_score”.
If a dictionary is passed, there should be a key for
each unique dataset_name factor in the self.fit_models dataframe
- with the value being the respective testing dataframe.</p></li>
<li><p><strong>overwrite</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to overwrite the validation_loss column if it already exists.
The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The self.fit_models dataframe with the validation loss added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.get_conditional_loss_df">
<span class="sig-name descname"><span class="pre">get_conditional_loss_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.get_conditional_loss_df" title="Link to this definition">¶</a></dt>
<dd><p>return a long form dataframe with columns
“dataset_name”, “scale_coeff_lasso_shift”,
“split” (“training” or “validation”),
“loss” (actual value), and “condition”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>query</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – The query to apply to the fit_models dataframe
before formatting the loss dataframe. The default is None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.mut_param_heatmap">
<span class="sig-name descname"><span class="pre">mut_param_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aggregate_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inner_merge_dataset_muts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_as_effect</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.mut_param_heatmap" title="Link to this definition">¶</a></dt>
<dd><p>Create lineplot and heatmap altair chart
across replicate datasets.
This function optionally applies a given <cite>pandas.query</cite>
on the fit_models dataframe that should result in a subset of
fit’s which make sense to aggregate mutational data across, e.g.
replicate datasets.
It then computes the mean or median mutational parameter value
(“beta”, “shift”, or “predicted_func_score”)
between the remaining fits. and creates an interactive altair chart.</p>
<p>Note that this will throw an error if the queried fits have more
than one unique hyper-parameter besides “dataset_name”.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – The query to apply to the fit_models dataframe. This should be
used to subset the fits to only those which make sense to aggregate
mutational data across, e.g. replicate datasets.
For example, if you have a collection of
fits with different epistatic models, you may want to query
for only those fits with the same epistatic model. e.g.
<cite>query=”epistatic_model == ‘Sigmoid’”</cite>. For more on the query
syntax, see the
<a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html">pandas.query</a>
documentation.</p></li>
<li><p><strong>mut_param</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – The mutational parameter to plot. The default is “shift”.
Must be one of “shift”, “predicted_func_score”, or “beta”.</p></li>
<li><p><strong>aggregate_func</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – The function to aggregate the mutational parameter values
between dataset fits. The default is “mean”.</p></li>
<li><p><strong>inner_merge_dataset_muts</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to toss mutations which are _not_ shared across all datasets
before aggregation of group mutation parameter values.
The default is True.</p></li>
<li><p><strong>times_seen_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The minimum number of times a mutation must be seen across
all conditions within a single fit to be included in the
aggregation. The default is 0.</p></li>
<li><p><strong>phenotype_as_effect</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – Passed to <cite>Model.get_mutations_df()</cite>,
Only applies if <cite>mut_param=”predicted_func_score”</cite>.</p></li>
<li><p><strong>**kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Keyword arguments to pass to
<code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.plot._lineplot_and_heatmap()</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A chart object which can be displayed in a jupyter notebook
or saved to a file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>altair.Chart</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.mut_param_traceplot">
<span class="sig-name descname"><span class="pre">mut_param_traceplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mut_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'scale_coeff_lasso_shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width_scalar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height_scalar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.mut_param_traceplot" title="Link to this definition">¶</a></dt>
<dd><p>visualize mutation parameter values across the lasso penalty weights
(by default) of a given subset of the mutations in the form of an
<cite>altair.FacetChart</cite>. This is useful when you would like to confirm
that a reported mutational parameter value carries through across the
individual fits.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A chart object which can be displayed in a jupyter notebook
or saved to a file.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>altair.Chart</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.shift_sparsity">
<span class="sig-name descname"><span class="pre">shift_sparsity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'scale_coeff_lasso_shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width_scalar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height_scalar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.shift_sparsity" title="Link to this definition">¶</a></dt>
<dd><p>Visualize shift parameter set sparsity across the lasso penalty weights
(by default) in the form of an <cite>altair.FacetChart</cite>.
We will group the mutations according to their status as either a
a “stop” (e.g. A15*), or “nonsynonymous” (e.g. A15G) mutation before calculating
the sparsity. This is because in a way, mutations to stop codons act as a
False positive rate, as we expect their mutational effect to be equally
deleterious in all experiments, and thus have a shift parameter value of zero.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A chart object which can be displayed in a jupyter notebook
or saved to a file. If <cite>return_data=True</cite>, then a tuple
containing the chart and the underlying data will be returned.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>altair.Chart or Tuple(pd.DataFrame, altair.Chart)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="multidms.ModelCollection.mut_param_dataset_correlation">
<span class="sig-name descname"><span class="pre">mut_param_dataset_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'scale_coeff_lasso_shift'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width_scalar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.ModelCollection.mut_param_dataset_correlation" title="Link to this definition">¶</a></dt>
<dd><p>Visualize the correlation between replicate datasets across the lasso penalty
weights (by default) in the form of an <cite>altair.FacetChart</cite>.
We compute correlation of mutation parameters accross each pair of datasets
in the collection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – The parameter to plot on the x-axis.
The default is “scale_coeff_lasso_shift”.</p></li>
<li><p><strong>width_scalar</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The width of the chart. The default is 150.</p></li>
<li><p><strong>height</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The height of the chart. The default is 200.</p></li>
<li><p><strong>return_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to return the underlying data. The default is False.</p></li>
<li><p><strong>r</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The exponential of the correlation coefficient reported.
May be either 1 for pearson,
2 for coefficient of determination (r-squared),
The default is 2.</p></li>
<li><p><strong>**kwargs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – The keyword arguments to pass to the
<a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.split_apply_combine_muts" title="multidms.model_collection.ModelCollection.split_apply_combine_muts"><code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.model_collection.ModelCollection.split_apply_combine_muts()</span></code></a>
method. See the method docstring for details.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A chart object which can be displayed in a jupyter notebook
or saved to a file. If <cite>return_data=True</cite>, then a tuple
containing the chart and the underlying data will be returned.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>altair.Chart or Tuple(altair.Chart, pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="multidms.fit_models">
<span class="sig-prename descclassname"><span class="pre">multidms.</span></span><span class="sig-name descname"><span class="pre">fit_models</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">failures</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'error'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#multidms.fit_models" title="Link to this definition">¶</a></dt>
<dd><p>Fit collection of <a class="reference internal" href="multidms.model.html#multidms.model.Model" title="multidms.model.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">multidms.model.Model</span></code></a> models.</p>
<p>Enables fitting of multiple models simultaneously using multiple threads.
Most commonly, this function is used to fit a set of models across combinations
of replicate training datasets, and lasso coefficients for model selection and
evaluation. The returned dataframe is meant to be passed into the
<a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection" title="multidms.model_collection.ModelCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">multidms.model_collection.ModelCollection</span></code></a> class for comparison
and visualization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>params</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a>) – Dictionary which defines the parameter space of all models you
wish to run. Each value in the dictionary must be a list of
values, even in the case of singletons.
This function will compute all combinations of the parameter
space and pass each combination to <code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.utils.fit_one_model()</span></code>
to be run in parallel, thus only key-value pairs which
match the kwargs are allowed.
See the docstring of <a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.fit_one_model" title="multidms.model_collection.fit_one_model"><code class="xref py py-func docutils literal notranslate"><span class="pre">multidms.model_collection.fit_one_model()</span></code></a> for
a description of the allowed parameters.</p></li>
<li><p><strong>n_threads</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Number of threads (CPUs, cores) to use for fitting. Set to -1 to use
all CPUs available.</p></li>
<li><p><strong>failures</strong> (<em>{&quot;error&quot;</em><em>, </em><em>&quot;tolerate&quot;}</em>) – What if fitting fails for a model? If “error” then raise an error,
if “ignore” then just return <cite>None</cite> for models that failed optimization.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of models that fit successfully, number of models that failed,
and a dataframe which contains a row for each of the <cite>multidms.Model</cite>
object references along with the parameters each was fit with for convenience.
The dataframe is ultimately meant to be passed into the ModelCollection class.
for comparison and visualization.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>(n_fit, n_failed, fit_models)</p>
</dd>
</dl>
</dd></dl>

<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html">biophysical</a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.additive_model"><code class="docutils literal notranslate"><span class="pre">additive_model()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.sigmoidal_global_epistasis"><code class="docutils literal notranslate"><span class="pre">sigmoidal_global_epistasis()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.softplus_global_epistasis"><code class="docutils literal notranslate"><span class="pre">softplus_global_epistasis()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.nn_global_epistasis"><code class="docutils literal notranslate"><span class="pre">nn_global_epistasis()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.identity_activation"><code class="docutils literal notranslate"><span class="pre">identity_activation()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.biophysical.html#multidms.biophysical.softplus_activation"><code class="docutils literal notranslate"><span class="pre">softplus_activation()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.data.html">data</a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.data.html#multidms.data.split_sub"><code class="docutils literal notranslate"><span class="pre">split_sub()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.data.html#multidms.data.split_subs"><code class="docutils literal notranslate"><span class="pre">split_subs()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.data.html#multidms.data.Data"><code class="docutils literal notranslate"><span class="pre">Data</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.counter"><code class="docutils literal notranslate"><span class="pre">Data.counter</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.name"><code class="docutils literal notranslate"><span class="pre">Data.name</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.conditions"><code class="docutils literal notranslate"><span class="pre">Data.conditions</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.reference"><code class="docutils literal notranslate"><span class="pre">Data.reference</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.mutations"><code class="docutils literal notranslate"><span class="pre">Data.mutations</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.mutations_df"><code class="docutils literal notranslate"><span class="pre">Data.mutations_df</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.variants_df"><code class="docutils literal notranslate"><span class="pre">Data.variants_df</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.site_map"><code class="docutils literal notranslate"><span class="pre">Data.site_map</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.non_identical_mutations"><code class="docutils literal notranslate"><span class="pre">Data.non_identical_mutations</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.non_identical_sites"><code class="docutils literal notranslate"><span class="pre">Data.non_identical_sites</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.reference_sequence_conditions"><code class="docutils literal notranslate"><span class="pre">Data.reference_sequence_conditions</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.training_data"><code class="docutils literal notranslate"><span class="pre">Data.training_data</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.binarymaps"><code class="docutils literal notranslate"><span class="pre">Data.binarymaps</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.targets"><code class="docutils literal notranslate"><span class="pre">Data.targets</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.mutparser"><code class="docutils literal notranslate"><span class="pre">Data.mutparser</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.parse_mut"><code class="docutils literal notranslate"><span class="pre">Data.parse_mut</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.parse_muts"><code class="docutils literal notranslate"><span class="pre">Data.parse_muts</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.convert_subs_wrt_ref_seq"><code class="docutils literal notranslate"><span class="pre">Data.convert_subs_wrt_ref_seq()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.plot_times_seen_hist"><code class="docutils literal notranslate"><span class="pre">Data.plot_times_seen_hist()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.data.html#multidms.data.Data.plot_func_score_boxplot"><code class="docutils literal notranslate"><span class="pre">Data.plot_func_score_boxplot()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model.html">model</a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model.html#multidms.model.Model"><code class="docutils literal notranslate"><span class="pre">Model</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.counter"><code class="docutils literal notranslate"><span class="pre">Model.counter</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.name"><code class="docutils literal notranslate"><span class="pre">Model.name</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.params"><code class="docutils literal notranslate"><span class="pre">Model.params</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.state"><code class="docutils literal notranslate"><span class="pre">Model.state</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.converged"><code class="docutils literal notranslate"><span class="pre">Model.converged</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.data"><code class="docutils literal notranslate"><span class="pre">Model.data</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.model_components"><code class="docutils literal notranslate"><span class="pre">Model.model_components</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.loss"><code class="docutils literal notranslate"><span class="pre">Model.loss</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.conditional_loss"><code class="docutils literal notranslate"><span class="pre">Model.conditional_loss</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.variants_df"><code class="docutils literal notranslate"><span class="pre">Model.variants_df</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.get_variants_df"><code class="docutils literal notranslate"><span class="pre">Model.get_variants_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.mutations_df"><code class="docutils literal notranslate"><span class="pre">Model.mutations_df</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.get_mutations_df"><code class="docutils literal notranslate"><span class="pre">Model.get_mutations_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.get_df_loss"><code class="docutils literal notranslate"><span class="pre">Model.get_df_loss()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.add_phenotypes_to_df"><code class="docutils literal notranslate"><span class="pre">Model.add_phenotypes_to_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.wildtype_df"><code class="docutils literal notranslate"><span class="pre">Model.wildtype_df</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.mutation_site_summary_df"><code class="docutils literal notranslate"><span class="pre">Model.mutation_site_summary_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.get_condition_params"><code class="docutils literal notranslate"><span class="pre">Model.get_condition_params()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.phenotype_fromsubs"><code class="docutils literal notranslate"><span class="pre">Model.phenotype_fromsubs()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.latent_fromsubs"><code class="docutils literal notranslate"><span class="pre">Model.latent_fromsubs()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.phenotype_frombinary"><code class="docutils literal notranslate"><span class="pre">Model.phenotype_frombinary()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.latent_frombinary"><code class="docutils literal notranslate"><span class="pre">Model.latent_frombinary()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.fit_reference_beta"><code class="docutils literal notranslate"><span class="pre">Model.fit_reference_beta()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.fit"><code class="docutils literal notranslate"><span class="pre">Model.fit()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_pred_accuracy"><code class="docutils literal notranslate"><span class="pre">Model.plot_pred_accuracy()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_epistasis"><code class="docutils literal notranslate"><span class="pre">Model.plot_epistasis()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_param_hist"><code class="docutils literal notranslate"><span class="pre">Model.plot_param_hist()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_param_heatmap"><code class="docutils literal notranslate"><span class="pre">Model.plot_param_heatmap()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_shifts_by_site"><code class="docutils literal notranslate"><span class="pre">Model.plot_shifts_by_site()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.plot_fit_param_comp_scatter"><code class="docutils literal notranslate"><span class="pre">Model.plot_fit_param_comp_scatter()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model.html#multidms.model.Model.mut_param_heatmap"><code class="docutils literal notranslate"><span class="pre">Model.mut_param_heatmap()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model_collection.html"><code class="docutils literal notranslate"><span class="pre">ModelCollectionFitError</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.fit_one_model"><code class="docutils literal notranslate"><span class="pre">fit_one_model()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.stack_fit_models"><code class="docutils literal notranslate"><span class="pre">stack_fit_models()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.fit_models"><code class="docutils literal notranslate"><span class="pre">fit_models()</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection"><code class="docutils literal notranslate"><span class="pre">ModelCollection</span></code></a><ul>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.site_map_union"><code class="docutils literal notranslate"><span class="pre">ModelCollection.site_map_union</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.conditions"><code class="docutils literal notranslate"><span class="pre">ModelCollection.conditions</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.reference"><code class="docutils literal notranslate"><span class="pre">ModelCollection.reference</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.shared_mutations"><code class="docutils literal notranslate"><span class="pre">ModelCollection.shared_mutations</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.all_mutations"><code class="docutils literal notranslate"><span class="pre">ModelCollection.all_mutations</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.split_apply_combine_muts"><code class="docutils literal notranslate"><span class="pre">ModelCollection.split_apply_combine_muts()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.add_validation_loss"><code class="docutils literal notranslate"><span class="pre">ModelCollection.add_validation_loss()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.get_conditional_loss_df"><code class="docutils literal notranslate"><span class="pre">ModelCollection.get_conditional_loss_df()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.mut_param_heatmap"><code class="docutils literal notranslate"><span class="pre">ModelCollection.mut_param_heatmap()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.mut_param_traceplot"><code class="docutils literal notranslate"><span class="pre">ModelCollection.mut_param_traceplot()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.shift_sparsity"><code class="docutils literal notranslate"><span class="pre">ModelCollection.shift_sparsity()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="multidms.model_collection.html#multidms.model_collection.ModelCollection.mut_param_dataset_correlation"><code class="docutils literal notranslate"><span class="pre">ModelCollection.mut_param_dataset_correlation()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="multidms.plot.html">plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="multidms.plot.html#multidms.plot.color_gradient_hex"><code class="docutils literal notranslate"><span class="pre">color_gradient_hex()</span></code></a></li>
</ul>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/2020-logo-1000px-transparent.png" alt="Logo" />
    
    <h1 class="logo logo-name">multidms</h1>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=matsengrp&repo=multidms&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">multidms documentation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulation_validation.html">Simulation validation</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_delta_BA1_example.html"><code class="docutils literal notranslate"><span class="pre">multidms</span></code> usage examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">multidms package</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgments</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="fit_delta_BA1_example.html" title="previous chapter"><code class="docutils literal notranslate"><span class="pre">multidms</span></code> usage examples</a></li>
      <li>Next: <a href="multidms.biophysical.html" title="next chapter">biophysical</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2023, Jared Galloway, Hugh Haddox.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/multidms.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>