{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd18d90e",
   "metadata": {},
   "source": [
    "# `multidms` fitting pipeline\n",
    "\n",
    "Here, we demonstrate the pipeline for fitting a `multidms` model on some data using the tools available in the package. Currently, we show how to pre-process data from various dms experiments targeting Delta and Omicron spike protein homologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "53fa951c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# built-in libraries\n",
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "from itertools import combinations\n",
    "import importlib\n",
    "import math\n",
    "import re\n",
    "from timeit import default_timer as timer\n",
    "import json\n",
    "\n",
    "# external dependencies\n",
    "import pandas as pd\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "from jax.experimental import sparse\n",
    "from jaxopt import ProximalGradient\n",
    "import jaxopt\n",
    "import numpy as onp\n",
    "from scipy.stats import pearsonr\n",
    "import binarymap as bmap\n",
    "from tqdm.notebook import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotnine\n",
    "\n",
    "# local dependencies\n",
    "sys.path.append(\"..\")\n",
    "import multidms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3484460c",
   "metadata": {},
   "source": [
    "**First, we'll define a few global variables.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b91f7ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The substitution column to use from the functional score dataframes\n",
    "substitution_column = 'aa_substitutions_reference'\n",
    "\n",
    "# The column providing which individual conditions each variant observation was derived from.\n",
    "experiment_column = 'homolog_exp'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1216729a",
   "metadata": {},
   "source": [
    "**Define all the knobs which define the pipeline's behavior**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8182d3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit_params = {\n",
    "    # pre-processing params\n",
    "    \"scale_counts\" : False,\n",
    "    \"fs_scaling_group_column\" : \"homolog_exp\",\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"pseudocount\" : 0.1,\n",
    "    \"agg_variants\" : True,\n",
    "    \"sample\" : False, # 10000,\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"clip_target\" : (-3.5, 2.5),\n",
    "    \"include_stop_variants\" : True,\n",
    "    \"shift_func_score_target_nonref\" : False,\n",
    "    # conditions to include in fit\n",
    "    \"reference_condition\" : \"Delta-2-1\",\n",
    "    \"conditions\" : [\"Delta-2-1\", \"Omicron_BA.1-2-1\"],\n",
    "    # fitting hyper-parameters\n",
    "    \"maxiter\" : 10000,\n",
    "    \"λ_lasso\" : 1e-5,\n",
    "    \"λ_ridge\" : 0,\n",
    "    \"model\" : \"non-linear\",\n",
    "    \"lower_bound\" : -3.5,\n",
    "    \"hinge_scale\" : 2.0\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "846274a7",
   "metadata": {},
   "source": [
    "**Define the attributes you'll have access to after each fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb0739c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = list(fit_params.keys()) + [\n",
    "    \"tuned_model_params\", \"mutation_effects_df\", \"variant_prediction_df\", \"site_map\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42708714",
   "metadata": {},
   "source": [
    "**uncomment if you want to reset all results, or if this is the first run**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3023164a",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.DataFrame(columns = cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89330f69",
   "metadata": {},
   "source": [
    "## Pre-Proccessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f8e5e1",
   "metadata": {},
   "source": [
    "**Read in the dms data and list all available experimental conditions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee7c2d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>125127</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "0   2021-10-28_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "1   2021-10-28_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "2   2021-10-28_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "3   2021-10-28_thaw-1_VSVG_control_2   Lib-3      thaw-1          2   \n",
       "4   2021-10-28_thaw-1_VSVG_control_1   Lib-4      thaw-1          1   \n",
       "5   2021-10-28_thaw-1_VSVG_control_2   Lib-4      thaw-1          2   \n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "7   2021-10-28_thaw-1_VSVG_control_2   Lib-2      thaw-1          2   \n",
       "8   2022-03-25_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "9   2022-03-25_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "11  2022-06-22_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "12  2022-10-22_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "13  2022-10-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "14  2022-10-22_thaw-2_VSVG_control_1   Lib-1      thaw-2          2   \n",
       "15  2022-10-22_thaw-2_VSVG_control_1   Lib-2      thaw-2          2   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "0   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "0   Lib-1_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   Lib-1_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   Lib-3_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   Lib-3_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   Lib-4_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   Lib-4_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   Lib-2_2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   Lib-1_2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   Lib-1_2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "0   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "1   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "2   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "3   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "4   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "5   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "7   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "8   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "9   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "11  ../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...   \n",
       "12  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "13  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "14  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "15  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "0         library                        pre_sampl...                   45131   \n",
       "1         library                        pre_sampl...                   45131   \n",
       "2         library                        pre_sampl...                   37703   \n",
       "3         library                        pre_sampl...                   37703   \n",
       "4         library                        pre_sampl...                   39694   \n",
       "5         library                        pre_sampl...                   39694   \n",
       "6         library                        pre_sampl...                   41586   \n",
       "7         library                        pre_sampl...                   41586   \n",
       "8         library                        pre_sampl...                   94347   \n",
       "9         library                        pre_sampl...                   94347   \n",
       "10         library                        pre_samp...                  140643   \n",
       "11         library                        pre_samp...                  125127   \n",
       "12        library                        pre_sampl...                   90992   \n",
       "13        library                        pre_sampl...                   85210   \n",
       "14        library                        pre_sampl...                   90992   \n",
       "15        library                        pre_sampl...                   85210   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "0          Delta         Delta-1-1  \n",
       "1          Delta         Delta-1-2  \n",
       "2          Delta         Delta-3-1  \n",
       "3          Delta         Delta-3-2  \n",
       "4          Delta         Delta-4-1  \n",
       "5          Delta         Delta-4-2  \n",
       "6          Delta         Delta-2-1  \n",
       "7          Delta         Delta-2-2  \n",
       "8   Omicron_BA.1  Omicron_BA.1-1-1  \n",
       "9   Omicron_BA.1  Omicron_BA.1-1-2  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  \n",
       "11  Omicron_BA.1  Omicron_BA.1-3-1  \n",
       "12  Omicron_BA.2  Omicron_BA.2-1-1  \n",
       "13  Omicron_BA.2  Omicron_BA.2-2-1  \n",
       "14  Omicron_BA.2  Omicron_BA.2-1-2  \n",
       "15  Omicron_BA.2  Omicron_BA.2-2-2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = pd.DataFrame()\n",
    "sites = {}\n",
    "wt_seqs = {}\n",
    "\n",
    "for homolog in [\"Delta\", \"Omicron_BA.1\", \"Omicron_BA.2\"]:\n",
    "    \n",
    "    # functional scores\n",
    "    func_sel = pd.read_csv(f\"../results/{homolog}/functional_selections.csv\")\n",
    "    func_sel = func_sel.assign(\n",
    "        filename = f\"../results/{homolog}/\" + \n",
    "        func_sel.library + \"_\" + \n",
    "        func_sel.preselection_sample + \n",
    "        \"_vs_\" + func_sel.postselection_sample + \n",
    "        \"_func_scores.csv\"\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        func_sel_scores_df = func_sel.filename.apply(lambda f: pd.read_csv(f))\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        len_func_sel_scores_df = func_sel.func_sel_scores_df.apply(lambda x: len(x))\n",
    "    )\n",
    "    fun_sel = func_sel.assign(homolog = homolog)\n",
    "    func_score_data = pd.concat([func_score_data, fun_sel]).reset_index(drop=True)\n",
    "\n",
    "# Add a column that gives a unique ID to each homolog/DMS experiment\n",
    "func_score_data['homolog_exp'] = func_score_data.apply(\n",
    "    lambda row: f\"{row['homolog']}-{row['library']}-{row['replicate']}\".replace('-Lib',''),\n",
    "    axis=1\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79383a1d",
   "metadata": {},
   "source": [
    "**Query the conditions to be included in a fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6408eb32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "6         library                        pre_sampl...                   41586   \n",
       "10         library                        pre_samp...                  140643   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "6          Delta         Delta-2-1  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = func_score_data.query(\n",
    "    f\"{experiment_column}.isin({fit_params['conditions']})\"\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0ec32631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/json": {
       "ascii": false,
       "bar_format": null,
       "colour": null,
       "elapsed": 0.009624719619750977,
       "initial": 0,
       "n": 0,
       "ncols": null,
       "nrows": 46,
       "postfix": null,
       "prefix": "",
       "rate": null,
       "total": 2,
       "unit": "it",
       "unit_divisor": 1000,
       "unit_scale": false
      },
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d89fe86d9d0148d2bf729447798b79ef",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df = pd.DataFrame()\n",
    "for idx, row in tqdm(func_score_data.iterrows(), total=len(func_score_data)):\n",
    "    df = row.func_sel_scores_df.assign(homolog=row.homolog)\n",
    "    df = df.assign(library = row.library)\n",
    "    df = df.assign(replicate = row.replicate)\n",
    "    exp_func_score_df = df.assign(homolog_exp=row.homolog_exp)\n",
    "    func_score_df = pd.concat([func_score_df, exp_func_score_df])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed08ea8f",
   "metadata": {},
   "source": [
    "**Optionally subset the variants**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ecc3a404",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params[\"sample\"]:\n",
    "    func_score_df = func_score_df.sample(fit_params[\"sample\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82e1ad2",
   "metadata": {},
   "source": [
    "**remove all variants with string sites and stop codon wildtypes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "349dfa11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/json": {
       "ascii": false,
       "bar_format": null,
       "colour": null,
       "elapsed": 0.009277105331420898,
       "initial": 0,
       "n": 0,
       "ncols": null,
       "nrows": 46,
       "postfix": null,
       "prefix": "",
       "rate": null,
       "total": 182229,
       "unit": "it",
       "unit_divisor": 1000,
       "unit_scale": false
      },
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "df6af4b0283a4c378d243f90e7173a99",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/182229 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df.aa_substitutions_reference.fillna(\"\", inplace=True)\n",
    "gapped_sub_vars = []\n",
    "stop_wt_vars = []\n",
    "non_numeric_sites = []\n",
    "for idx, row in tqdm(func_score_df.iterrows(), total=len(func_score_df)):\n",
    "    if \"-\" in row[substitution_column]:\n",
    "        gapped_sub_vars.append(idx)\n",
    "    for sub in row[substitution_column].split():\n",
    "        if sub[0] == \"*\":\n",
    "            stop_wt_vars.append(idx)\n",
    "        if not sub[-2].isnumeric():\n",
    "            non_numeric_sites.append(idx)\n",
    "\n",
    "to_drop = set.union(set(gapped_sub_vars), set(stop_wt_vars), set(non_numeric_sites))\n",
    "func_score_df.drop(to_drop, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf8faff",
   "metadata": {},
   "source": [
    "**Optionally, scale the counts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "364e1f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO re-write and make function\n",
    "# def normalize_by_freq()?\n",
    "if fit_params['scale_counts']:\n",
    "    dfs = []\n",
    "    for (h, hdf) in func_score_df.groupby(fit_params[\"fs_scaling_group_column\"]):\n",
    "        n_post_counts = sum(hdf['post_count'])\n",
    "        if 'Delta' in h:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        else:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        hdf['orig_post_count'] = hdf['post_count']\n",
    "        hdf['post_count'] *= scaling_factor\n",
    "        hdf['post_count_wt'] *= scaling_factor\n",
    "        print(h, n_post_counts, round(scaling_factor, 2), round(sum(hdf['post_count']),2))\n",
    "\n",
    "        # Recompute enrichment ratios with new counts\n",
    "        hdf['pre_count_ps'] = hdf['pre_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_ps'] = hdf['post_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['pre_count_wt_ps'] = hdf['pre_count_wt'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_wt_ps'] = hdf['post_count_wt'] + fit_params[\"pseudocount\"]\n",
    "\n",
    "        total_pre_count = sum(hdf['pre_count_ps'])\n",
    "        total_post_count = sum(hdf['post_count_ps'])\n",
    "\n",
    "        hdf['pre_freq'] = hdf['pre_count_ps'] / total_pre_count\n",
    "        hdf['post_freq'] = hdf['post_count_ps'] / total_post_count\n",
    "        hdf['pre_freq_wt'] = hdf['pre_count_wt_ps'] / total_pre_count\n",
    "        hdf['post_freq_wt'] = hdf['post_count_wt_ps'] / total_post_count\n",
    "\n",
    "        hdf['wt_e'] = hdf['post_freq_wt'] / hdf['pre_freq_wt']\n",
    "        hdf['var_e'] = hdf['post_freq'] / hdf['pre_freq']\n",
    "        hdf['e'] = hdf['var_e'] / hdf['wt_e']\n",
    "        #hdf.dropna(subset=['e'], inplace=True)\n",
    "        hdf['func_score'] = hdf['e'].apply(lambda x: math.log(x, 2))\n",
    "        dfs.append(hdf)\n",
    "    func_score_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0310eb5",
   "metadata": {},
   "source": [
    "**Drop all variants with pre-counts below a threshold.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "678b9752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 169856 variants, 16326 had fewer than 100 counts before selection, and were filtered out\n"
     ]
    }
   ],
   "source": [
    "# Drop barcoded variants with pre-counts below a threshold\n",
    "n_pre_threshold = len(func_score_df)\n",
    "func_score_df = func_score_df[func_score_df['pre_count'] >= fit_params[\"min_pre_counts\"]]\n",
    "print(f\"Of {n_pre_threshold} variants, {n_pre_threshold - len(func_score_df)} had fewer than {fit_params['min_pre_counts']} counts before selection, and were filtered out\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39fd464c",
   "metadata": {},
   "source": [
    "**Optionally throw all variants with stop codons.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "aa118736",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results[\"include_stop_variants\"] = True\n",
    "if not fit_params[\"include_stop_variants\"]:\n",
    "    stop_variants = [idx for idx, row in func_score_df.iterrows() if \"*\" in row[substitution_column]]\n",
    "    func_score_df = func_score_df.drop(stop_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12087cb5",
   "metadata": {},
   "source": [
    "**Optionally, aggregate variant functional scores across barcode replicates**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bf2ed1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params[\"agg_variants\"]:\n",
    "    func_score_df = func_score_df.groupby([substitution_column, experiment_column]).mean().reset_index()\n",
    "func_score_df[\"pre_count\"] = func_score_df[\"pre_count\"].astype(int)\n",
    "func_score_df[\"post_count\"] = func_score_df[\"post_count\"].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12f74dad",
   "metadata": {},
   "source": [
    "**Optionally, clip the target functional scores**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "72e18352",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params['clip_target']:\n",
    "    func_score_df[\"func_score\"] = func_score_df[\"func_score\"].clip(*fit_params[\"clip_target\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31db55e5",
   "metadata": {},
   "source": [
    "## Prep the data for model fit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6dc7bc7",
   "metadata": {},
   "source": [
    "**From the functional score dataframe, we prep the modeling data**\n",
    "1. (X, y) - the sparse-binarymap and target jax ndarrays for training. Both X and y are dictionaries with entries for each of the conditions.\n",
    "2. df - the functional score dataframe which has been curated to include only shared sites between conditions and adds the variant substitutions string with respect to the reference condition.\n",
    "3. All the substitutions given in the order which parameters are assigned (i.e. their column index into X)\n",
    "4. The site map for all inferred wildtype aa at each site in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d081b1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 11 site(s) lacking data in at least one homolog.\n",
      "2766 of the 93822 variants were removed because they had mutations at the above sites, leaving 91056 variants.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 24%|███████████▊                                      | 15857/66785 [00:31<01:40, 504.60it/s]"
     ]
    }
   ],
   "source": [
    "importlib.reload(multidms.utils)\n",
    "\n",
    "# multidms = importlib.reload(multidms)\n",
    "(X, y), df, all_subs, site_map = multidms.utils.create_homolog_modeling_data(\n",
    "                            func_score_df, \n",
    "                            experiment_column,\n",
    "                            fit_params[\"reference_condition\"],\n",
    "                            substitution_column,\n",
    "                            \"func_score\"\n",
    "                        )\n",
    "\n",
    "sig_upper = df[\"func_score\"].quantile(0.95)\n",
    "sig_lower = df[\"func_score\"].quantile(0.05)\n",
    "sig_range = sig_upper - sig_lower"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfd2ec25",
   "metadata": {},
   "source": [
    "**Initialize parameters for the model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc306e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize all params\n",
    "params = multidms.utils.initialize_model_params(\n",
    "    func_score_df[experiment_column].unique(), \n",
    "    n_beta_shift_params=X[fit_params['reference_condition']].shape[1],\n",
    "    include_alpha=True if fit_params[\"model\"] == \"non-linear\" else False,\n",
    "    init_sig_range=sig_range,\n",
    "    init_sig_min=sig_lower\n",
    ")\n",
    "params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04321919",
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(multidms.model)\n",
    "cost_func = multidms.model.cost_smooth\n",
    "prox_func = multidms.model.prox\n",
    "pred_func = multidms.model.f\n",
    "if fit_params[\"model\"] == \"linear\":\n",
    "    cost_func = multidms.model.cost_smooth_linear\n",
    "    pred_func = multidms.model.f_linear"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cff04978",
   "metadata": {},
   "source": [
    "## Fit the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46d28cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"\\nPre-Optimization\")\n",
    "print(f\"----------------\")\n",
    "pred = cost_func(\n",
    "    params, \n",
    "    data=(X, y),\n",
    "    λ_ridge=fit_params['λ_ridge'],\n",
    "    lower_bound=fit_params['lower_bound'], \n",
    "    hinge_scale=fit_params['hinge_scale']\n",
    ")\n",
    "print(f\"cost = {pred:.2e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29844021",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference_condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202ecaa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tol = 1e-6\n",
    "start = timer()\n",
    "\n",
    "# initialize a solver\n",
    "solver = ProximalGradient(\n",
    "    cost_func,\n",
    "    prox_func, \n",
    "    tol=tol, \n",
    "    maxiter=fit_params['maxiter']\n",
    ")\n",
    "\n",
    "# the reference shift and gamma parameters forced to be zero\n",
    "lock_params = {\n",
    "    f\"S_{fit_params['reference_condition']}\" : jnp.zeros(len(params['β'])),\n",
    "    f\"γ_{fit_params['reference_condition']}\" : jnp.zeros(shape=(1,)),\n",
    "}\n",
    "\n",
    "# currently we lock C_h because of odd model behavior\n",
    "for condition in fit_params['conditions']:\n",
    "    lock_params[f\"C_{condition}\"] = jnp.zeros(shape=(1,))\n",
    "    lock_params[f\"γ_{condition}\"] = jnp.zeros(shape=(1,))\n",
    "    \n",
    "\n",
    "# lasso regularization on the Shift parameters\n",
    "lasso_params = {}\n",
    "for non_ref_condition in fit_params['conditions'][1:]:\n",
    "    lasso_params[f\"S_{non_ref_condition}\"] = fit_params['λ_lasso']\n",
    "\n",
    "# run the optimizer\n",
    "print('Fitting model')\n",
    "params, state = solver.run(\n",
    "    params,\n",
    "    hyperparams_prox = dict(\n",
    "        lasso_params = lasso_params,\n",
    "        lock_params = lock_params\n",
    "    ),\n",
    "    data=(X, y),\n",
    "    λ_ridge=fit_params['λ_ridge'],\n",
    "    lower_bound=fit_params['lower_bound'],\n",
    "    hinge_scale=fit_params['hinge_scale']\n",
    ")\n",
    "end = timer()\n",
    "\n",
    "# evaluate\n",
    "print(f\"\\nPost-Optimization\")\n",
    "print(f\"-----------------\")\n",
    "print(f\"Full model optimization: {state.iter_num} iterations\")\n",
    "print(f\"error = {state.error:.2e}\")\n",
    "pred = cost_func(\n",
    "    params, \n",
    "    (X, y), \n",
    "    λ_ridge=fit_params['λ_ridge'], \n",
    "    lower_bound=fit_params['lower_bound'], \n",
    "    hinge_scale=fit_params['hinge_scale']\n",
    ")\n",
    "print(f\"cost = {pred:.2e}\")\n",
    "print(f\"Wall time for fit: {end - start}\")\n",
    "print(f\"\\nDONE :)\")\n",
    "print(f\"-------------------\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebcfa6e7",
   "metadata": {},
   "source": [
    "**Create all single mutant effects dataframe**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1e5ac37",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df[\"predicted_latent_phenotype\"] = onp.nan\n",
    "df[f\"predicted_func_score\"] = onp.nan\n",
    "df[f\"corrected_func_score\"] = df[f\"func_score\"]\n",
    "\n",
    "print(f\"\\nRunning Predictions\")\n",
    "print(f\"-------------------\")\n",
    "\n",
    "# sub string, wt, site, mut, Beta, {S_c}, {F_c}\n",
    "# TODO number of times seen?\n",
    "\n",
    "mut_effect_df = pd.DataFrame({\n",
    "    \"substitution\" : all_subs,\n",
    "    \"β\" : params[\"β\"]\n",
    "})\n",
    "\n",
    "mut_effect_df[\"wt\"], mut_effect_df[\"site\"], mut_effect_df[\"mut\"] = zip(\n",
    "  *mut_effect_df[\"substitution\"].map(multidms.utils.split_sub)\n",
    ")\n",
    "\n",
    "binary_single_subs = sparse.BCOO.fromdense(onp.identity(len(all_subs)))\n",
    "\n",
    "for homolog, hdf in df.groupby(experiment_column):\n",
    "    \n",
    "    # collect relevant params\n",
    "    h_params = {\n",
    "        \"β\":params[\"β\"], \n",
    "        \"C_ref\":params[\"C_ref\"],\n",
    "        \"S\":params[f\"S_{homolog}\"], \n",
    "        \"C\":params[f\"C_{homolog}\"],\n",
    "        \"γ\":params[f\"γ_{homolog}\"]\n",
    "    }\n",
    "    \n",
    "    if fit_params[\"model\"] == \"non-linear\": h_params[\"α\"] = params[\"α\"]\n",
    "    \n",
    "    # latent predictions on all variants\n",
    "    z_h = multidms.model.ϕ(h_params, X[homolog])\n",
    "    df.loc[hdf.index, \"predicted_latent_phenotype\"] = z_h\n",
    "\n",
    "    # full model predictions\n",
    "    y_h_pred = pred_func(\n",
    "        h_params, \n",
    "        X[homolog], \n",
    "        lower_bound=fit_params['lower_bound'], \n",
    "        hinge_scale=fit_params['hinge_scale'],\n",
    "    )\n",
    "    df.loc[hdf.index, f\"predicted_func_score\"] = y_h_pred\n",
    "    \n",
    "    # add the corrected functional scores by tuned gamma\n",
    "    # df.loc[hdf.index, f\"corrected_func_score\"] -= h_params[f\"γ\"]\n",
    "    \n",
    "    # attach relevent params to mut effects df\n",
    "    mut_effect_df[f\"S_{homolog}\"] = params[f\"S_{homolog}\"]\n",
    "    \n",
    "    # predictions for all single subs\n",
    "    mut_effect_df[f\"F_{homolog}\"] = pred_func(\n",
    "        h_params, \n",
    "        binary_single_subs, \n",
    "        lower_bound=fit_params['lower_bound'], \n",
    "        hinge_scale=fit_params['hinge_scale'],\n",
    "    )\n",
    "\n",
    "row = fit_params.copy()\n",
    "\n",
    "row[\"variant_prediction_df\"] = df.drop(\"index\", axis=1)\n",
    "row[\"mutation_effects_df\"] = mut_effect_df\n",
    "row[\"site_map\"] = site_map.copy()\n",
    "row[\"tuned_model_params\"] = params.copy()\n",
    "\n",
    "# results = pd.Series(row)\n",
    "results.loc[len(results)] = pd.Series(row)\n",
    "\n",
    "print(f\"\\nDONE :)\")\n",
    "print(f\"-------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ade451df",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "results.loc[:, [\n",
    "    \"agg_variants\", \n",
    "    \"clip_target\",\n",
    "    \"hinge_scale\",\n",
    "    \"lower_bound\",\n",
    "    \"scale_counts\", \n",
    "    \"model\", \n",
    "    \"conditions\"\n",
    "    ]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cabca3ec",
   "metadata": {},
   "source": [
    "## Visualize results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f7280ba",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "importlib.reload(multidms.plot)\n",
    "exps = [0]\n",
    "res = results.loc[exps, :]\n",
    "# res = results.loc[len(results)-1:len(results), :]\n",
    "\n",
    "multidms.plot.plot_pred_scatter(\n",
    "    res, \n",
    "    #annotation_vars=list(fit_params.keys()), \n",
    "    hue=True, \n",
    "    printrow=False,\n",
    "    annotate_params=True,\n",
    ")\n",
    "multidms.plot.plot_param_hist(\n",
    "    res,\n",
    "    printrow=False\n",
    ")\n",
    "multidms.plot.plot_param_heatmap(\n",
    "    res, \n",
    "    printrow=False\n",
    ")\n",
    "multidms.plot.plot_param_by_site(\n",
    "    res, \n",
    "    printrow=False\n",
    ")\n",
    "# multidms.plot.plot_fit_param_comp_scatter(\n",
    "#     res, \n",
    "#     {0:\"β\", 1:\"β\"}\n",
    "# )\n",
    "# multidms.plot.plot_fit_param_comp_scatter(\n",
    "#     res, \n",
    "#     {2:\"S_Omicron_BA.1-2-1\", 3:\"S_Omicron_BA.1-3-1\"}\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39d24f2e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
