{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7f0224-78d0-4cfc-ac06-48c1d1b4dfe2",
   "metadata": {},
   "source": [
    "# Homolog BinaryMap Encoding"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809b2cd8",
   "metadata": {},
   "source": [
    "## Import `Python` modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e3816f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import json\n",
    "import binarymap as bmap\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import jax\n",
    "jax.config.update(\"jax_enable_x64\", True)\n",
    "import jax.numpy as jnp\n",
    "from jax.experimental import sparse\n",
    "from jaxopt import ProximalGradient\n",
    "import jaxopt\n",
    "import numpy as onp\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "from collections import defaultdict\n",
    "from timeit import default_timer as timer\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "tqdm.pandas()\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b14e961",
   "metadata": {},
   "source": [
    "## Strategy for converting mutations to be relative to the reference homolog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1dd9def7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference = \"MT\"\n",
    "homologs = {\n",
    "    \"1\" : \"M-G\",\n",
    "    \"2\" : \"MPP\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "380d116f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G3R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G3P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>M1W</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P3R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>P3G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3R</td>\n",
       "      <td>-2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>P2T</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  homolog  variant  log2E\n",
       "0       1      M1E    2.0\n",
       "1       1      G3R   -7.0\n",
       "2       1      G3P   -0.5\n",
       "3       1      M1W    2.3\n",
       "4       2      M1E    1.0\n",
       "5       2      P3R   -5.0\n",
       "6       2      P3G    0.4\n",
       "7       2  M1E P3G    2.7\n",
       "8       2  M1E P3R   -2.7\n",
       "9       2      P2T    0.3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [\"1\",\"1\",\"1\",\"1\", \"2\",\"2\",\"2\",\"2\",\"2\",\"2\"],\n",
    "    'variant' : ['M1E', 'G3R', 'G3P', 'M1W', 'M1E', 'P3R', 'P3G', 'M1E P3G', 'M1E P3R', 'P2T'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, 1, -5, 0.4, 2.7, -2.7, 0.3],\n",
    "}\n",
    "test_df = pd.DataFrame(test_dict)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "211ba89b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_homolog_modeling_data(\n",
    "    func_score_df:pd.DataFrame,\n",
    "    homolog_name_col: str,\n",
    "    reference_homolog: str,\n",
    "    substitution_col: str,\n",
    "    func_score_col: str\n",
    "):\n",
    "    \"\"\"\n",
    "    Takes a dataframe for making a `BinaryMap` object, and adds\n",
    "    a column where each entry is a list of mutations in a variant\n",
    "    relative to the amino-acid sequence of the reference homolog.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    func_score_df : pandas.DataFrame\n",
    "        This should be in the same format as described in BinaryMap.\n",
    "    \n",
    "    homolog_name_col : str\n",
    "        The name of the column in func_score_df that identifies the\n",
    "        homolog for a given variant. We require that the\n",
    "        reference homolog variants are labeled as 'reference'\n",
    "        in this column.\n",
    "        \n",
    "    reference_homolog : str\n",
    "        The name of the homolog existing in ``homolog_name_col`` for\n",
    "        which we should convert all substitution to be with respect to.\n",
    "    \n",
    "    substitution_col : str \n",
    "        The name of the column in func_score_df that\n",
    "        lists mutations in each variant relative to the homolog wildtype\n",
    "        amino-acid sequence where sites numbers must come from an alignment\n",
    "        to a reference sequence (which may or may not be the same as the\n",
    "        reference homolog).\n",
    "        \n",
    "    func_score_col : str\n",
    "        Column in func_scores_df giving functional score for each variant.\n",
    "        \n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "        \n",
    "    tuple : (dict[BinaryMap], dict[jnp.array]), pd.DataFrame, np.array, pd.DataFrame\n",
    "    \n",
    "        This function return a tuple which can be unpacked into the following:\n",
    "        \n",
    "        - (X, y) Where X and y are both dictionaries containing the prepped data\n",
    "            for training our JAX multidms model. The dictionary keys\n",
    "            stratify the datasets by homolog\n",
    "            \n",
    "        - A pandas dataframe which primary contains the information from\n",
    "            func_score_df, but has been curated to include only the variants\n",
    "            deemed appropriate for training, as well as the substitutions\n",
    "            converted to be wrt to the reference homolog.\n",
    "            \n",
    "        - A numpy array giving the substitutions (beta's) of the binary maps\n",
    "            in the order that is preserved to match the matrices in X.\n",
    "            \n",
    "        - A pandas dataframe providing the site map indexed by alignment site to\n",
    "            a column for each homolog wt amino acid. \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # TODO: strip gapped substitutions (insertions) variants\n",
    "    # from the func_score_df?\n",
    "    \n",
    "    def split_sub(sub_string):\n",
    "        \"\"\"String match the wt, site, and sub aa\n",
    "        in a given string denoting a single substitution\"\"\"\n",
    "        \n",
    "        pattern = r'(?P<aawt>\\w)(?P<site>[\\d\\w]+)(?P<aamut>[\\w\\*])'\n",
    "        match = re.search(pattern, sub_string)\n",
    "        assert match != None, sub_string\n",
    "        return match.group('aawt'), str(match.group('site')), match.group('aamut')\n",
    "    \n",
    "    def split_subs(subs_string):\n",
    "        \"\"\"wrap the split_sub func to work for a \n",
    "        string contining multiple substitutions\"\"\"\n",
    "        \n",
    "        wts, sites, muts = [], [], []\n",
    "        for sub in subs_string.split():\n",
    "            wt, site, mut = split_sub(sub)\n",
    "            wts.append(wt); sites.append(site); muts.append(mut)\n",
    "        return wts, sites, muts\n",
    "   \n",
    "    # Add columns that parse mutations into wt amino acid, site,\n",
    "    # and mutant amino acid\n",
    "    ret_fs_df = func_score_df.copy()\n",
    "    ret_fs_df[\"wts\"], ret_fs_df[\"sites\"], ret_fs_df[\"muts\"] = zip(\n",
    "        *ret_fs_df[substitution_col].map(split_subs)\n",
    "    )\n",
    "\n",
    "    # Use the substitution_col to infer the wildtype\n",
    "    # amino-acid sequence of each homolog, storing this\n",
    "    # information in a dataframe.\n",
    "    site_map = pd.DataFrame(dtype=\"string\")\n",
    "    for hom, hom_func_df in ret_fs_df.groupby(homolog_name_col):\n",
    "        for idx, row in hom_func_df.iterrows():\n",
    "            for wt, site  in zip(row.wts, row.sites):\n",
    "                site_map.loc[site, hom] = wt\n",
    "    \n",
    "    # Find all sites for which at least one homolog lacks data\n",
    "    # (this can happen if there is a gap in the alignment)\n",
    "    na_rows = site_map.isna().any(axis=1)\n",
    "    print(f\"Found {sum(na_rows)} site(s) lacking data in at least one homolog.\")\n",
    "    sites_to_throw = na_rows[na_rows].index\n",
    "    site_map.dropna(inplace=True)\n",
    "    \n",
    "    # Remove all variants with a mutation at one of the above\n",
    "    # \"disallowed\" sites lacking data\n",
    "    def flags_disallowed(disallowed_sites, sites_list):\n",
    "        \"\"\"Check to see if a sites list contains \n",
    "        any disallowed sites\"\"\"\n",
    "        for site in sites_list:\n",
    "            if site in disallowed_sites:\n",
    "                return False\n",
    "        return True\n",
    "    \n",
    "    ret_fs_df[\"allowed_variant\"] = ret_fs_df.sites.apply(\n",
    "        lambda sl: flags_disallowed(sites_to_throw,sl)\n",
    "    )\n",
    "    n_var_pre_filter = len(ret_fs_df)\n",
    "    ret_fs_df = ret_fs_df[ret_fs_df[\"allowed_variant\"]]\n",
    "    print(f\"{n_var_pre_filter-len(ret_fs_df)} of the {n_var_pre_filter} variants\"\n",
    "          f\" were removed because they had mutations at the above sites, leaving\"\n",
    "          f\" {len(ret_fs_df)} variants.\")\n",
    "\n",
    "    # Duplicate the substitutions_col, then convert the respective subs to be wrt ref\n",
    "    # using the function above\n",
    "    ret_fs_df = ret_fs_df.assign(var_wrt_ref = ret_fs_df[substitution_col])\n",
    "    for hom, hom_func_df in ret_fs_df.groupby(homolog_name_col):\n",
    "        if hom == reference_homolog: continue\n",
    "\n",
    "        # compute bundle muts for a specific site\n",
    "        for idx, row in tqdm(hom_func_df.iterrows(), total=len(hom_func_df)):\n",
    "            var_map = site_map[[reference_homolog, hom]].copy()\n",
    "            for wt, site, mut in zip(row.wts, row.sites, row.muts):\n",
    "                var_map.loc[site, hom] = mut\n",
    "            nis = var_map.where(\n",
    "                var_map[reference_homolog] != var_map[hom]\n",
    "            ).dropna()\n",
    "            muts = nis[reference_homolog] + nis.index + nis[hom]\n",
    "            ret_fs_df.loc[idx, \"var_wrt_ref\"] = \" \".join(muts.values)\n",
    "\n",
    "    # Get list of all allowed substitutions for which we will tune beta parameters\n",
    "    allowed_subs = {\n",
    "        s for subs in ret_fs_df.var_wrt_ref\n",
    "        for s in subs.split()\n",
    "    }\n",
    "    \n",
    "    # Make BinaryMap representations for each homolog\n",
    "    X, y = {}, {}\n",
    "    for homolog, homolog_func_score_df in ret_fs_df.groupby(\"homolog\"):\n",
    "        ref_bmap = bmap.BinaryMap(\n",
    "            homolog_func_score_df,\n",
    "            substitutions_col=\"var_wrt_ref\",\n",
    "            allowed_subs=allowed_subs\n",
    "        )\n",
    "        \n",
    "        # convert binarymaps into sparse arrays for model input\n",
    "        X[homolog] = sparse.BCOO.from_scipy_sparse(ref_bmap.binary_variants)\n",
    "        \n",
    "        # create jax array for functional score targets\n",
    "        y[homolog] = jnp.array(homolog_func_score_df[func_score_col].values)\n",
    "    \n",
    "    ret_fs_df.drop([\"wts\", \"sites\", \"muts\"], axis=1, inplace=True)\n",
    "\n",
    "    return (X, y), ret_fs_df, ref_bmap.all_subs, site_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "41a2c434",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 site(s) lacking data in at least one homolog.\n",
      "1 of the 10 variants were removed because they had mutations at the above sites, leaving 9 variants.\n"
     ]
    },
    {
     "data": {
      "application/json": {
       "ascii": false,
       "bar_format": null,
       "colour": null,
       "elapsed": 0.00935816764831543,
       "initial": 0,
       "n": 0,
       "ncols": null,
       "nrows": 7,
       "postfix": null,
       "prefix": "",
       "rate": null,
       "total": 5,
       "unit": "it",
       "unit_divisor": 1000,
       "unit_scale": false
      },
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ddb68babe9d54ff0873761713477551b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G3R</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>True</td>\n",
       "      <td>G3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G3P</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>G3P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>M1W</td>\n",
       "      <td>2.3</td>\n",
       "      <td>True</td>\n",
       "      <td>M1W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E G3P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P3R</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>G3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>P3G</td>\n",
       "      <td>0.4</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3G</td>\n",
       "      <td>2.7</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3R</td>\n",
       "      <td>-2.7</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E G3R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  homolog  variant  log2E  allowed_variant var_wrt_ref\n",
       "0       1      M1E    2.0             True         M1E\n",
       "1       1      G3R   -7.0             True         G3R\n",
       "2       1      G3P   -0.5             True         G3P\n",
       "3       1      M1W    2.3             True         M1W\n",
       "4       2      M1E    1.0             True     M1E G3P\n",
       "5       2      P3R   -5.0             True         G3R\n",
       "6       2      P3G    0.4             True            \n",
       "7       2  M1E P3G    2.7             True         M1E\n",
       "8       2  M1E P3R   -2.7             True     M1E G3R"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(X, y), updated_func_score_df, all_subs, site_map = create_homolog_modeling_data(\n",
    "    test_df, \n",
    "    \"homolog\", \n",
    "    \"1\", \n",
    "    \"variant\", \n",
    "    \"log2E\"\n",
    ")\n",
    "updated_func_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "052713ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   1  2\n",
       "1  M  M\n",
       "3  G  P"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "92279d63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M1E', 'M1W', 'G3P', 'G3R']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_subs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd266716",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
