{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7f0224-78d0-4cfc-ac06-48c1d1b4dfe2",
   "metadata": {},
   "source": [
    "# Homolog BinaryMap Encoding"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809b2cd8",
   "metadata": {},
   "source": [
    "## Import `Python` modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "e3816f0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import json\n",
    "import binarymap as bmap\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import jax\n",
    "jax.config.update(\"jax_enable_x64\", True)\n",
    "import jax.numpy as jnp\n",
    "from jax.experimental import sparse\n",
    "from jaxopt import ProximalGradient\n",
    "import jaxopt\n",
    "import numpy as onp\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "from collections import defaultdict\n",
    "from timeit import default_timer as timer\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "edfc0022",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import sys\n",
    "# sys.path.append(\"..\")\n",
    "# from multidms.utils import create_homolog_modeling_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b14e961",
   "metadata": {},
   "source": [
    "## Strategy for converting mutations to be relative to the reference homolog"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dd2195a",
   "metadata": {},
   "source": [
    "As described above, our strategy involves converting all mutations in all homologs to be relative to the amino-acid sequence of the reference homolog. The below function performs this step.\n",
    "\n",
    "Note 1: we will probably need some special-purpose code to handle gaps. This isn't done yet."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7222d340",
   "metadata": {},
   "source": [
    "Next, we will test the above function with a small test case. Below, we define variants from two imaginary homologs: \"reference\" and \"2\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "id": "1dd9def7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference = \"MT\"\n",
    "homologs = {\n",
    "    \"1\" : \"M-G\",\n",
    "    \"2\" : \"MPP\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "id": "380d116f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G3R</td>\n",
       "      <td>-7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G3P</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P3R</td>\n",
       "      <td>-5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P3G</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3G</td>\n",
       "      <td>2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3R</td>\n",
       "      <td>-2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>P2T</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  homolog  variant  log2E\n",
       "0       1      M1E    2.0\n",
       "1       1      G3R   -7.0\n",
       "2       1      G3P   -0.5\n",
       "3       2      M1E    2.3\n",
       "4       2      P3R   -5.0\n",
       "5       2      P3G    0.4\n",
       "6       2  M1E P3G    2.7\n",
       "7       2  M1E P3R   -2.7\n",
       "8       2      P2T    0.3"
      ]
     },
     "execution_count": 422,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dict = {\n",
    "    'homolog' : [\"1\",\"1\",\"1\",\"2\",\"2\",\"2\",\"2\",\"2\",\"2\"],\n",
    "    'variant' : ['M1E', 'G3R', 'G3P', 'M1E', 'P3R', 'P3G', 'M1E P3G', 'M1E P3R', 'P2T'],\n",
    "    'log2E' : [2, -7, -0.5, 2.3, -5, 0.4, 2.7, -2.7, 0.3],\n",
    "}\n",
    "test_df = pd.DataFrame(test_dict)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "id": "211ba89b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_homolog_modeling_data(\n",
    "    func_score_df:pd.DataFrame,\n",
    "    homolog_name_col: str,\n",
    "    substitution_col: str,\n",
    "    func_score_col: str\n",
    "):\n",
    "    \"\"\"\n",
    "    Takes a dataframe for making a `BinaryMap` object, and adds\n",
    "    a column where each entry is a list of mutations in a variant\n",
    "    relative to the amino-acid sequence of the reference homolog.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    func_score_df : pandas.DataFrame\n",
    "        This should be in the same format as described in BinaryMap.\n",
    "    \n",
    "    homolog_name_col : str\n",
    "        The name of the column in ret_fs_df that identifies the\n",
    "        homolog for a given variant. We require that the\n",
    "        reference homolog variants are labeled as 'reference'\n",
    "        in this column.\n",
    "    \n",
    "    substitution_col : str \n",
    "        The name of the column in ret_fs_df that\n",
    "        lists mutations in each variant relative to the homolog wildtype\n",
    "        amino-acid sequence where sites must come from a sequence alignment\n",
    "        to the reference.\n",
    "        \n",
    "    func_score_col : str\n",
    "        Column in func_scores_df giving functional score for each variant.\n",
    "        \n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "        \n",
    "    TODO\n",
    "    \n",
    "    \"\"\"    \n",
    "    \n",
    "    \n",
    "    def split_sub(sub_string):\n",
    "        \"\"\"String match the wt, site, and sub aa\n",
    "        in a given string denoting a single substitution\"\"\"\n",
    "        \n",
    "        pattern = r'(?P<aawt>\\w)(?P<site>[\\d\\w]+)(?P<aamut>[\\w\\*])'\n",
    "        match = re.search(pattern, sub_string)\n",
    "        assert match != None, sub_string\n",
    "        return match.group('aawt'), str(match.group('site')), match.group('aamut')\n",
    "    \n",
    "    def split_subs(subs_string):\n",
    "        \"\"\"wrap the split_sub func to work for a \n",
    "        string contining multiple substitutions\"\"\"\n",
    "        \n",
    "        wts, sites, muts = [], [], []\n",
    "        for sub in subs_string.split():\n",
    "            wt, site, mut = split_sub(sub)\n",
    "            wts.append(wt); sites.append(site); muts.append(mut)\n",
    "        return wts, sites, muts\n",
    "   \n",
    "    ret_fs_df = func_score_df.copy()\n",
    "    ret_fs_df[\"wts\"], ret_fs_df[\"sites\"], ret_fs_df[\"muts\"] = zip(\n",
    "        *ret_fs_df[substitution_col].map(split_subs)\n",
    "    )\n",
    "\n",
    "    # we want to build a site map top infer the wildtype\n",
    "    # this will have a dictionary for each homolog\n",
    "    # that dictionary will have keys for sites seen, \n",
    "    # and value as homolog wildtype\n",
    "    site_map = pd.DataFrame(dtype=\"string\")\n",
    "    # for each homolog, find available sites\n",
    "    for hom, hom_func_df in ret_fs_df.groupby(homolog_name_col):\n",
    "        for idx, row in hom_func_df.iterrows():\n",
    "            for wt, site  in zip(row.wts, row.sites):\n",
    "                # TODO do we continue if \"-\" in mut col? or in split_subs()?\n",
    "                site_map.loc[site, hom] = wt\n",
    "    \n",
    "    # Find all sites where any of the homologs don't have an entry\n",
    "    na_rows = site_map.isna().any(axis=1)\n",
    "    sites_to_throw = na_rows[na_rows].index\n",
    "    site_map.dropna(inplace=True)\n",
    "    \n",
    "    # remove all variant's that contain disallowed site\n",
    "    def flags_disallowed(sites, sites_list):\n",
    "        \"\"\"Check to see if a sites list contains \n",
    "        any disallowed sites\"\"\"\n",
    "        for site in sites:\n",
    "            if site in sites_list:\n",
    "                return False\n",
    "        return True\n",
    "    \n",
    "    ret_fs_df[\"allowed_variant\"] = ret_fs_df.sites.apply(\n",
    "        lambda sl: flags_disallowed(sites_to_throw,sl)\n",
    "    )\n",
    "    ret_fs_df = ret_fs_df[ret_fs_df[\"allowed_variant\"]]\n",
    "    \n",
    "    def subs_wrt_ref(subs_str, sites_map, ref, hom):\n",
    "        \"\"\"Takes in a string of substitutions wrt homolog seq,\n",
    "        a sites map containing seqs for both the homolog\n",
    "        and reference, as well as the column names for\n",
    "        both. It then copies and mutates the homolog\n",
    "        sequence before\n",
    "        \"\"\"\n",
    "        \n",
    "        var_map = sites_map.copy()\n",
    "        for sub in subs_str.split():\n",
    "            wt, site, mut = split_sub(sub)\n",
    "            var_map.loc[site, hom] = mut\n",
    "\n",
    "        ref_muts = [\n",
    "            f\"{row[ref]}{i}{row[hom]}\" \n",
    "            for i, row in var_map.iterrows()\n",
    "            if row[ref] != row[hom]\n",
    "        ]\n",
    "        \n",
    "        return \" \".join(ref_muts)\n",
    "\n",
    "    # Duplicate the substitutions_col, then convert the respective subs to be wrt ref\n",
    "    # using the function above\n",
    "    ret_fs_df = ret_fs_df.assign(\n",
    "            var_wrt_ref = ret_fs_df[substitution_col].values\n",
    "    )\n",
    "    for hom, hom_func_df in ret_fs_df.groupby(homolog_name_col):\n",
    "        if hom == \"1\": continue\n",
    "        \n",
    "        hom_var_wrt_ref = hom_func_df[substitution_col].apply(\n",
    "            lambda subs: subs_wrt_ref(subs, site_map, \"1\", hom)\n",
    "        )\n",
    "        ret_fs_df.loc[hom_func_df.index.values, \"var_wrt_ref\"] = hom_var_wrt_ref   \n",
    "\n",
    "    # Get list of all allowed substitutions that we will tune beta parameters for\n",
    "    allowed_subs = {\n",
    "        s for subs in ret_fs_df.var_wrt_ref\n",
    "        for s in subs.split()\n",
    "    }\n",
    "    \n",
    "    # Make BinaryMap representations for each homolog\n",
    "    X, y = {}, {}\n",
    "    for homolog, homolog_func_score_df in ret_fs_df.groupby(\"homolog\"):\n",
    "        ref_bmap = bmap.BinaryMap(\n",
    "            homolog_func_score_df,\n",
    "            substitutions_col=\"var_wrt_ref\",\n",
    "            allowed_subs=allowed_subs\n",
    "        )\n",
    "        \n",
    "        # convert binarymaps into sparse arrays for model input\n",
    "        X[homolog] = sparse.BCOO.fromdense(ref_bmap.binary_variants.toarray())\n",
    "        \n",
    "        # create jax array for functional score targets\n",
    "        y[homolog] = jnp.array(homolog_func_score_df[func_score_col].values)\n",
    "    \n",
    "    return (X, y), ret_fs_df.drop([\"wts\", \"sites\", \"muts\"], axis=1), ref_bmap.all_subs    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "id": "75d95406",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1': 'M-G', '2': 'MPP'}\n"
     ]
    }
   ],
   "source": [
    "print(homologs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "id": "41a2c434",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homolog</th>\n",
       "      <th>variant</th>\n",
       "      <th>log2E</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>G3R</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>True</td>\n",
       "      <td>G3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>G3P</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>G3P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E</td>\n",
       "      <td>2.3</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E G3P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>P3R</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>G3R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>P3G</td>\n",
       "      <td>0.4</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3G</td>\n",
       "      <td>2.7</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>M1E P3R</td>\n",
       "      <td>-2.7</td>\n",
       "      <td>True</td>\n",
       "      <td>M1E G3R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  homolog  variant  log2E  allowed_variant var_wrt_ref\n",
       "0       1      M1E    2.0             True         M1E\n",
       "1       1      G3R   -7.0             True         G3R\n",
       "2       1      G3P   -0.5             True         G3P\n",
       "3       2      M1E    2.3             True     M1E G3P\n",
       "4       2      P3R   -5.0             True         G3R\n",
       "5       2      P3G    0.4             True            \n",
       "6       2  M1E P3G    2.7             True         M1E\n",
       "7       2  M1E P3R   -2.7             True     M1E G3R"
      ]
     },
     "execution_count": 425,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(X, y), updated_func_score_df, all_subs = create_homolog_modeling_data(test_df, \"homolog\", \"variant\", \"log2E\")\n",
    "updated_func_score_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
