{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd18d90e",
   "metadata": {},
   "source": [
    "# `multidms` fitting pipeline\n",
    "\n",
    "Here, we demonstrate the pipeline for fitting a `multidms` model on some data using the tools available in the package. Currently, we show how to pre-process data from various dms experiments targeting Delta and Omicron spike protein homologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "53fa951c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# built-in libraries\n",
    "import os\n",
    "import sys\n",
    "# import pickle\n",
    "from itertools import combinations\n",
    "import importlib\n",
    "import math\n",
    "import re\n",
    "# from timeit import default_timer as timer\n",
    "import time\n",
    "import json\n",
    "\n",
    "# external dependencies\n",
    "import pandas as pd\n",
    "# import jax\n",
    "# import jax.numpy as jnp\n",
    "# from jax.experimental import sparse\n",
    "# from jaxopt import ProximalGradient\n",
    "# import jaxopt\n",
    "import numpy as onp\n",
    "# from scipy.stats import pearsonr\n",
    "# import binarymap as bmap\n",
    "from tqdm.notebook import tqdm\n",
    "# import matplotlib.pyplot as plt\n",
    "# import seaborn as sns\n",
    "# import plotnine\n",
    "\n",
    "# local dependencies\n",
    "# sys.path.append(\"..\")\n",
    "# import multidms\n",
    "import multidms\n",
    "# from multidms import Multidms, AAS_WITHSTOP\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3484460c",
   "metadata": {},
   "source": [
    "**First, we'll define a few global variables.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b91f7ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The substitution column to use from the functional score dataframes\n",
    "substitution_column = 'aa_substitutions_reference'\n",
    "\n",
    "# The column providing which individual conditions each variant observation was derived from.\n",
    "experiment_column = 'homolog_exp'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1216729a",
   "metadata": {},
   "source": [
    "**Define all the knobs which define the pipeline's behavior**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8182d3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit_params = {\n",
    "    # pre-processing params\n",
    "    \"scale_counts\" : False,\n",
    "    \"fs_scaling_group_column\" : \"homolog_exp\",\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"pseudocount\" : 0.1,\n",
    "    \"agg_variants\" : True,\n",
    "    \"sample\" : False, #3000,\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"clip_target\" : (-3.5, 2.5),\n",
    "    \"include_stop_variants\" : True,\n",
    "    \"shift_func_score_target_nonref\" : False,\n",
    "    # conditions to include in fit\n",
    "    \"reference_condition\" : \"Delta-2-1\",\n",
    "    \"conditions\" : [\"Delta-2-1\", \"Omicron_BA.1-2-1\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "846274a7",
   "metadata": {},
   "source": [
    "**Define the attributes you'll have access to after each fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fb0739c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cols = list(fit_params.keys()) + [\n",
    "#     \"tuned_model_params\", \"mutation_effects_df\", \"variant_prediction_df\", \"site_map\"\n",
    "# ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42708714",
   "metadata": {},
   "source": [
    "**uncomment if you want to reset all results, or if this is the first run**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3023164a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results = pd.DataFrame(columns = cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89330f69",
   "metadata": {},
   "source": [
    "## Pre-Proccessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f8e5e1",
   "metadata": {},
   "source": [
    "**Read in the dms data and list all available experimental conditions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ee7c2d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>125127</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "0   2021-10-28_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "1   2021-10-28_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "2   2021-10-28_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "3   2021-10-28_thaw-1_VSVG_control_2   Lib-3      thaw-1          2   \n",
       "4   2021-10-28_thaw-1_VSVG_control_1   Lib-4      thaw-1          1   \n",
       "5   2021-10-28_thaw-1_VSVG_control_2   Lib-4      thaw-1          2   \n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "7   2021-10-28_thaw-1_VSVG_control_2   Lib-2      thaw-1          2   \n",
       "8   2022-03-25_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "9   2022-03-25_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "11  2022-06-22_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "12  2022-10-22_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "13  2022-10-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "14  2022-10-22_thaw-2_VSVG_control_1   Lib-1      thaw-2          2   \n",
       "15  2022-10-22_thaw-2_VSVG_control_1   Lib-2      thaw-2          2   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "0   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "0   Lib-1_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   Lib-1_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   Lib-3_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   Lib-3_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   Lib-4_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   Lib-4_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   Lib-2_2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   Lib-1_2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   Lib-1_2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "0   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "1   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "2   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "3   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "4   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "5   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "7   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "8   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "9   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "11  ../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...   \n",
       "12  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "13  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "14  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "15  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "0         library                        pre_sampl...                   45131   \n",
       "1         library                        pre_sampl...                   45131   \n",
       "2         library                        pre_sampl...                   37703   \n",
       "3         library                        pre_sampl...                   37703   \n",
       "4         library                        pre_sampl...                   39694   \n",
       "5         library                        pre_sampl...                   39694   \n",
       "6         library                        pre_sampl...                   41586   \n",
       "7         library                        pre_sampl...                   41586   \n",
       "8         library                        pre_sampl...                   94347   \n",
       "9         library                        pre_sampl...                   94347   \n",
       "10         library                        pre_samp...                  140643   \n",
       "11         library                        pre_samp...                  125127   \n",
       "12        library                        pre_sampl...                   90992   \n",
       "13        library                        pre_sampl...                   85210   \n",
       "14        library                        pre_sampl...                   90992   \n",
       "15        library                        pre_sampl...                   85210   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "0          Delta         Delta-1-1  \n",
       "1          Delta         Delta-1-2  \n",
       "2          Delta         Delta-3-1  \n",
       "3          Delta         Delta-3-2  \n",
       "4          Delta         Delta-4-1  \n",
       "5          Delta         Delta-4-2  \n",
       "6          Delta         Delta-2-1  \n",
       "7          Delta         Delta-2-2  \n",
       "8   Omicron_BA.1  Omicron_BA.1-1-1  \n",
       "9   Omicron_BA.1  Omicron_BA.1-1-2  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  \n",
       "11  Omicron_BA.1  Omicron_BA.1-3-1  \n",
       "12  Omicron_BA.2  Omicron_BA.2-1-1  \n",
       "13  Omicron_BA.2  Omicron_BA.2-2-1  \n",
       "14  Omicron_BA.2  Omicron_BA.2-1-2  \n",
       "15  Omicron_BA.2  Omicron_BA.2-2-2  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = pd.DataFrame()\n",
    "sites = {}\n",
    "wt_seqs = {}\n",
    "\n",
    "for homolog in [\"Delta\", \"Omicron_BA.1\", \"Omicron_BA.2\"]:\n",
    "    \n",
    "    # functional scores\n",
    "    func_sel = pd.read_csv(f\"../results/{homolog}/functional_selections.csv\")\n",
    "    func_sel = func_sel.assign(\n",
    "        filename = f\"../results/{homolog}/\" + \n",
    "        func_sel.library + \"_\" + \n",
    "        func_sel.preselection_sample + \n",
    "        \"_vs_\" + func_sel.postselection_sample + \n",
    "        \"_func_scores.csv\"\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        func_sel_scores_df = func_sel.filename.apply(lambda f: pd.read_csv(f))\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        len_func_sel_scores_df = func_sel.func_sel_scores_df.apply(lambda x: len(x))\n",
    "    )\n",
    "    fun_sel = func_sel.assign(homolog = homolog)\n",
    "    func_score_data = pd.concat([func_score_data, fun_sel]).reset_index(drop=True)\n",
    "\n",
    "# Add a column that gives a unique ID to each homolog/DMS experiment\n",
    "func_score_data['homolog_exp'] = func_score_data.apply(\n",
    "    lambda row: f\"{row['homolog']}-{row['library']}-{row['replicate']}\".replace('-Lib',''),\n",
    "    axis=1\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79383a1d",
   "metadata": {},
   "source": [
    "**Query the conditions to be included in a fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "6408eb32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "6         library                        pre_sampl...                   41586   \n",
       "10         library                        pre_samp...                  140643   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "6          Delta         Delta-2-1  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = func_score_data.query(\n",
    "    f\"{experiment_column}.isin({fit_params['conditions']})\"\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0ec32631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "01a3ae8ee20d401fb5d23ca257e99b63",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df = pd.DataFrame()\n",
    "for idx, row in tqdm(func_score_data.iterrows(), total=len(func_score_data)):\n",
    "    df = row.func_sel_scores_df.assign(homolog=row.homolog)\n",
    "    df = df.assign(library = row.library)\n",
    "    df = df.assign(replicate = row.replicate)\n",
    "    exp_func_score_df = df.assign(homolog_exp=row.homolog_exp)\n",
    "    func_score_df = pd.concat([func_score_df, exp_func_score_df])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed08ea8f",
   "metadata": {},
   "source": [
    "**Optionally subset the variants**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ecc3a404",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params[\"sample\"]:\n",
    "    func_score_df = func_score_df.sample(fit_params[\"sample\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82e1ad2",
   "metadata": {},
   "source": [
    "**remove all variants with string sites and stop codon wildtypes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "349dfa11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e42aee4104c14ed2a4effd22c3ef22e6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/182229 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df.aa_substitutions_reference.fillna(\"\", inplace=True)\n",
    "gapped_sub_vars = []\n",
    "stop_wt_vars = []\n",
    "non_numeric_sites = []\n",
    "for idx, row in tqdm(func_score_df.iterrows(), total=len(func_score_df)):\n",
    "    if \"-\" in row[substitution_column]:\n",
    "        gapped_sub_vars.append(idx)\n",
    "    for sub in row[substitution_column].split():\n",
    "        if sub[0] == \"*\":\n",
    "            stop_wt_vars.append(idx)\n",
    "        if not sub[-2].isnumeric():\n",
    "            non_numeric_sites.append(idx)\n",
    "\n",
    "to_drop = set.union(set(gapped_sub_vars), set(stop_wt_vars), set(non_numeric_sites))\n",
    "func_score_df.drop(to_drop, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf8faff",
   "metadata": {},
   "source": [
    "**Optionally, scale the counts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "364e1f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO re-write and make function\n",
    "# def normalize_by_freq()?\n",
    "if fit_params['scale_counts']:\n",
    "    dfs = []\n",
    "    for (h, hdf) in func_score_df.groupby(fit_params[\"fs_scaling_group_column\"]):\n",
    "        n_post_counts = sum(hdf['post_count'])\n",
    "        if 'Delta' in h:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        else:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        hdf['orig_post_count'] = hdf['post_count']\n",
    "        hdf['post_count'] *= scaling_factor\n",
    "        hdf['post_count_wt'] *= scaling_factor\n",
    "        print(h, n_post_counts, round(scaling_factor, 2), round(sum(hdf['post_count']),2))\n",
    "\n",
    "        # Recompute enrichment ratios with new counts\n",
    "        hdf['pre_count_ps'] = hdf['pre_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_ps'] = hdf['post_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['pre_count_wt_ps'] = hdf['pre_count_wt'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_wt_ps'] = hdf['post_count_wt'] + fit_params[\"pseudocount\"]\n",
    "\n",
    "        total_pre_count = sum(hdf['pre_count_ps'])\n",
    "        total_post_count = sum(hdf['post_count_ps'])\n",
    "\n",
    "        hdf['pre_freq'] = hdf['pre_count_ps'] / total_pre_count\n",
    "        hdf['post_freq'] = hdf['post_count_ps'] / total_post_count\n",
    "        hdf['pre_freq_wt'] = hdf['pre_count_wt_ps'] / total_pre_count\n",
    "        hdf['post_freq_wt'] = hdf['post_count_wt_ps'] / total_post_count\n",
    "\n",
    "        hdf['wt_e'] = hdf['post_freq_wt'] / hdf['pre_freq_wt']\n",
    "        hdf['var_e'] = hdf['post_freq'] / hdf['pre_freq']\n",
    "        hdf['e'] = hdf['var_e'] / hdf['wt_e']\n",
    "        #hdf.dropna(subset=['e'], inplace=True)\n",
    "        hdf['func_score'] = hdf['e'].apply(lambda x: math.log(x, 2))\n",
    "        dfs.append(hdf)\n",
    "    func_score_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0310eb5",
   "metadata": {},
   "source": [
    "**Drop all variants with pre-counts below a threshold.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "678b9752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 169856 variants, 16326 had fewer than 100 counts before selection, and were filtered out\n"
     ]
    }
   ],
   "source": [
    "# Drop barcoded variants with pre-counts below a threshold\n",
    "n_pre_threshold = len(func_score_df)\n",
    "func_score_df = func_score_df[func_score_df['pre_count'] >= fit_params[\"min_pre_counts\"]]\n",
    "print(f\"Of {n_pre_threshold} variants, {n_pre_threshold - len(func_score_df)} had fewer than {fit_params['min_pre_counts']} counts before selection, and were filtered out\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39fd464c",
   "metadata": {},
   "source": [
    "**Optionally throw all variants with stop codons.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "aa118736",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results[\"include_stop_variants\"] = True\n",
    "if not fit_params[\"include_stop_variants\"]:\n",
    "    stop_variants = [idx for idx, row in func_score_df.iterrows() if \"*\" in row[substitution_column]]\n",
    "    func_score_df = func_score_df.drop(stop_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12087cb5",
   "metadata": {},
   "source": [
    "**Optionally, aggregate variant functional scores across barcode replicates**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "bf2ed1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if fit_params[\"agg_variants\"]:\n",
    "#     func_score_df = func_score_df.groupby([substitution_column, experiment_column]).mean().reset_index()\n",
    "# func_score_df[\"pre_count\"] = func_score_df[\"pre_count\"].astype(int)\n",
    "# func_score_df[\"post_count\"] = func_score_df[\"post_count\"].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12f74dad",
   "metadata": {},
   "source": [
    "**Optionally, clip the target functional scores**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "72e18352",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params['clip_target']:\n",
    "    func_score_df[\"func_score\"] = func_score_df[\"func_score\"].clip(*fit_params[\"clip_target\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5a57739c",
   "metadata": {},
   "outputs": [],
   "source": [
    "func_score_df.rename({\"aa_substitutions_reference\":\"aa_substitutions\"}, axis=1, inplace=True)\n",
    "func_score_df.rename({\"homolog_exp\":\"condition\"}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b91fcb28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>pre_sample</th>\n",
       "      <th>post_sample</th>\n",
       "      <th>barcode</th>\n",
       "      <th>func_score</th>\n",
       "      <th>func_score_var</th>\n",
       "      <th>pre_count</th>\n",
       "      <th>post_count</th>\n",
       "      <th>pre_count_wt</th>\n",
       "      <th>post_count_wt</th>\n",
       "      <th>pseudocount</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>aa_substitutions_sequential</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>pre_count_threshold</th>\n",
       "      <th>homolog</th>\n",
       "      <th>replicate</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>CCGAGTAATCCAATAA</td>\n",
       "      <td>-0.4039</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>10237</td>\n",
       "      <td>7909</td>\n",
       "      <td>1909460</td>\n",
       "      <td>1951855</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3</td>\n",
       "      <td>D251A M1235V</td>\n",
       "      <td>2</td>\n",
       "      <td>D253A M1237V</td>\n",
       "      <td>20</td>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>TAAGGACACTCACAAA</td>\n",
       "      <td>-0.3398</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>8528</td>\n",
       "      <td>6888</td>\n",
       "      <td>1909460</td>\n",
       "      <td>1951855</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3</td>\n",
       "      <td>H623L A1014V T1115M</td>\n",
       "      <td>3</td>\n",
       "      <td>H625L A1016V T1117M</td>\n",
       "      <td>20</td>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>ATATTAGTTTATGCCT</td>\n",
       "      <td>-0.1499</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>7979</td>\n",
       "      <td>7351</td>\n",
       "      <td>1909460</td>\n",
       "      <td>1951855</td>\n",
       "      <td>0.5</td>\n",
       "      <td>2</td>\n",
       "      <td>G767H</td>\n",
       "      <td>1</td>\n",
       "      <td>G769H</td>\n",
       "      <td>20</td>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>GAAGATCACTGGTCGA</td>\n",
       "      <td>-0.1554</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>7956</td>\n",
       "      <td>7302</td>\n",
       "      <td>1909460</td>\n",
       "      <td>1951855</td>\n",
       "      <td>0.5</td>\n",
       "      <td>3</td>\n",
       "      <td>P26T S475D</td>\n",
       "      <td>2</td>\n",
       "      <td>P26T S477D</td>\n",
       "      <td>20</td>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>CAAGTGGGTAAATGAT</td>\n",
       "      <td>-1.0530</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>7340</td>\n",
       "      <td>3616</td>\n",
       "      <td>1909460</td>\n",
       "      <td>1951855</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4</td>\n",
       "      <td>G75W D176E P320Q H1099R</td>\n",
       "      <td>4</td>\n",
       "      <td>G75W D178E P322Q H1101R</td>\n",
       "      <td>20</td>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library                        pre_sample  \\\n",
       "0   Lib-2  2021-10-28_thaw-1_VSVG_control_1   \n",
       "1   Lib-2  2021-10-28_thaw-1_VSVG_control_1   \n",
       "2   Lib-2  2021-10-28_thaw-1_VSVG_control_1   \n",
       "3   Lib-2  2021-10-28_thaw-1_VSVG_control_1   \n",
       "4   Lib-2  2021-10-28_thaw-1_VSVG_control_1   \n",
       "\n",
       "                               post_sample           barcode  func_score  \\\n",
       "0  2021-11-28_thaw-1_no-antibody_control_1  CCGAGTAATCCAATAA     -0.4039   \n",
       "1  2021-11-28_thaw-1_no-antibody_control_1  TAAGGACACTCACAAA     -0.3398   \n",
       "2  2021-11-28_thaw-1_no-antibody_control_1  ATATTAGTTTATGCCT     -0.1499   \n",
       "3  2021-11-28_thaw-1_no-antibody_control_1  GAAGATCACTGGTCGA     -0.1554   \n",
       "4  2021-11-28_thaw-1_no-antibody_control_1  CAAGTGGGTAAATGAT     -1.0530   \n",
       "\n",
       "   func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  \\\n",
       "0          0.0005      10237        7909       1909460        1951855   \n",
       "1          0.0005       8528        6888       1909460        1951855   \n",
       "2          0.0005       7979        7351       1909460        1951855   \n",
       "3          0.0005       7956        7302       1909460        1951855   \n",
       "4          0.0009       7340        3616       1909460        1951855   \n",
       "\n",
       "   pseudocount  n_codon_substitutions aa_substitutions_sequential  \\\n",
       "0          0.5                      3                D251A M1235V   \n",
       "1          0.5                      3         H623L A1014V T1115M   \n",
       "2          0.5                      2                       G767H   \n",
       "3          0.5                      3                  P26T S475D   \n",
       "4          0.5                      4     G75W D176E P320Q H1099R   \n",
       "\n",
       "   n_aa_substitutions         aa_substitutions  pre_count_threshold homolog  \\\n",
       "0                   2             D253A M1237V                   20   Delta   \n",
       "1                   3      H625L A1016V T1117M                   20   Delta   \n",
       "2                   1                    G769H                   20   Delta   \n",
       "3                   2               P26T S477D                   20   Delta   \n",
       "4                   4  G75W D178E P322Q H1101R                   20   Delta   \n",
       "\n",
       "   replicate  condition  \n",
       "0          1  Delta-2-1  \n",
       "1          1  Delta-2-1  \n",
       "2          1  Delta-2-1  \n",
       "3          1  Delta-2-1  \n",
       "4          1  Delta-2-1  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb34ecc",
   "metadata": {},
   "source": [
    "## `MultiDmsData`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "8bfc72bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = multidms.MultiDmsData(\n",
    "    func_score_df,\n",
    "    alphabet= multidms.AAS_WITHSTOP,\n",
    "    reference=\"Delta-2-1\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "1a4df78e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('M1F', 'M1I', 'M1K', 'M1L', 'M1T')"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.mutations[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "10448926",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1F</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1K</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1L</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen\n",
       "0      M1F   M      1    F           2\n",
       "1      M1I   M      1    I           8\n",
       "2      M1K   M      1    K           3\n",
       "3      M1L   M      1    L           4\n",
       "4      M1T   M      1    T           7"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "5dde7b39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>4344</td>\n",
       "      <td>-0.158614</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.676400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.118400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.500000</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1016S</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.222967</td>\n",
       "      <td>True</td>\n",
       "      <td>A1016S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                     4344   -0.158614             True   \n",
       "1  Delta-2-1           A1015D       1   -1.676400             True   \n",
       "2  Delta-2-1    A1015D E1188Q       1   -1.118400             True   \n",
       "3  Delta-2-1    A1015D T1027S       1   -3.500000             True   \n",
       "4  Delta-2-1           A1016S       3   -0.222967             True   \n",
       "\n",
       "     var_wrt_ref  \n",
       "0                 \n",
       "1         A1015D  \n",
       "2  A1015D E1188Q  \n",
       "3  A1015D T1027S  \n",
       "4         A1016S  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "677f9109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Delta-2-1', 'Omicron_BA.1-2-1')"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "509b466f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Delta-2-1'"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "7f111753",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Delta-2-1': BCOO(int8[24271, 9655], nse=65004),\n",
       " 'Omicron_BA.1-2-1': BCOO(int8[66785, 9655], nse=2225339)}"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.binarymaps['X'] # this is poorly named, 'y' also exists here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "615f4f84",
   "metadata": {},
   "source": [
    "## `MultiDmsModel`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2483815f",
   "metadata": {},
   "source": [
    "In short the `MultiDmsModel` object initialization includes:\n",
    "   1. initialize attributes\n",
    "   2. initialize params depending on model composits chosen\n",
    "   3. compile model composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6e9f3864",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = multidms.MultiDmsModel(\n",
    "    data,\n",
    "    latent_model=\"phi\",\n",
    "    epistatic_model=\"identity\",\n",
    "    output_activation=\"identity\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "6d1d9a10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['gamma_corrected', 'conditional_shifts', 'conditional_c', '_data', 'params', '_model'])"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.__dict__.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "0fe7c86e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitting, the loss of this model (with initial params) is 11.408\n",
      "Fitting took 18.7742919921875\n",
      "After fitting, the loss of this model (with initial params) is 0.873\n"
     ]
    }
   ],
   "source": [
    "print(f\"Before fitting, the loss of this model (with initial params) is {round(model.loss, 3)}\")\n",
    "start = time.time()\n",
    "model.fit(lasso_shift=1e-5, maxiter=1000)\n",
    "end = time.time()\n",
    "print(f\"Fitting took {end - start}\")\n",
    "print(f\"After fitting, the loss of this model (with initial params) is {round(model.loss, 3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "d4592dbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1F</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1K</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1L</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen\n",
       "0      M1F   M      1    F           2\n",
       "1      M1I   M      1    I           8\n",
       "2      M1K   M      1    K           3\n",
       "3      M1L   M      1    L           4\n",
       "4      M1T   M      1    T           7"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "f022f962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "      <th></th>\n",
       "      <th>S_Delta-2-1</th>\n",
       "      <th>F_Delta-2-1</th>\n",
       "      <th>S_Omicron_BA.1-2-1</th>\n",
       "      <th>F_Omicron_BA.1-2-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1F</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>2</td>\n",
       "      <td>1.361793</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000780</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>8</td>\n",
       "      <td>-1.012618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.373631</td>\n",
       "      <td>-0.235320</td>\n",
       "      <td>-1.608950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1K</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>3</td>\n",
       "      <td>0.092072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.268941</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.268941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1L</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.864001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.225014</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.225014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>7</td>\n",
       "      <td>1.593224</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.232211</td>\n",
       "      <td>-0.454679</td>\n",
       "      <td>0.777532</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen           S_Delta-2-1  F_Delta-2-1  \\\n",
       "0      M1F   M      1    F           2  1.361793          0.0     1.000780   \n",
       "1      M1I   M      1    I           8 -1.012618          0.0    -1.373631   \n",
       "2      M1K   M      1    K           3  0.092072          0.0    -0.268941   \n",
       "3      M1L   M      1    L           4 -0.864001          0.0    -1.225014   \n",
       "4      M1T   M      1    T           7  1.593224          0.0     1.232211   \n",
       "\n",
       "   S_Omicron_BA.1-2-1  F_Omicron_BA.1-2-1  \n",
       "0            0.000000            1.000780  \n",
       "1           -0.235320           -1.608950  \n",
       "2            0.000000           -0.268941  \n",
       "3            0.000000           -1.225014  \n",
       "4           -0.454679            0.777532  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "32cbeb0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>4344</td>\n",
       "      <td>-0.158614</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.676400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.118400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.500000</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1016S</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.222967</td>\n",
       "      <td>True</td>\n",
       "      <td>A1016S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                     4344   -0.158614             True   \n",
       "1  Delta-2-1           A1015D       1   -1.676400             True   \n",
       "2  Delta-2-1    A1015D E1188Q       1   -1.118400             True   \n",
       "3  Delta-2-1    A1015D T1027S       1   -3.500000             True   \n",
       "4  Delta-2-1           A1016S       3   -0.222967             True   \n",
       "\n",
       "     var_wrt_ref  \n",
       "0                 \n",
       "1         A1015D  \n",
       "2  A1015D E1188Q  \n",
       "3  A1015D T1027S  \n",
       "4         A1016S  "
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "7ac38b82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "      <th>predicted_latent</th>\n",
       "      <th>predicted_func_score</th>\n",
       "      <th>corrected_func_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>4344</td>\n",
       "      <td>-0.158614</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>-0.361013</td>\n",
       "      <td>-0.361013</td>\n",
       "      <td>-0.158614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.676400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D</td>\n",
       "      <td>-1.196098</td>\n",
       "      <td>-1.196098</td>\n",
       "      <td>-1.676400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.118400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D E1188Q</td>\n",
       "      <td>-0.999916</td>\n",
       "      <td>-0.999916</td>\n",
       "      <td>-1.118400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.500000</td>\n",
       "      <td>True</td>\n",
       "      <td>A1015D T1027S</td>\n",
       "      <td>-1.297847</td>\n",
       "      <td>-1.297847</td>\n",
       "      <td>-3.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1016S</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.222967</td>\n",
       "      <td>True</td>\n",
       "      <td>A1016S</td>\n",
       "      <td>-0.663131</td>\n",
       "      <td>-0.663131</td>\n",
       "      <td>-0.222967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                     4344   -0.158614             True   \n",
       "1  Delta-2-1           A1015D       1   -1.676400             True   \n",
       "2  Delta-2-1    A1015D E1188Q       1   -1.118400             True   \n",
       "3  Delta-2-1    A1015D T1027S       1   -3.500000             True   \n",
       "4  Delta-2-1           A1016S       3   -0.222967             True   \n",
       "\n",
       "     var_wrt_ref  predicted_latent  predicted_func_score  corrected_func_score  \n",
       "0                        -0.361013             -0.361013             -0.158614  \n",
       "1         A1015D         -1.196098             -1.196098             -1.676400  \n",
       "2  A1015D E1188Q         -0.999916             -0.999916             -1.118400  \n",
       "3  A1015D T1027S         -1.297847             -1.297847             -3.500000  \n",
       "4         A1016S         -0.663131             -0.663131             -0.222967  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "63878edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sigmoid_model = multidms.MultiDmsModel(\n",
    "    data,\n",
    "    latent_model=\"phi\",\n",
    "    epistatic_model=\"sigmoid\",\n",
    "    output_activation=\"identity\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fbedb3f",
   "metadata": {},
   "source": [
    "**The model classes share references to the same data to keep things effecient.** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "bbf5169a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data is sigmoid_model.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "780947da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitting, the loss of this model (with initial params) is 1.8940000000000001\n",
      "After fitting, the loss of this model (with initial params) is 0.9380000000000001\n"
     ]
    }
   ],
   "source": [
    "print(f\"Before fitting, the loss of this model (with initial params) is {round(sigmoid_model.loss, 3)}\")\n",
    "sigmoid_model.fit(lasso_shift=1e-5, maxiter=1000)\n",
    "print(f\"After fitting, the loss of this model (with initial params) is {round(sigmoid_model.loss, 3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "904b45e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C_Delta-2-1': DeviceArray([0.], dtype=float64),\n",
       " 'C_Omicron_BA.1-2-1': DeviceArray([0.], dtype=float64),\n",
       " 'C_ref': DeviceArray([3.16808994], dtype=float64),\n",
       " 'S_Delta-2-1': DeviceArray([0., 0., 0., ..., 0., 0., 0.], dtype=float64),\n",
       " 'S_Omicron_BA.1-2-1': DeviceArray([ 0.        , -0.50970376,  0.        , ...,  0.0292292 ,\n",
       "               0.64317001,  0.54683618], dtype=float64),\n",
       " '': {'ge_bias': DeviceArray([-3.70341545], dtype=float64),\n",
       "  'ge_scale': DeviceArray([3.43341845], dtype=float64)},\n",
       " '': DeviceArray([ 1.76247481, -1.76742336,  0.69877739, ...,  0.00863932,\n",
       "               1.3728258 ,  0.89143468], dtype=float64),\n",
       " '_Delta-2-1': DeviceArray([0.], dtype=float64),\n",
       " '_Omicron_BA.1-2-1': DeviceArray([-0.3866507], dtype=float64)}"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sigmoid_model.params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "da82d388",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loss w/ identity output activation of this model (with initial params) is 0.9380000000000001\n",
      "loss w/ softplus output activation of this model (with initial params) is 0.96\n",
      "loss w/ gelu output activation of this model (with initial params) is 0.988\n"
     ]
    }
   ],
   "source": [
    "for output_act in [\"identity\", \"softplus\", \"gelu\"]:\n",
    "    imodel = multidms.MultiDmsModel(\n",
    "        data,\n",
    "        latent_model=\"phi\",\n",
    "        epistatic_model=\"sigmoid\",\n",
    "        output_activation=output_act\n",
    "    )\n",
    "    imodel.fit(lasso_shift=1e-5, maxiter=1000)\n",
    "    print(f\"loss w/ {output_act} output activation of this model (with initial params) is {round(imodel.loss, 3)}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
