{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd18d90e",
   "metadata": {},
   "source": [
    "# `multidms` fitting pipeline\n",
    "\n",
    "Here, we demonstrate the pipeline for fitting a `multidms` model on some data using the tools available in the package. Currently, we show how to pre-process data from various dms experiments targeting Delta and Omicron spike protein homologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "53fa951c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# built-in libraries\n",
    "import os\n",
    "import sys\n",
    "# import pickle\n",
    "from itertools import combinations\n",
    "import importlib\n",
    "import math\n",
    "import re\n",
    "from timeit import default_timer as timer\n",
    "import json\n",
    "\n",
    "# external dependencies\n",
    "import pandas as pd\n",
    "# import jax\n",
    "# import jax.numpy as jnp\n",
    "# from jax.experimental import sparse\n",
    "# from jaxopt import ProximalGradient\n",
    "# import jaxopt\n",
    "import numpy as onp\n",
    "# from scipy.stats import pearsonr\n",
    "# import binarymap as bmap\n",
    "from tqdm.notebook import tqdm\n",
    "# import matplotlib.pyplot as plt\n",
    "# import seaborn as sns\n",
    "# import plotnine\n",
    "\n",
    "# local dependencies\n",
    "# sys.path.append(\"..\")\n",
    "# import multidms\n",
    "import multidms\n",
    "# from multidms import Multidms, AAS_WITHSTOP\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3484460c",
   "metadata": {},
   "source": [
    "**First, we'll define a few global variables.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "b91f7ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The substitution column to use from the functional score dataframes\n",
    "substitution_column = 'aa_substitutions_reference'\n",
    "\n",
    "# The column providing which individual conditions each variant observation was derived from.\n",
    "experiment_column = 'homolog_exp'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1216729a",
   "metadata": {},
   "source": [
    "**Define all the knobs which define the pipeline's behavior**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8182d3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit_params = {\n",
    "    # pre-processing params\n",
    "    \"scale_counts\" : False,\n",
    "    \"fs_scaling_group_column\" : \"homolog_exp\",\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"pseudocount\" : 0.1,\n",
    "    \"agg_variants\" : True,\n",
    "    \"sample\" : 3000,\n",
    "    \"min_pre_counts\" : 100,\n",
    "    \"clip_target\" : (-3.5, 2.5),\n",
    "    \"include_stop_variants\" : True,\n",
    "    \"shift_func_score_target_nonref\" : False,\n",
    "    # conditions to include in fit\n",
    "    \"reference_condition\" : \"Delta-2-1\",\n",
    "    \"conditions\" : [\"Delta-2-1\", \"Omicron_BA.1-2-1\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "846274a7",
   "metadata": {},
   "source": [
    "**Define the attributes you'll have access to after each fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fb0739c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cols = list(fit_params.keys()) + [\n",
    "#     \"tuned_model_params\", \"mutation_effects_df\", \"variant_prediction_df\", \"site_map\"\n",
    "# ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42708714",
   "metadata": {},
   "source": [
    "**uncomment if you want to reset all results, or if this is the first run**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3023164a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results = pd.DataFrame(columns = cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89330f69",
   "metadata": {},
   "source": [
    "## Pre-Proccessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f8e5e1",
   "metadata": {},
   "source": [
    "**Read in the dms data and list all available experimental conditions**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ee7c2d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>45131</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-3_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>37703</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-3-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-4_2021-12-14_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>39694</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-4-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2</td>\n",
       "      <td>Lib-1_2022-04-13_thaw-1_no-antibody_control_2</td>\n",
       "      <td>Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>94347</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>125127</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-3-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>90992</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-1-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-2</td>\n",
       "      <td>2</td>\n",
       "      <td>2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>85210</td>\n",
       "      <td>Omicron_BA.2</td>\n",
       "      <td>Omicron_BA.2-2-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "0   2021-10-28_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "1   2021-10-28_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "2   2021-10-28_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "3   2021-10-28_thaw-1_VSVG_control_2   Lib-3      thaw-1          2   \n",
       "4   2021-10-28_thaw-1_VSVG_control_1   Lib-4      thaw-1          1   \n",
       "5   2021-10-28_thaw-1_VSVG_control_2   Lib-4      thaw-1          2   \n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "7   2021-10-28_thaw-1_VSVG_control_2   Lib-2      thaw-1          2   \n",
       "8   2022-03-25_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "9   2022-03-25_thaw-1_VSVG_control_2   Lib-1      thaw-1          2   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "11  2022-06-22_thaw-1_VSVG_control_1   Lib-3      thaw-1          1   \n",
       "12  2022-10-22_thaw-1_VSVG_control_1   Lib-1      thaw-1          1   \n",
       "13  2022-10-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "14  2022-10-22_thaw-2_VSVG_control_1   Lib-1      thaw-2          2   \n",
       "15  2022-10-22_thaw-2_VSVG_control_1   Lib-2      thaw-2          2   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "0   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "0   Lib-1_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "1   Lib-1_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "2   Lib-3_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "3   Lib-3_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "4   Lib-4_2021-12-14_thaw-1_no-antibody_control_1   \n",
       "5   Lib-4_2021-12-14_thaw-1_no-antibody_control_2   \n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "7   Lib-2_2021-11-28_thaw-1_no-antibody_control_2   \n",
       "8   Lib-1_2022-04-13_thaw-1_no-antibody_control_1   \n",
       "9   Lib-1_2022-04-13_thaw-1_no-antibody_control_2   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "11  Lib-3_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "12  Lib-1_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "13  Lib-2_2022-10-22_thaw-1_no-antibody_control_1   \n",
       "14  Lib-1_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "15  Lib-2_2022-10-22_thaw-2_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "0   Lib-1_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "1   Lib-1_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "2   Lib-3_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "3   Lib-3_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "4   Lib-4_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "5   Lib-4_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "7   Lib-2_2021-10-28_thaw-1_VSVG_control_2_vs_2021...   \n",
       "8   Lib-1_2022-03-25_thaw-1_VSVG_control_1_vs_2022...   \n",
       "9   Lib-1_2022-03-25_thaw-1_VSVG_control_2_vs_2022...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "11  Lib-3_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "12  Lib-1_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "13  Lib-2_2022-10-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "14  Lib-1_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "15  Lib-2_2022-10-22_thaw-2_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "0   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "1   ../results/Delta/Lib-1_2021-10-28_thaw-1_VSVG_...   \n",
       "2   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "3   ../results/Delta/Lib-3_2021-10-28_thaw-1_VSVG_...   \n",
       "4   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "5   ../results/Delta/Lib-4_2021-10-28_thaw-1_VSVG_...   \n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "7   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "8   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "9   ../results/Omicron_BA.1/Lib-1_2022-03-25_thaw-...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "11  ../results/Omicron_BA.1/Lib-3_2022-06-22_thaw-...   \n",
       "12  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "13  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "14  ../results/Omicron_BA.2/Lib-1_2022-10-22_thaw-...   \n",
       "15  ../results/Omicron_BA.2/Lib-2_2022-10-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "0         library                        pre_sampl...                   45131   \n",
       "1         library                        pre_sampl...                   45131   \n",
       "2         library                        pre_sampl...                   37703   \n",
       "3         library                        pre_sampl...                   37703   \n",
       "4         library                        pre_sampl...                   39694   \n",
       "5         library                        pre_sampl...                   39694   \n",
       "6         library                        pre_sampl...                   41586   \n",
       "7         library                        pre_sampl...                   41586   \n",
       "8         library                        pre_sampl...                   94347   \n",
       "9         library                        pre_sampl...                   94347   \n",
       "10         library                        pre_samp...                  140643   \n",
       "11         library                        pre_samp...                  125127   \n",
       "12        library                        pre_sampl...                   90992   \n",
       "13        library                        pre_sampl...                   85210   \n",
       "14        library                        pre_sampl...                   90992   \n",
       "15        library                        pre_sampl...                   85210   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "0          Delta         Delta-1-1  \n",
       "1          Delta         Delta-1-2  \n",
       "2          Delta         Delta-3-1  \n",
       "3          Delta         Delta-3-2  \n",
       "4          Delta         Delta-4-1  \n",
       "5          Delta         Delta-4-2  \n",
       "6          Delta         Delta-2-1  \n",
       "7          Delta         Delta-2-2  \n",
       "8   Omicron_BA.1  Omicron_BA.1-1-1  \n",
       "9   Omicron_BA.1  Omicron_BA.1-1-2  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  \n",
       "11  Omicron_BA.1  Omicron_BA.1-3-1  \n",
       "12  Omicron_BA.2  Omicron_BA.2-1-1  \n",
       "13  Omicron_BA.2  Omicron_BA.2-2-1  \n",
       "14  Omicron_BA.2  Omicron_BA.2-1-2  \n",
       "15  Omicron_BA.2  Omicron_BA.2-2-2  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = pd.DataFrame()\n",
    "sites = {}\n",
    "wt_seqs = {}\n",
    "\n",
    "for homolog in [\"Delta\", \"Omicron_BA.1\", \"Omicron_BA.2\"]:\n",
    "    \n",
    "    # functional scores\n",
    "    func_sel = pd.read_csv(f\"../results/{homolog}/functional_selections.csv\")\n",
    "    func_sel = func_sel.assign(\n",
    "        filename = f\"../results/{homolog}/\" + \n",
    "        func_sel.library + \"_\" + \n",
    "        func_sel.preselection_sample + \n",
    "        \"_vs_\" + func_sel.postselection_sample + \n",
    "        \"_func_scores.csv\"\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        func_sel_scores_df = func_sel.filename.apply(lambda f: pd.read_csv(f))\n",
    "    )\n",
    "    func_sel = func_sel.assign(\n",
    "        len_func_sel_scores_df = func_sel.func_sel_scores_df.apply(lambda x: len(x))\n",
    "    )\n",
    "    fun_sel = func_sel.assign(homolog = homolog)\n",
    "    func_score_data = pd.concat([func_score_data, fun_sel]).reset_index(drop=True)\n",
    "\n",
    "# Add a column that gives a unique ID to each homolog/DMS experiment\n",
    "func_score_data['homolog_exp'] = func_score_data.apply(\n",
    "    lambda row: f\"{row['homolog']}-{row['library']}-{row['replicate']}\".replace('-Lib',''),\n",
    "    axis=1\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79383a1d",
   "metadata": {},
   "source": [
    "**Query the conditions to be included in a fit**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6408eb32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>preselection_sample</th>\n",
       "      <th>library</th>\n",
       "      <th>virus_batch</th>\n",
       "      <th>replicate</th>\n",
       "      <th>postselection_sample</th>\n",
       "      <th>preselection_library_sample</th>\n",
       "      <th>postselection_library_sample</th>\n",
       "      <th>selection_name</th>\n",
       "      <th>filename</th>\n",
       "      <th>func_sel_scores_df</th>\n",
       "      <th>len_func_sel_scores_df</th>\n",
       "      <th>homolog</th>\n",
       "      <th>homolog_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2021-11-28_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...</td>\n",
       "      <td>../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...</td>\n",
       "      <td>library                        pre_sampl...</td>\n",
       "      <td>41586</td>\n",
       "      <td>Delta</td>\n",
       "      <td>Delta-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2</td>\n",
       "      <td>thaw-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...</td>\n",
       "      <td>../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...</td>\n",
       "      <td>library                        pre_samp...</td>\n",
       "      <td>140643</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 preselection_sample library virus_batch  replicate  \\\n",
       "6   2021-10-28_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "10  2022-06-22_thaw-1_VSVG_control_1   Lib-2      thaw-1          1   \n",
       "\n",
       "                       postselection_sample  \\\n",
       "6   2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "               preselection_library_sample  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1   \n",
       "\n",
       "                     postselection_library_sample  \\\n",
       "6   Lib-2_2021-11-28_thaw-1_no-antibody_control_1   \n",
       "10  Lib-2_2022-06-22_thaw-1_no-antibody_control_1   \n",
       "\n",
       "                                       selection_name  \\\n",
       "6   Lib-2_2021-10-28_thaw-1_VSVG_control_1_vs_2021...   \n",
       "10  Lib-2_2022-06-22_thaw-1_VSVG_control_1_vs_2022...   \n",
       "\n",
       "                                             filename  \\\n",
       "6   ../results/Delta/Lib-2_2021-10-28_thaw-1_VSVG_...   \n",
       "10  ../results/Omicron_BA.1/Lib-2_2022-06-22_thaw-...   \n",
       "\n",
       "                                   func_sel_scores_df  len_func_sel_scores_df  \\\n",
       "6         library                        pre_sampl...                   41586   \n",
       "10         library                        pre_samp...                  140643   \n",
       "\n",
       "         homolog       homolog_exp  \n",
       "6          Delta         Delta-2-1  \n",
       "10  Omicron_BA.1  Omicron_BA.1-2-1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_data = func_score_data.query(\n",
    "    f\"{experiment_column}.isin({fit_params['conditions']})\"\n",
    ")\n",
    "func_score_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0ec32631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aec69dce5c004c7693cc12a9a0499d05",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df = pd.DataFrame()\n",
    "for idx, row in tqdm(func_score_data.iterrows(), total=len(func_score_data)):\n",
    "    df = row.func_sel_scores_df.assign(homolog=row.homolog)\n",
    "    df = df.assign(library = row.library)\n",
    "    df = df.assign(replicate = row.replicate)\n",
    "    exp_func_score_df = df.assign(homolog_exp=row.homolog_exp)\n",
    "    func_score_df = pd.concat([func_score_df, exp_func_score_df])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed08ea8f",
   "metadata": {},
   "source": [
    "**Optionally subset the variants**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ecc3a404",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params[\"sample\"]:\n",
    "    func_score_df = func_score_df.sample(fit_params[\"sample\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d82e1ad2",
   "metadata": {},
   "source": [
    "**remove all variants with string sites and stop codon wildtypes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "349dfa11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5eeeee77072e426da5b425bf60a55c05",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "func_score_df.aa_substitutions_reference.fillna(\"\", inplace=True)\n",
    "gapped_sub_vars = []\n",
    "stop_wt_vars = []\n",
    "non_numeric_sites = []\n",
    "for idx, row in tqdm(func_score_df.iterrows(), total=len(func_score_df)):\n",
    "    if \"-\" in row[substitution_column]:\n",
    "        gapped_sub_vars.append(idx)\n",
    "    for sub in row[substitution_column].split():\n",
    "        if sub[0] == \"*\":\n",
    "            stop_wt_vars.append(idx)\n",
    "        if not sub[-2].isnumeric():\n",
    "            non_numeric_sites.append(idx)\n",
    "\n",
    "to_drop = set.union(set(gapped_sub_vars), set(stop_wt_vars), set(non_numeric_sites))\n",
    "func_score_df.drop(to_drop, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf8faff",
   "metadata": {},
   "source": [
    "**Optionally, scale the counts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "364e1f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO re-write and make function\n",
    "# def normalize_by_freq()?\n",
    "if fit_params['scale_counts']:\n",
    "    dfs = []\n",
    "    for (h, hdf) in func_score_df.groupby(fit_params[\"fs_scaling_group_column\"]):\n",
    "        n_post_counts = sum(hdf['post_count'])\n",
    "        if 'Delta' in h:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        else:\n",
    "            bottleneck = 1e5\n",
    "            scaling_factor = bottleneck / n_post_counts # scaling_factor = 0.05\n",
    "        hdf['orig_post_count'] = hdf['post_count']\n",
    "        hdf['post_count'] *= scaling_factor\n",
    "        hdf['post_count_wt'] *= scaling_factor\n",
    "        print(h, n_post_counts, round(scaling_factor, 2), round(sum(hdf['post_count']),2))\n",
    "\n",
    "        # Recompute enrichment ratios with new counts\n",
    "        hdf['pre_count_ps'] = hdf['pre_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_ps'] = hdf['post_count'] + fit_params[\"pseudocount\"]\n",
    "        hdf['pre_count_wt_ps'] = hdf['pre_count_wt'] + fit_params[\"pseudocount\"]\n",
    "        hdf['post_count_wt_ps'] = hdf['post_count_wt'] + fit_params[\"pseudocount\"]\n",
    "\n",
    "        total_pre_count = sum(hdf['pre_count_ps'])\n",
    "        total_post_count = sum(hdf['post_count_ps'])\n",
    "\n",
    "        hdf['pre_freq'] = hdf['pre_count_ps'] / total_pre_count\n",
    "        hdf['post_freq'] = hdf['post_count_ps'] / total_post_count\n",
    "        hdf['pre_freq_wt'] = hdf['pre_count_wt_ps'] / total_pre_count\n",
    "        hdf['post_freq_wt'] = hdf['post_count_wt_ps'] / total_post_count\n",
    "\n",
    "        hdf['wt_e'] = hdf['post_freq_wt'] / hdf['pre_freq_wt']\n",
    "        hdf['var_e'] = hdf['post_freq'] / hdf['pre_freq']\n",
    "        hdf['e'] = hdf['var_e'] / hdf['wt_e']\n",
    "        #hdf.dropna(subset=['e'], inplace=True)\n",
    "        hdf['func_score'] = hdf['e'].apply(lambda x: math.log(x, 2))\n",
    "        dfs.append(hdf)\n",
    "    func_score_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0310eb5",
   "metadata": {},
   "source": [
    "**Drop all variants with pre-counts below a threshold.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "678b9752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of 2849 variants, 280 had fewer than 100 counts before selection, and were filtered out\n"
     ]
    }
   ],
   "source": [
    "# Drop barcoded variants with pre-counts below a threshold\n",
    "n_pre_threshold = len(func_score_df)\n",
    "func_score_df = func_score_df[func_score_df['pre_count'] >= fit_params[\"min_pre_counts\"]]\n",
    "print(f\"Of {n_pre_threshold} variants, {n_pre_threshold - len(func_score_df)} had fewer than {fit_params['min_pre_counts']} counts before selection, and were filtered out\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39fd464c",
   "metadata": {},
   "source": [
    "**Optionally throw all variants with stop codons.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "aa118736",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results[\"include_stop_variants\"] = True\n",
    "if not fit_params[\"include_stop_variants\"]:\n",
    "    stop_variants = [idx for idx, row in func_score_df.iterrows() if \"*\" in row[substitution_column]]\n",
    "    func_score_df = func_score_df.drop(stop_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12087cb5",
   "metadata": {},
   "source": [
    "**Optionally, aggregate variant functional scores across barcode replicates**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bf2ed1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if fit_params[\"agg_variants\"]:\n",
    "#     func_score_df = func_score_df.groupby([substitution_column, experiment_column]).mean().reset_index()\n",
    "# func_score_df[\"pre_count\"] = func_score_df[\"pre_count\"].astype(int)\n",
    "# func_score_df[\"post_count\"] = func_score_df[\"post_count\"].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12f74dad",
   "metadata": {},
   "source": [
    "**Optionally, clip the target functional scores**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "72e18352",
   "metadata": {},
   "outputs": [],
   "source": [
    "if fit_params['clip_target']:\n",
    "    func_score_df[\"func_score\"] = func_score_df[\"func_score\"].clip(*fit_params[\"clip_target\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5a57739c",
   "metadata": {},
   "outputs": [],
   "source": [
    "func_score_df.rename({\"aa_substitutions_reference\":\"aa_substitutions\"}, axis=1, inplace=True)\n",
    "func_score_df.rename({\"homolog_exp\":\"condition\"}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b91fcb28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>pre_sample</th>\n",
       "      <th>post_sample</th>\n",
       "      <th>barcode</th>\n",
       "      <th>func_score</th>\n",
       "      <th>func_score_var</th>\n",
       "      <th>pre_count</th>\n",
       "      <th>post_count</th>\n",
       "      <th>pre_count_wt</th>\n",
       "      <th>post_count_wt</th>\n",
       "      <th>pseudocount</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>aa_substitutions_sequential</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>pre_count_threshold</th>\n",
       "      <th>homolog</th>\n",
       "      <th>replicate</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4951</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>ACCGGTAACAGGTCAC</td>\n",
       "      <td>-3.5000</td>\n",
       "      <td>0.0294</td>\n",
       "      <td>1469</td>\n",
       "      <td>74</td>\n",
       "      <td>11338665</td>\n",
       "      <td>14040452</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4</td>\n",
       "      <td>D136A R1011G C1233Y G1243S</td>\n",
       "      <td>4</td>\n",
       "      <td>D138A R1014G C1236Y G1246S</td>\n",
       "      <td>66</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2815</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>CGGGGACATAACGACA</td>\n",
       "      <td>-0.1798</td>\n",
       "      <td>0.0023</td>\n",
       "      <td>1738</td>\n",
       "      <td>1900</td>\n",
       "      <td>11338665</td>\n",
       "      <td>14040452</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36799</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>CCTGAAAGTTAGCATA</td>\n",
       "      <td>0.2163</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>590</td>\n",
       "      <td>849</td>\n",
       "      <td>11338665</td>\n",
       "      <td>14040452</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63538</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>TGCGACCACACACGAG</td>\n",
       "      <td>0.3322</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>377</td>\n",
       "      <td>588</td>\n",
       "      <td>11338665</td>\n",
       "      <td>14040452</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55782</th>\n",
       "      <td>Lib-2</td>\n",
       "      <td>2022-06-22_thaw-1_VSVG_control_1</td>\n",
       "      <td>2022-06-22_thaw-1_no-antibody_control_1</td>\n",
       "      <td>TGCTCCATTAATGAAA</td>\n",
       "      <td>-0.1248</td>\n",
       "      <td>0.0092</td>\n",
       "      <td>427</td>\n",
       "      <td>485</td>\n",
       "      <td>11338665</td>\n",
       "      <td>14040452</td>\n",
       "      <td>0.5</td>\n",
       "      <td>4</td>\n",
       "      <td>V67N S146K E658K</td>\n",
       "      <td>3</td>\n",
       "      <td>V67N S151K E661K</td>\n",
       "      <td>66</td>\n",
       "      <td>Omicron_BA.1</td>\n",
       "      <td>1</td>\n",
       "      <td>Omicron_BA.1-2-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      library                        pre_sample  \\\n",
       "4951    Lib-2  2022-06-22_thaw-1_VSVG_control_1   \n",
       "2815    Lib-2  2022-06-22_thaw-1_VSVG_control_1   \n",
       "36799   Lib-2  2022-06-22_thaw-1_VSVG_control_1   \n",
       "63538   Lib-2  2022-06-22_thaw-1_VSVG_control_1   \n",
       "55782   Lib-2  2022-06-22_thaw-1_VSVG_control_1   \n",
       "\n",
       "                                   post_sample           barcode  func_score  \\\n",
       "4951   2022-06-22_thaw-1_no-antibody_control_1  ACCGGTAACAGGTCAC     -3.5000   \n",
       "2815   2022-06-22_thaw-1_no-antibody_control_1  CGGGGACATAACGACA     -0.1798   \n",
       "36799  2022-06-22_thaw-1_no-antibody_control_1  CCTGAAAGTTAGCATA      0.2163   \n",
       "63538  2022-06-22_thaw-1_no-antibody_control_1  TGCGACCACACACGAG      0.3322   \n",
       "55782  2022-06-22_thaw-1_no-antibody_control_1  TGCTCCATTAATGAAA     -0.1248   \n",
       "\n",
       "       func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  \\\n",
       "4951           0.0294       1469          74      11338665       14040452   \n",
       "2815           0.0023       1738        1900      11338665       14040452   \n",
       "36799          0.0060        590         849      11338665       14040452   \n",
       "63538          0.0091        377         588      11338665       14040452   \n",
       "55782          0.0092        427         485      11338665       14040452   \n",
       "\n",
       "       pseudocount  n_codon_substitutions aa_substitutions_sequential  \\\n",
       "4951           0.5                      4  D136A R1011G C1233Y G1243S   \n",
       "2815           0.5                      0                         NaN   \n",
       "36799          0.5                      1                         NaN   \n",
       "63538          0.5                      0                         NaN   \n",
       "55782          0.5                      4            V67N S146K E658K   \n",
       "\n",
       "       n_aa_substitutions            aa_substitutions  pre_count_threshold  \\\n",
       "4951                    4  D138A R1014G C1236Y G1246S                   66   \n",
       "2815                    0                                               66   \n",
       "36799                   0                                               66   \n",
       "63538                   0                                               66   \n",
       "55782                   3            V67N S151K E661K                   66   \n",
       "\n",
       "            homolog  replicate         condition  \n",
       "4951   Omicron_BA.1          1  Omicron_BA.1-2-1  \n",
       "2815   Omicron_BA.1          1  Omicron_BA.1-2-1  \n",
       "36799  Omicron_BA.1          1  Omicron_BA.1-2-1  \n",
       "63538  Omicron_BA.1          1  Omicron_BA.1-2-1  \n",
       "55782  Omicron_BA.1          1  Omicron_BA.1-2-1  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_score_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb34ecc",
   "metadata": {},
   "source": [
    "## `MultiDmsData`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "8bfc72bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 495 site(s) lacking data in at least one condition.\n",
      "901 of the 2076 variants were removed because they had mutations at the above sites, leaving 1175 variants.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 799/799 [00:02<00:00, 272.52it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were 0 cache hits in total for condition Omicron_BA.1-2-1.\n"
     ]
    }
   ],
   "source": [
    "data = multidms.MultiDmsData(\n",
    "    func_score_df,\n",
    "    alphabet= multidms.AAS_WITHSTOP,\n",
    "    reference=\"Delta-2-1\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1a4df78e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('M1I', 'M1T', 'V3A', 'V3F', 'V3G')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.mutations[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "10448926",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V3A</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V3F</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V3G</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen\n",
       "0      M1I   M      1    I           2\n",
       "1      M1T   M      1    T           1\n",
       "2      V3A   V      3    A           1\n",
       "3      V3F   V      3    F           3\n",
       "4      V3G   V      3    G           2"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5dde7b39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>-0.201139</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "      <td>1</td>\n",
       "      <td>0.029400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1080S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.541900</td>\n",
       "      <td>True</td>\n",
       "      <td>A1080S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.650000</td>\n",
       "      <td>True</td>\n",
       "      <td>A222S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222V</td>\n",
       "      <td>1</td>\n",
       "      <td>0.130900</td>\n",
       "      <td>True</td>\n",
       "      <td>A222V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                       66   -0.201139             True   \n",
       "1  Delta-2-1    A1020Y N1135S       1    0.029400             True   \n",
       "3  Delta-2-1           A1080S       1   -0.541900             True   \n",
       "7  Delta-2-1            A222S       1   -0.650000             True   \n",
       "8  Delta-2-1            A222V       1    0.130900             True   \n",
       "\n",
       "     var_wrt_ref  \n",
       "0                 \n",
       "1  A1020Y N1135S  \n",
       "3         A1080S  \n",
       "7          A222S  \n",
       "8          A222V  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "677f9109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Omicron_BA.1-2-1', 'Delta-2-1')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "509b466f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Delta-2-1'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "7f111753",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Delta-2-1': BCOO(int8[376, 1584], nse=836),\n",
       " 'Omicron_BA.1-2-1': BCOO(int8[799, 1584], nse=20461)}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.binarymaps['X'] # 'y' also exists here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "615f4f84",
   "metadata": {},
   "source": [
    "## `MultiDmsModel`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2483815f",
   "metadata": {},
   "source": [
    "In short the `MultiDmsModel` object initialization includes:\n",
    "   1. initialize attributes\n",
    "   2. initialize params depending on model composits chosen\n",
    "   3. compile model composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6e9f3864",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = multidms.MultiDmsModel(\n",
    "    data,\n",
    "    latent_model=\"phi\",\n",
    "    epistatic_model=\"identity\",\n",
    "    output_activation=\"identity\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "6d1d9a10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['gamma_corrected', 'conditional_shifts', 'conditional_c', '_data', 'params', '_model'])"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.__dict__.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0fe7c86e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitting, the loss of this model (with initial params) is 8.121\n",
      "After fitting, the loss of this model (with initial params) is 4.761\n"
     ]
    }
   ],
   "source": [
    "print(f\"Before fitting, the loss of this model (with initial params) is {round(model.loss, 3)}\")\n",
    "model.fit(lasso_shift=1e-5, maxiter=10)\n",
    "print(f\"After fitting, the loss of this model (with initial params) is {round(model.loss, 3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "d4592dbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V3A</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V3F</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V3G</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen\n",
       "0      M1I   M      1    I           2\n",
       "1      M1T   M      1    T           1\n",
       "2      V3A   V      3    A           1\n",
       "3      V3F   V      3    F           3\n",
       "4      V3G   V      3    G           2"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f022f962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>wts</th>\n",
       "      <th>sites</th>\n",
       "      <th>muts</th>\n",
       "      <th>times_seen</th>\n",
       "      <th></th>\n",
       "      <th>S_Omicron_BA.1-2-1</th>\n",
       "      <th>F_Omicron_BA.1-2-1</th>\n",
       "      <th>S_Delta-2-1</th>\n",
       "      <th>F_Delta-2-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M1I</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.824364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.433046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.433046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0.882230</td>\n",
       "      <td>-0.002143</td>\n",
       "      <td>4.137496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.139639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V3A</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.064535</td>\n",
       "      <td>-0.002143</td>\n",
       "      <td>2.190732</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.192874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V3F</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "      <td>0.886644</td>\n",
       "      <td>-0.002143</td>\n",
       "      <td>4.141911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.144054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V3G</td>\n",
       "      <td>V</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.660799</td>\n",
       "      <td>0.002002</td>\n",
       "      <td>1.598612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.596610</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation wts  sites muts  times_seen           S_Omicron_BA.1-2-1  \\\n",
       "0      M1I   M      1    I           2 -0.824364            0.000000   \n",
       "1      M1T   M      1    T           1  0.882230           -0.002143   \n",
       "2      V3A   V      3    A           1 -1.064535           -0.002143   \n",
       "3      V3F   V      3    F           3  0.886644           -0.002143   \n",
       "4      V3G   V      3    G           2 -1.660799            0.002002   \n",
       "\n",
       "   F_Omicron_BA.1-2-1  S_Delta-2-1  F_Delta-2-1  \n",
       "0            2.433046          0.0     2.433046  \n",
       "1            4.137496          0.0     4.139639  \n",
       "2            2.190732          0.0     2.192874  \n",
       "3            4.141911          0.0     4.144054  \n",
       "4            1.598612          0.0     1.596610  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.mutations_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "32cbeb0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>-0.201139</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "      <td>1</td>\n",
       "      <td>0.029400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1080S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.541900</td>\n",
       "      <td>True</td>\n",
       "      <td>A1080S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.650000</td>\n",
       "      <td>True</td>\n",
       "      <td>A222S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222V</td>\n",
       "      <td>1</td>\n",
       "      <td>0.130900</td>\n",
       "      <td>True</td>\n",
       "      <td>A222V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                       66   -0.201139             True   \n",
       "1  Delta-2-1    A1020Y N1135S       1    0.029400             True   \n",
       "3  Delta-2-1           A1080S       1   -0.541900             True   \n",
       "7  Delta-2-1            A222S       1   -0.650000             True   \n",
       "8  Delta-2-1            A222V       1    0.130900             True   \n",
       "\n",
       "     var_wrt_ref  \n",
       "0                 \n",
       "1  A1020Y N1135S  \n",
       "3         A1080S  \n",
       "7          A222S  \n",
       "8          A222V  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "7ac38b82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>weight</th>\n",
       "      <th>func_score</th>\n",
       "      <th>allowed_variant</th>\n",
       "      <th>var_wrt_ref</th>\n",
       "      <th>predicted_latent</th>\n",
       "      <th>predicted_func_score</th>\n",
       "      <th>corrected_func_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>-0.201139</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>3.257409</td>\n",
       "      <td>3.257409</td>\n",
       "      <td>-0.201139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "      <td>1</td>\n",
       "      <td>0.029400</td>\n",
       "      <td>True</td>\n",
       "      <td>A1020Y N1135S</td>\n",
       "      <td>5.090660</td>\n",
       "      <td>5.090660</td>\n",
       "      <td>0.029400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A1080S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.541900</td>\n",
       "      <td>True</td>\n",
       "      <td>A1080S</td>\n",
       "      <td>4.396266</td>\n",
       "      <td>4.396266</td>\n",
       "      <td>-0.541900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222S</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.650000</td>\n",
       "      <td>True</td>\n",
       "      <td>A222S</td>\n",
       "      <td>5.353698</td>\n",
       "      <td>5.353698</td>\n",
       "      <td>-0.650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Delta-2-1</td>\n",
       "      <td>A222V</td>\n",
       "      <td>1</td>\n",
       "      <td>0.130900</td>\n",
       "      <td>True</td>\n",
       "      <td>A222V</td>\n",
       "      <td>2.816859</td>\n",
       "      <td>2.816859</td>\n",
       "      <td>0.130900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   condition aa_substitutions  weight  func_score  allowed_variant  \\\n",
       "0  Delta-2-1                       66   -0.201139             True   \n",
       "1  Delta-2-1    A1020Y N1135S       1    0.029400             True   \n",
       "3  Delta-2-1           A1080S       1   -0.541900             True   \n",
       "7  Delta-2-1            A222S       1   -0.650000             True   \n",
       "8  Delta-2-1            A222V       1    0.130900             True   \n",
       "\n",
       "     var_wrt_ref  predicted_latent  predicted_func_score  corrected_func_score  \n",
       "0                         3.257409              3.257409             -0.201139  \n",
       "1  A1020Y N1135S          5.090660              5.090660              0.029400  \n",
       "3         A1080S          4.396266              4.396266             -0.541900  \n",
       "7          A222S          5.353698              5.353698             -0.650000  \n",
       "8          A222V          2.816859              2.816859              0.130900  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.variants_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "63878edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sigmoid_model = multidms.MultiDmsModel(\n",
    "    data,\n",
    "    latent_model=\"phi\",\n",
    "    epistatic_model=\"sigmoid\",\n",
    "    output_activation=\"identity\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fbedb3f",
   "metadata": {},
   "source": [
    "**The model classes share references to the same data to keep things effecient.** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "bbf5169a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.data is sigmoid_model.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "780947da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitting, the loss of this model (with initial params) is 1.437\n",
      "After fitting, the loss of this model (with initial params) is 1.437\n"
     ]
    }
   ],
   "source": [
    "print(f\"Before fitting, the loss of this model (with initial params) is {round(sigmoid_model.loss, 3)}\")\n",
    "sigmoid_model.fit(lasso_shift=1e-5, maxiter=10)\n",
    "print(f\"After fitting, the loss of this model (with initial params) is {round(sigmoid_model.loss, 3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "da82d388",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Before fitting, the loss of this model (with initial params) is 1.691\n",
      "After fitting, the loss of this model (with initial params) is 1.437\n"
     ]
    }
   ],
   "source": [
    "sigmoid_softplus_model = multidms.MultiDmsModel(\n",
    "    data,\n",
    "    latent_model=\"phi\",\n",
    "    epistatic_model=\"sigmoid\",\n",
    "    output_activation=\"softplus\"\n",
    ")\n",
    "print(f\"Before fitting, the loss of this model (with initial params) is {round(sigmoid_softplus_model.loss, 3)}\")\n",
    "sigmoid_softplus_model.fit(lasso_shift=1e-5, maxiter=10)\n",
    "print(f\"After fitting, the loss of this model (with initial params) is {round(sigmoid_softplus_model.loss, 3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "3f51a9a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sigmoid_model.variants_df.predicted_func_score.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "904b45e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C_Delta-2-1': DeviceArray([0.], dtype=float64),\n",
       " 'C_Omicron_BA.1-2-1': DeviceArray([0.], dtype=float64),\n",
       " 'C_ref': DeviceArray([4.96247717], dtype=float64),\n",
       " 'S_Delta-2-1': DeviceArray([0., 0., 0., ..., 0., 0., 0.], dtype=float64),\n",
       " 'S_Omicron_BA.1-2-1': DeviceArray([ 0.        , -0.00036382, -0.00036382, ...,  0.        ,\n",
       "               0.        ,  0.00152727], dtype=float64),\n",
       " '': {'ge_bias': DeviceArray([0.], dtype=float64),\n",
       "  'ge_scale': DeviceArray([0.], dtype=float64)},\n",
       " '': DeviceArray([-0.8346446 ,  0.88331895, -1.06344564, ..., -1.85269346,\n",
       "               0.87187642, -0.19739826], dtype=float64),\n",
       " '_Delta-2-1': DeviceArray([0.], dtype=float64),\n",
       " '_Omicron_BA.1-2-1': DeviceArray([0.61121915], dtype=float64)}"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sigmoid_model.params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47b2a0ec",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
