{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b0db34cc",
   "metadata": {},
   "source": [
    "# Biophysical model\n",
    "\n",
    "The `multidms` joint model applies to a case where you have DMS datasets\n",
    "for two or more experimental conditions and are interested in identifying shifts\n",
    "in mutational effects between conditions. \n",
    "Here we describe the biophysical model of conditional shifts that motivates the approach in this package.\n",
    "\n",
    "We suggest reading the [Otwinowski et al. 2018](https://www.pnas.org/doi/10.1073/pnas.1804015115) paper to understand the approach to modeling global epistasis before reading the rest of the documentation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f510168f",
   "metadata": {},
   "source": [
    "## Model overview\n",
    "\n",
    "`multidms` extends the traditional global epistasis model by informing the parameters with multiple dms experiments under differing experimental conditions. For example, distinct experimental conditions (referred to as $h$ throughout the documentation) may include sets of experiments that are performed with differing homology of wildtype sequence (i.e. SARS-CoV-2 Delta Vs. Omicron spike). This approach could also be used for experimental conditions which share homology of the wildtype but differing selection targets of the same protein. This model makes the assumption that differing conditions should result in _mostly_ the same effect of mutations -- but some shifts in mutational effect due to biological mechanisms (i.e. epistasis). Ultimately, this model was designed to identify those shifts using lasso regularization applied to the shift parameters described in the additive latent phenotype section below.\n",
    "\n",
    "[//]: # \"We find that the qualitative results are robust to choice for lasso strength, and generally this lasso\"\n",
    "[//]: # \"acts as a single dial to increase signal and noise in a linear fasion\"\n",
    "\n",
    "\n",
    "At a high level, the model is a composition of three\n",
    "functions which describe the expected biophysical interactions underlying a given phenotype;\n",
    "(1) an additive model, $\\phi$, describing a variant's _latent_ phenotype under a given condition,\n",
    "(2) a global epistasis model, $g$, to disentangle the effects of multiple mutations on the same variant, and\n",
    "(3) a final output activation function, $t$, accounting for an expected _lower bound_ on the variant's phenotype,\n",
    "where the observed functional score may sit below due to experimental sensitivity.\n",
    "\n",
    "$$\n",
    "f(v,h) = t_{\\gamma}(g_{\\alpha}(\\phi_{\\beta, S, C_{r}}(v,h))\n",
    "$$\n",
    "\n",
    "Where \n",
    "$\\gamma$, $\\alpha$, $\\beta$, $S$, and $C_{r}$\n",
    "are _free_ parameters inferred from experimental observations during the fitting process.\n",
    "\n",
    "**Note** The motivation behind defining an abstract model in terms of its components offers us (1) modularity for method testing and development, and (2) The flexibility of multiple options for model components that encompass the needs of differing research goals and experimental techniques. While there is only a single option in for the $\\phi$ latent prediction, we offer a few options both for the global epistasis ($g$) and output activation ($t$) functions. Generally, the package defaults for these components described below should be sufficient for most purposes, and in this case feel free to ignore the `multidms.biophysical` module all-together as this functionality is generally hidden unless explicitly specified during the instantiation of a `MultiDmsModel` object. \n",
    "\n",
    "Below, we'll describe the individual components in more detail."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f967d19d",
   "metadata": {},
   "source": [
    "## Additive latent phenotype, $\\phi$\n",
    "\n",
    "The model defines one condition as a _reference_ condition.\n",
    "For each mutation $m$, the model fits a single mutation effect parameter, $\\beta_{m}$.\n",
    "Additionally, the model fits set of shift parameters, $S_{m,h}$, \n",
    "that quantifies the shift a given mutation's\n",
    "effect. Each mutation is associated with an independent shift parameter \n",
    "for each non-reference condition. \n",
    "For example if there exists 3 total experimental conditions, $h_{1}$, $h_{2}$, & $h_{3}$, \n",
    "then each mutation, $m$, will be get assigned a\n",
    "single $\\beta_{m}$ parameter, \n",
    "and two non-reference condition _shift_ parameters $S_{m, h_{2}}$, $S_{m, h_{3}}$ for the latent prediction\n",
    "\n",
    "\n",
    "Shift parameters can be regularized, encouraging most of them to be\n",
    "close to zero. This regularization step is a useful way to eliminate\n",
    "the effects of experimental noise, and is most useful in cases where\n",
    "you expect most mutations to have the same effects between conditions,\n",
    "such as for conditions that are close relatives. \n",
    "\n",
    "Concretely, the latent phenotype of any variant, $v$, from the experimental condition, $h$,\n",
    "is computed like so:\n",
    "\n",
    "$$\n",
    "\\phi(v,h) = c_{r} + \\sum_{m \\in v} (\\beta_{m} + S_{m,h})\n",
    "$$\n",
    "\n",
    "where:\n",
    "\n",
    "* $c_{r}$ is the wild type latent phenotype for the reference condition.\n",
    "* $\\beta_{m}$ is the latent phenotypic effect of mutation $m$. See the note below\n",
    "* $s_{m,h}$ is the shift of the effect of mutation $m$ in condition $h$.\n",
    "  These parameters are fixed to zero for the reference condition. For\n",
    "  non-reference conditions, they are defined in the same way as $x_m$ parameters.\n",
    "* $v$ is the set of all mutations relative to the reference wild type sequence\n",
    "  (including all mutations that separate condition $h$ from the reference condition).\n",
    "\n",
    "**Note** This function is implimented in `multidms.biophysical.additive_model`\n",
    "\n",
    "**Note** The $\\beta_m$ variable is defined such that mutations are always relative to the\n",
    "reference condition. For example, if the wild type amino acid at site 30 is an\n",
    "A in the reference condition, and a G in a non-reference condition,\n",
    "then a Y30G mutation in the non-reference condition is recorded as an A30G\n",
    "mutation relative to the reference. This way, each condition informs\n",
    "the exact same parameters, even at sites that differ in wild type amino acid.\n",
    "These are encoded in a `BinaryMap` object, where all sites that are non-identical\n",
    "to the reference are 1's."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd8613ea",
   "metadata": {},
   "source": [
    "## Sigmoidal global epistasis, $g$\n",
    "\n",
    "In contrast to the common family of spline functions typically applied to these models, \n",
    "the global-epistasis function here assumes a sigmoidal relationship between\n",
    "a protein's latent property and it's functional score measured in the experiment\n",
    "(e.g., log enrichment score). Using free parameters, the sigmoid\n",
    "can flexibly conform to an optimal shape informed by the data. \n",
    "Note that this function is independent from the\n",
    "experimental condition from which a variant is observed.\n",
    "\n",
    "The sigmoidal function that relates a given _latent phenotype_, $z$, to its functional score is given by:\n",
    "\n",
    "$$\n",
    "g(z) = \\alpha_{scale} * \\frac{1}{1 + e^{-z}} + \\alpha_{bias}\n",
    "$$\n",
    "\n",
    "where:\n",
    "* $\\alpha_{scale}$ is a free parameter defining the range of the sigmoid\n",
    "* $\\alpha_{bias}$ is a free parameter defining the lower bound of the sigmoid.\n",
    "\n",
    "Below is an interactive plot showing the effect of the sigmoidal global epistasis as a function of an adjustable $\\alpha_{scale}$, and $\\alpha_{bias}$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8d3e310c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-533c847c40d143efa8aae53488e20ae4\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-533c847c40d143efa8aae53488e20ae4\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-533c847c40d143efa8aae53488e20ae4\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.6.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.6.1\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-5df0d8288b8899ccb5d4c64cc03b51be\"}, \"mark\": {\"type\": \"line\"}, \"encoding\": {\"x\": {\"field\": \"latent\", \"scale\": {\"domain\": [-10, 10]}, \"title\": \"latent phenotype\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"phenotype\", \"scale\": {\"domain\": [-10, 10]}, \"title\": \"predicted phenotype\", \"type\": \"quantitative\"}}, \"params\": [{\"name\": \"alpha_scale\", \"bind\": {\"input\": \"range\", \"max\": 10, \"min\": 0.1}, \"value\": 1}, {\"name\": \"alpha_bias\", \"bind\": {\"input\": \"range\", \"max\": 5, \"min\": -10}, \"value\": 0}], \"transform\": [{\"calculate\": \"(((1 / (1 + exp((-1 * datum['latent'])))) * alpha_scale) + alpha_bias)\", \"as\": \"phenotype\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.6.1.json\", \"datasets\": {\"data-5df0d8288b8899ccb5d4c64cc03b51be\": [{\"latent\": -10.0}, {\"latent\": -9.797979797979798}, {\"latent\": -9.595959595959595}, {\"latent\": -9.393939393939394}, {\"latent\": -9.191919191919192}, {\"latent\": -8.98989898989899}, {\"latent\": -8.787878787878787}, {\"latent\": -8.585858585858587}, {\"latent\": -8.383838383838384}, {\"latent\": -8.181818181818182}, {\"latent\": -7.979797979797979}, {\"latent\": -7.777777777777778}, {\"latent\": -7.575757575757576}, {\"latent\": -7.373737373737374}, {\"latent\": -7.171717171717171}, {\"latent\": -6.96969696969697}, {\"latent\": -6.767676767676768}, {\"latent\": -6.565656565656566}, {\"latent\": -6.363636363636363}, {\"latent\": -6.161616161616162}, {\"latent\": -5.959595959595959}, {\"latent\": -5.757575757575758}, {\"latent\": -5.555555555555555}, {\"latent\": -5.353535353535354}, {\"latent\": -5.151515151515151}, {\"latent\": -4.94949494949495}, {\"latent\": -4.747474747474747}, {\"latent\": -4.545454545454546}, {\"latent\": -4.343434343434343}, {\"latent\": -4.141414141414142}, {\"latent\": -3.9393939393939394}, {\"latent\": -3.737373737373738}, {\"latent\": -3.5353535353535355}, {\"latent\": -3.333333333333333}, {\"latent\": -3.1313131313131315}, {\"latent\": -2.929292929292929}, {\"latent\": -2.7272727272727275}, {\"latent\": -2.525252525252525}, {\"latent\": -2.3232323232323235}, {\"latent\": -2.121212121212121}, {\"latent\": -1.9191919191919187}, {\"latent\": -1.717171717171718}, {\"latent\": -1.5151515151515156}, {\"latent\": -1.3131313131313131}, {\"latent\": -1.1111111111111107}, {\"latent\": -0.9090909090909101}, {\"latent\": -0.7070707070707076}, {\"latent\": -0.5050505050505052}, {\"latent\": -0.30303030303030276}, {\"latent\": -0.10101010101010033}, {\"latent\": 0.10101010101010033}, {\"latent\": 0.30303030303030276}, {\"latent\": 0.5050505050505052}, {\"latent\": 0.7070707070707076}, {\"latent\": 0.9090909090909083}, {\"latent\": 1.1111111111111107}, {\"latent\": 1.3131313131313131}, {\"latent\": 1.5151515151515156}, {\"latent\": 1.7171717171717162}, {\"latent\": 1.9191919191919187}, {\"latent\": 2.121212121212121}, {\"latent\": 2.3232323232323235}, {\"latent\": 2.525252525252524}, {\"latent\": 2.7272727272727266}, {\"latent\": 2.929292929292929}, {\"latent\": 3.1313131313131315}, {\"latent\": 3.333333333333334}, {\"latent\": 3.5353535353535346}, {\"latent\": 3.737373737373737}, {\"latent\": 3.9393939393939394}, {\"latent\": 4.141414141414142}, {\"latent\": 4.3434343434343425}, {\"latent\": 4.545454545454545}, {\"latent\": 4.747474747474747}, {\"latent\": 4.94949494949495}, {\"latent\": 5.1515151515151505}, {\"latent\": 5.353535353535353}, {\"latent\": 5.555555555555555}, {\"latent\": 5.757575757575758}, {\"latent\": 5.9595959595959584}, {\"latent\": 6.161616161616163}, {\"latent\": 6.363636363636363}, {\"latent\": 6.565656565656564}, {\"latent\": 6.767676767676768}, {\"latent\": 6.969696969696969}, {\"latent\": 7.171717171717173}, {\"latent\": 7.373737373737374}, {\"latent\": 7.575757575757574}, {\"latent\": 7.777777777777779}, {\"latent\": 7.979797979797979}, {\"latent\": 8.18181818181818}, {\"latent\": 8.383838383838384}, {\"latent\": 8.585858585858585}, {\"latent\": 8.787878787878789}, {\"latent\": 8.98989898989899}, {\"latent\": 9.19191919191919}, {\"latent\": 9.393939393939394}, {\"latent\": 9.595959595959595}, {\"latent\": 9.7979797979798}, {\"latent\": 10.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"latent\": numpy.linspace(-10, 10, 100)})\n",
    "\n",
    "slider_s = alt.binding_range(min=0.1, max=10)\n",
    "var_s = alt.param(bind=slider_s, value=1, name=\"alpha_scale\")\n",
    "\n",
    "slider_b = alt.binding_range(min=-10, max=5) #, step=1)\n",
    "var_b = alt.param(bind=slider_b, value=0, name=\"alpha_bias\")\n",
    "\n",
    "(\n",
    "    alt.Chart(df)\n",
    "    .transform_calculate(\n",
    "        phenotype=(1 / (1 + alt.expr.exp(-1*alt.datum['latent'])))\n",
    "        * var_s\n",
    "        + var_b\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"latent\", title=\"latent phenotype\", scale=alt.Scale(domain=[-10, 10])),\n",
    "        y=alt.Y(\"phenotype:Q\", title=\"predicted phenotype\", scale=alt.Scale(domain=[-10, 10]))\n",
    "    )\n",
    "    .mark_line()\n",
    "    .add_params(var_s, var_b)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cef6407",
   "metadata": {},
   "source": [
    "## Output activation, $t$\n",
    "\n",
    "DMS reported phenotypes are often the result of a fold-change calculation on barcode read counts after a selection stage as compared to the pre-selection barcode read counts. Often this means that a lower frequency of variants in the initial library will often result in far more extreme observation of phenotype than we expect the sensitivity of this protocol to detect. In this case, one option is to filter the data to exclude these low frequency barcodes -- However in certain cases even a modest filtering threshold cuts out many of the individual mutations you may want to learn about from the dataset. In this case, it's common for researchers to clip their data at some lower bound, $l$. This type of clipping causes a specific unwanted behavior in the model. Put simply, the model, $g(\\phi(v,h))$ _learns_ this lower bound and thus is encouraged to limit it's shape to conform to this lower bound. \n",
    "\n",
    "We account for this unwanted behavior by applying a final activation on $z'$, where the output of the global epistasis model $g(\\phi(v,h)) = z$. In essence, this is a generic softplus activation, softplus($z'$) = $\\log(1 + e^{z'})$, a smooth approximation of the well-known ReLU activation function. Because sofplus is a rigid function with a lower bound at zero, we introduced; (1) a user defined parameter, $\\lambda_{sp}$ which scales the rate at which the \"hinge\" interpolates from the lower bound to the identity function, and (2) a user-defined parameter to shift the lower bound to where the data was clipped, $l$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5772b058",
   "metadata": {},
   "source": [
    "Given the output of a our global epistasis function, $z'$, the final predicted phenotype of our model is computed like so:\n",
    "\n",
    "$$\n",
    "t(z') = \\lambda_{sp}\\log(1 + e^{\\frac{z' - l}{\\lambda_{sp}}}) + l\n",
    "$$\n",
    "\n",
    "\n",
    "Below is an interactive plot showing the effect of the modified softplus activation as a function of an adjustable $\\lambda_{sp}$ scaling parameter, and lower bound, $l$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "61188e01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-8dc8da26022b45689d6443ea6eb2108e\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-8dc8da26022b45689d6443ea6eb2108e\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-8dc8da26022b45689d6443ea6eb2108e\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.6.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.6.1\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-5df0d8288b8899ccb5d4c64cc03b51be\"}, \"mark\": {\"type\": \"line\"}, \"encoding\": {\"x\": {\"field\": \"latent\", \"scale\": {\"domain\": [-10, 10]}, \"title\": \"global epistasis prediction (z')\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"phenotype\", \"scale\": {\"domain\": [-10, 10]}, \"title\": \"predicted phenotype\", \"type\": \"quantitative\"}}, \"params\": [{\"name\": \"lambda_sp\", \"bind\": {\"input\": \"range\", \"max\": 10, \"min\": 0.1}, \"value\": 1}, {\"name\": \"lower_bound\", \"bind\": {\"input\": \"range\", \"max\": 0, \"min\": -10}, \"value\": -3.5}], \"transform\": [{\"calculate\": \"((log((1 + exp(((datum['latent'] - lower_bound) / lambda_sp)))) * lambda_sp) + lower_bound)\", \"as\": \"phenotype\"}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.6.1.json\", \"datasets\": {\"data-5df0d8288b8899ccb5d4c64cc03b51be\": [{\"latent\": -10.0}, {\"latent\": -9.797979797979798}, {\"latent\": -9.595959595959595}, {\"latent\": -9.393939393939394}, {\"latent\": -9.191919191919192}, {\"latent\": -8.98989898989899}, {\"latent\": -8.787878787878787}, {\"latent\": -8.585858585858587}, {\"latent\": -8.383838383838384}, {\"latent\": -8.181818181818182}, {\"latent\": -7.979797979797979}, {\"latent\": -7.777777777777778}, {\"latent\": -7.575757575757576}, {\"latent\": -7.373737373737374}, {\"latent\": -7.171717171717171}, {\"latent\": -6.96969696969697}, {\"latent\": -6.767676767676768}, {\"latent\": -6.565656565656566}, {\"latent\": -6.363636363636363}, {\"latent\": -6.161616161616162}, {\"latent\": -5.959595959595959}, {\"latent\": -5.757575757575758}, {\"latent\": -5.555555555555555}, {\"latent\": -5.353535353535354}, {\"latent\": -5.151515151515151}, {\"latent\": -4.94949494949495}, {\"latent\": -4.747474747474747}, {\"latent\": -4.545454545454546}, {\"latent\": -4.343434343434343}, {\"latent\": -4.141414141414142}, {\"latent\": -3.9393939393939394}, {\"latent\": -3.737373737373738}, {\"latent\": -3.5353535353535355}, {\"latent\": -3.333333333333333}, {\"latent\": -3.1313131313131315}, {\"latent\": -2.929292929292929}, {\"latent\": -2.7272727272727275}, {\"latent\": -2.525252525252525}, {\"latent\": -2.3232323232323235}, {\"latent\": -2.121212121212121}, {\"latent\": -1.9191919191919187}, {\"latent\": -1.717171717171718}, {\"latent\": -1.5151515151515156}, {\"latent\": -1.3131313131313131}, {\"latent\": -1.1111111111111107}, {\"latent\": -0.9090909090909101}, {\"latent\": -0.7070707070707076}, {\"latent\": -0.5050505050505052}, {\"latent\": -0.30303030303030276}, {\"latent\": -0.10101010101010033}, {\"latent\": 0.10101010101010033}, {\"latent\": 0.30303030303030276}, {\"latent\": 0.5050505050505052}, {\"latent\": 0.7070707070707076}, {\"latent\": 0.9090909090909083}, {\"latent\": 1.1111111111111107}, {\"latent\": 1.3131313131313131}, {\"latent\": 1.5151515151515156}, {\"latent\": 1.7171717171717162}, {\"latent\": 1.9191919191919187}, {\"latent\": 2.121212121212121}, {\"latent\": 2.3232323232323235}, {\"latent\": 2.525252525252524}, {\"latent\": 2.7272727272727266}, {\"latent\": 2.929292929292929}, {\"latent\": 3.1313131313131315}, {\"latent\": 3.333333333333334}, {\"latent\": 3.5353535353535346}, {\"latent\": 3.737373737373737}, {\"latent\": 3.9393939393939394}, {\"latent\": 4.141414141414142}, {\"latent\": 4.3434343434343425}, {\"latent\": 4.545454545454545}, {\"latent\": 4.747474747474747}, {\"latent\": 4.94949494949495}, {\"latent\": 5.1515151515151505}, {\"latent\": 5.353535353535353}, {\"latent\": 5.555555555555555}, {\"latent\": 5.757575757575758}, {\"latent\": 5.9595959595959584}, {\"latent\": 6.161616161616163}, {\"latent\": 6.363636363636363}, {\"latent\": 6.565656565656564}, {\"latent\": 6.767676767676768}, {\"latent\": 6.969696969696969}, {\"latent\": 7.171717171717173}, {\"latent\": 7.373737373737374}, {\"latent\": 7.575757575757574}, {\"latent\": 7.777777777777779}, {\"latent\": 7.979797979797979}, {\"latent\": 8.18181818181818}, {\"latent\": 8.383838383838384}, {\"latent\": 8.585858585858585}, {\"latent\": 8.787878787878789}, {\"latent\": 8.98989898989899}, {\"latent\": 9.19191919191919}, {\"latent\": 9.393939393939394}, {\"latent\": 9.595959595959595}, {\"latent\": 9.7979797979798}, {\"latent\": 10.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"latent\": numpy.linspace(-10, 10, 100)})\n",
    "\n",
    "slider_lsp = alt.binding_range(min=0.1, max=10)\n",
    "var_lambda_sp = alt.param(bind=slider_lsp, value=1, name=\"lambda_sp\")\n",
    "\n",
    "slider_lb = alt.binding_range(min=-10, max=0)\n",
    "var_lower_bound = alt.param(bind=slider_lb, value=-3.5, name=\"lower_bound\")\n",
    "\n",
    "(\n",
    "    alt.Chart(df)\n",
    "    .transform_calculate(\n",
    "        phenotype=alt.expr.log(1 + alt.expr.exp((alt.datum['latent']-var_lower_bound)/var_lambda_sp))\n",
    "        * var_lambda_sp\n",
    "        + var_lower_bound\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\"latent\", title=\"global epistasis prediction (z')\", scale=alt.Scale(domain=[-10, 10])),\n",
    "        y=alt.Y(\"phenotype:Q\", title=\"predicted phenotype\", scale=alt.Scale(domain=[-10, 10]))\n",
    "    )\n",
    "    .mark_line()\n",
    "    .add_params(var_lambda_sp, var_lower_bound)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5f419be",
   "metadata": {},
   "source": [
    "## Accounting for the experimental bias using $\\gamma_{h}$ "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ab6fb6a",
   "metadata": {},
   "source": [
    "We expect that there exists some constant bias in the observed functional scores between differing experimental conditions. We believe this to be the case esspecially when experimental conditions have differing wildtypes. Ideally, in this scenerio we could spike-in the reference wildtype in each non reference experiment to guage exactly what that difference may be and adjust accordingly. However, this is not always feasible in the library design, and an analogous procedure for experiments sharing homology is unclear. To account for this bias, we introduce a free parameter, $\\gamma$\n",
    "\n",
    "TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38bc52e5",
   "metadata": {},
   "source": [
    "## TODO - Fitting procedure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "109ff5f0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
