
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>API reference &#8212; multidms 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script src="_static/jquery.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Acknowledgments" href="acknowledgments.html" />
    <link rel="prev" title="multidms fitting pipeline" href="fit_delta_BA1_example.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="autodoc"></span><section id="api-reference">
<h1>API reference<a class="headerlink" href="#api-reference" title="Permalink to this heading"></a></h1>
<p>Here, we document the the details of the <cite>multidms</cite> Python interface.</p>
<span class="target" id="module-data"></span><section id="multidmsdata">
<h2>MultiDmsData<a class="headerlink" href="#multidmsdata" title="Permalink to this heading"></a></h2>
<p>Defines <code class="xref py py-class docutils literal notranslate"><span class="pre">Multidms</span></code> objects for handling data from one or more
dms experiments under various conditions.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="data.MultiDmsData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">data.</span></span><span class="sig-name descname"><span class="pre">MultiDmsData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variants_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_identical_variants</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('#0072B2',</span> <span class="pre">'#009E73',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#56B4E9',</span> <span class="pre">'#F0E442')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letter_suffixed_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assert_site_integrity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_non_shared_subs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#data.MultiDmsData" title="Permalink to this definition"></a></dt>
<dd><p>Prep data for multidms model(s),
Summarize, and provide static data attributes.
Individual objects of this type can be shared
by multiple <code class="docutils literal notranslate"><span class="pre">multidms.MultiDmsModel</span></code> Objects
for effeciently fitting various models to the same data.</p>
<section id="note">
<h2>Note<a class="headerlink" href="#note" title="Permalink to this heading"></a></h2>
<p>You can initialize a <code class="xref py py-class docutils literal notranslate"><span class="pre">Multidms</span></code> object with a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>
with a row for each variant sampled and annotations
provided in the required columns:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><cite>condition</cite> - Experimental condition from</dt><dd><p>which a sample measurement was obtained.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>aa_substitutions</cite> - Defines each variant</dt><dd><p><span class="math notranslate nohighlight">\(v\)</span> as a string of substitutions (e.g., ‘M3A K5G’).
Note that while conditions may have differing wild types
at a given site, the sites between conditions should reference
the same site when alignment is performed between
condition wild types.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>func_score</cite> - The functional score computed from experimental</dt><dd><p>measurements.</p>
</dd>
</dl>
</li>
</ol>
</section>
<section id="parameters">
<h2>Parameters<a class="headerlink" href="#parameters" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>variants_df<span class="classifier">pandas.DataFrame or None</span></dt><dd><p>The variant level information from all experiments you
wish to analyze. Should have columns named ‘condition’,
‘aa_substitutions’, and ‘func_score’.
See the class note for descriptions of each of the features.</p>
</dd>
<dt>reference<span class="classifier">str</span></dt><dd><p>Name of the condition which annotates the reference.
variants. Note that for model fitting this class will convert all
amino acid substitutions for non-reference condition groups
to relative to the reference condition.
For example, if the wild type amino acid at site 30 is an
A in the reference condition, and a G in a non-reference condition,
then a Y30G mutation in the non-reference condition is recorded as an A30G
mutation relative to the reference. This way, each condition informs
the exact same parameters, even at sites that differ in wild type amino acid.
These are encoded in a <code class="docutils literal notranslate"><span class="pre">BinaryMap</span></code> object for each condtion,
where all sites that are non-identical to the reference are 1’s.
For motivation, see the <cite>Model overview</cite> section in <cite>multidms.MultiDmsModel</cite>
class notes.</p>
</dd>
<dt>collapse_identical_variants<span class="classifier">{‘mean’, ‘median’, False}</span></dt><dd><p>If identical variants in <code class="docutils literal notranslate"><span class="pre">variants_df</span></code> (same ‘aa_substitutions’),
exist within individual condition groups,
collapse them by taking mean or median of ‘func_score’, or
(if <cite>False</cite>) do not collapse at all. Collapsing will make fitting faster,
but <em>not</em> a good idea if you are doing bootstrapping.</p>
</dd>
<dt>alphabet<span class="classifier">array-like</span></dt><dd><p>Allowed characters in mutation strings.</p>
</dd>
<dt>condition_colors<span class="classifier">array-like or dict</span></dt><dd><p>Maps each condition to the color used for plotting. Either a dict keyed
by each condition, or an array of colors that are sequentially assigned
to the conditions.</p>
</dd>
<dt>letter_suffixed_sites: bool</dt><dd><p>True if sites are sequential and integer, False otherwise.</p>
</dd>
</dl>
</section>
<section id="attributes">
<h2>Attributes<a class="headerlink" href="#attributes" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>variants_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>Data to fit as passed when initializing this <code class="xref py py-class docutils literal notranslate"><span class="pre">Multidms</span></code> object
less those variants which were thrown due to mutations
outside the union of sites seen in across all condition variants.
If using <code class="docutils literal notranslate"><span class="pre">collapse_identical_variants</span></code>, then identical variants
are collapsed on columns ‘condition’, ‘aa_substitutions’,
and a column ‘weight’ is added to represent number
of collapsed variants.
Also, row-order may be changed.</p>
</dd>
<dt>mutations_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>A dataframe summarizing all valid single mutations
in a dataset. The dataframe will contain the
mutation definitions (wt, site, mut) as well as the number
of times seen in each condition.</p>
</dd>
<dt>mutations<span class="classifier">tuple</span></dt><dd><p>A tuple with all mutations in the order reletive to their index into
the binarymap.</p>
</dd>
<dt>reference<span class="classifier">str</span></dt><dd><p>The reference factor level in conditions. All mutations will be converted to
be with respect to this condition’s inferred wild type sequence. See
The class description for more.</p>
</dd>
<dt>conditions<span class="classifier">tuple</span></dt><dd><p>Names of all conditions.</p>
</dd>
<dt>condition_colors<span class="classifier">dict</span></dt><dd><p>Maps each condition to its color.</p>
</dd>
<dt>alphabet<span class="classifier">tuple</span></dt><dd><p>Allowed characters in mutation strings.</p>
</dd>
<dt>site_map<span class="classifier">tuple</span></dt><dd><p>Inferred from <code class="docutils literal notranslate"><span class="pre">variants_df</span></code>, this attribute will provide the wild type
amino acid at all sites, for all conditions.</p>
</dd>
</dl>
</section>
<section id="example">
<h2>Example<a class="headerlink" href="#example" title="Permalink to this heading"></a></h2>
<p>Simple example with two conditions (<cite>1</cite> and <cite>2</cite>)</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">multidms</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_data</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;condition&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span><span class="s2">&quot;1&quot;</span><span class="p">,</span><span class="s2">&quot;1&quot;</span><span class="p">,</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;2&quot;</span><span class="p">,</span><span class="s2">&quot;2&quot;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;aa_substitutions&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="s1">&#39;M1E&#39;</span><span class="p">,</span> <span class="s1">&#39;G3R&#39;</span><span class="p">,</span> <span class="s1">&#39;G3P&#39;</span><span class="p">,</span> <span class="s1">&#39;M1W&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E&#39;</span><span class="p">,</span> <span class="s1">&#39;P3R&#39;</span><span class="p">,</span> <span class="s1">&#39;P3G&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E P3G&#39;</span><span class="p">,</span> <span class="s1">&#39;M1E P3R&#39;</span><span class="p">,</span> <span class="s1">&#39;P2T&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;func_score&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">7</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">2.3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">2.7</span><span class="p">,</span> <span class="o">-</span><span class="mf">2.7</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">func_score_data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_score_df</span>
<span class="go">  condition aa_substitutions  func_score</span>
<span class="go">  0         1              M1E         2.0</span>
<span class="go">  1         1              G3R        -7.0</span>
<span class="go">  2         1              G3P        -0.5</span>
<span class="go">  3         1              M1W         2.3</span>
<span class="go">  4         2              M1E         1.0</span>
<span class="go">  5         2              P3R        -5.0</span>
<span class="go">  6         2              P3G         0.4</span>
<span class="go">  7         2          M1E P3G         2.7</span>
<span class="go">  8         2          M1E P3R        -2.7</span>
<span class="go">  9         2              P2T         0.3</span>
</pre></div>
</div>
<p>Instantiate a <code class="docutils literal notranslate"><span class="pre">MultiDmsData</span></code> Object allowing for stop codon variants
and declaring condition ‘1’ as the reference condition.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">MultiDmsData</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">func_score_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">alphabet</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">AAS_WITHSTOP</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;1&quot;</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>Note this may take some time due to the string
operations that must be performed when converting
amino acid substitutions to be with respect to a
reference wild type sequence.</p>
<p>After the object has finished being instantiated,
we now have access to a few ‘static’ attributes
of our data.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">reference</span>
<span class="go">&#39;1&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">conditions</span>
<span class="go">(&#39;1&#39;, &#39;2&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1E&#39;, &#39;M1W&#39;, &#39;G3P&#39;, &#39;G3R&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">site_map</span>
<span class="go">   1  2</span>
<span class="go">   3  G  P</span>
<span class="go">   1  M  M</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">mutations_df</span>
<span class="go">  mutation wts  sites muts  times_seen_1  times_seen_2</span>
<span class="go">0      M1E   M      1    E             1           3.0</span>
<span class="go">1      M1W   M      1    W             1           0.0</span>
<span class="go">2      G3P   G      3    P             1           1.0</span>
<span class="go">3      G3R   G      3    R             1           2.0</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">variants_df</span>
<span class="go">  condition aa_substitutions  weight  func_score var_wrt_ref</span>
<span class="go">0         1              G3P       1        -0.5         G3P</span>
<span class="go">1         1              G3R       1        -7.0         G3R</span>
<span class="go">2         1              M1E       1         2.0         M1E</span>
<span class="go">3         1              M1W       1         2.3         M1W</span>
<span class="go">4         2              M1E       1         1.0     G3P M1E</span>
<span class="go">5         2          M1E P3G       1         2.7         M1E</span>
<span class="go">6         2          M1E P3R       1        -2.7     G3R M1E</span>
<span class="go">8         2              P3G       1         0.4</span>
<span class="go">9         2              P3R       1        -5.0         G3R</span>
</pre></div>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.non_identical_mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">non_identical_mutations</span></span><a class="headerlink" href="#data.MultiDmsData.non_identical_mutations" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.non_identical_sites">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">non_identical_sites</span></span><a class="headerlink" href="#data.MultiDmsData.non_identical_sites" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.conditions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">conditions</span></span><a class="headerlink" href="#data.MultiDmsData.conditions" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#data.MultiDmsData.reference" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.mutations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations</span></span><a class="headerlink" href="#data.MultiDmsData.mutations" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.mutations_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_df</span></span><a class="headerlink" href="#data.MultiDmsData.mutations_df" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.variants_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants_df</span></span><a class="headerlink" href="#data.MultiDmsData.variants_df" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.site_map">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">site_map</span></span><a class="headerlink" href="#data.MultiDmsData.site_map" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.training_data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">training_data</span></span><a class="headerlink" href="#data.MultiDmsData.training_data" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.binarymaps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">binarymaps</span></span><a class="headerlink" href="#data.MultiDmsData.binarymaps" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.targets">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">targets</span></span><a class="headerlink" href="#data.MultiDmsData.targets" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="data.MultiDmsData.mut_parser">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mut_parser</span></span><a class="headerlink" href="#data.MultiDmsData.mut_parser" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="data.MultiDmsData.plot_times_seen_hist">
<span class="sig-name descname"><span class="pre">plot_times_seen_hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#data.MultiDmsData.plot_times_seen_hist" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="data.MultiDmsData.plot_func_score_boxplot">
<span class="sig-name descname"><span class="pre">plot_func_score_boxplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#data.MultiDmsData.plot_func_score_boxplot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
</dd></dl>

<span class="target" id="module-model"></span><section id="multidmsmodel">
<h2>MultiDmsModel<a class="headerlink" href="#multidmsmodel" title="Permalink to this heading"></a></h2>
<p>Defines <a class="reference internal" href="#model.MultiDmsModel" title="model.MultiDmsModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">MultiDmsModel</span></code></a> objects.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="model.MultiDmsModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">model.</span></span><span class="sig-name descname"><span class="pre">MultiDmsModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data:</span> <span class="pre">~multidms.data.MultiDmsData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">latent_model=&lt;function</span> <span class="pre">additive_model&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epistatic_model=&lt;function</span> <span class="pre">identity_activation&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_activation=&lt;function</span> <span class="pre">identity_activation&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gamma_corrected=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional_shifts=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">conditional_c=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_g_range=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_g_min=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_C_ref=5.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PRNGKey=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_percep_units=5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel" title="Permalink to this definition"></a></dt>
<dd><p>Represent one or more DMS experiments
to obtain tuned parameters that provide insight into
individual mutational effects and conditional shifts
of those effects on all non-reference conditions.</p>
<section id="id1">
<h2>Note<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h2>
<p>For each variant <span class="math notranslate nohighlight">\(v\)</span> from condition <span class="math notranslate nohighlight">\(h\)</span>,
we use a global-epistasis function <span class="math notranslate nohighlight">\(g\)</span> to convert a latent phenotype
<span class="math notranslate nohighlight">\(\phi\)</span> to a functional score <span class="math notranslate nohighlight">\(f\)</span>:</p>
<div class="math notranslate nohighlight">
\[f(v,h) = g_{\alpha}(\phi(v,h)) + γ_h\]</div>
<p>where <span class="math notranslate nohighlight">\(g\)</span> is a post-latent model
and <span class="math notranslate nohighlight">\(\alpha\)</span> is the set of parameters that define the model</p>
<p>The latent phenotype is computed in the following way:</p>
<div class="math notranslate nohighlight">
\[\phi(v,h) = c + \sum_{m \in v} (x\ *m + s*\ {m,h})\]</div>
<p>where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(c\)</span> is the wild type latent phenotype for the reference condition.</p></li>
<li><p><span class="math notranslate nohighlight">\(x_m\)</span> is the latent phenotypic effect of mutation $m$. See details below.</p></li>
<li><p><span class="math notranslate nohighlight">\(s_{m,h}\)</span> is the shift of the effect of mutation $m$ in condition $h$.
These parameters are fixed to zero for the reference condition. For
non-reference conditions, they are defined in the same way as <span class="math notranslate nohighlight">\(x_m\)</span> parameters.</p></li>
<li><p><span class="math notranslate nohighlight">\(v\)</span> is the set of all mutations relative to the reference wild type sequence
(including all mutations that separate condition <span class="math notranslate nohighlight">\(h\)</span> from the reference condition).</p></li>
</ul>
<p>For more see the <a class="reference external" href="todo">biophysical model documentation</a></p>
</section>
<section id="id2">
<h2>Attributes<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>variants_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>A copy of this object’s data.variants_df attribute,
but with the added model predictions (latent and epistatic)
given the current parameters of this model.</p>
</dd>
<dt>mutations_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>A copy of this object’s data.mutations_df attribute,
but with the added respective parameter values
(Beta and contional shift) as well as
model predictions (latent and epistatic)
given the current parameters of this model.</p>
</dd>
<dt>mutation_site_summary_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>Similar to the mutations_df attribute, but aggregate
the numerical columns by site using a chosen function.</p>
</dd>
<dt>wildtype_df<span class="classifier">pandas.DataFrame</span></dt><dd><p>Get a dataframe indexed by condition wildtype
containing the prediction features for each.</p>
</dd>
<dt>loss<span class="classifier">float</span></dt><dd><p>Compute and return the current loss of the model
with current parameters.</p>
</dd>
</dl>
</section>
<section id="id3">
<h2>Parameters<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h2>
<dl class="simple">
<dt>data<span class="classifier">multidms.MultiDmsData</span></dt><dd><p>A reference to the dataset which will define the parameters
of the model to be fit.</p>
</dd>
<dt>latent_model<span class="classifier">str</span></dt><dd><p>A string encoding of the compiled function for
a latent prediction from binary one-hot encodings
of variants. Currently, we only support the ‘additive_model’
model.
See Model Description section for more.</p>
</dd>
<dt>epistatic_model<span class="classifier">str</span></dt><dd><p>A string encoding of the compiled function for
the function mapping latent phenotype, to
some monotonic function.
Set to ‘identity’ if you wish to effectively
skip this step in the model.
See Model Description section for more.</p>
</dd>
<dt>output_activation<span class="classifier">str</span></dt><dd><p>A string encoding of the compiled function for
a final transformation on the data - primarily
we have this for our gamma corrected clipping
functions.
Set to ‘identity’ if you wish to effectively
skip this step in the model.
See Model Description section for more.</p>
</dd>
<dt>gamma_corrected<span class="classifier">bool</span></dt><dd><p>If true (default), introduce the ‘gamma’ parameter
for each non-reference parameter to
account for differences between wild type
behavior relative to it’s variants. This
is essentially a bias added to the functional
scores during fitting.
See Model Description section for more.</p>
</dd>
<dt>conditional_shifts<span class="classifier">bool</span></dt><dd><p>If true (default) initialize and fit the shift
parameters for each non-reference condition.
See Model Description section for more.</p>
</dd>
<dt>conditional_c<span class="classifier">bool</span></dt><dd><p>If True (default False) introduce a bias parameter
on the latent space parameter for each condition.
See Model Description section for more.</p>
</dd>
<dt>init_g_range<span class="classifier">float</span></dt><dd><p>Initialize the range of two-parameter epistatic models
(Sigmoid or Softplus).</p>
</dd>
<dt>init_g_min<span class="classifier">float</span></dt><dd><p>Initialize the min of a two parameter epistatic models.
(Sigmoid or Softplus).</p>
</dd>
<dt>init_C_ref<span class="classifier">float</span></dt><dd><p>Initialize the $C_{ref}$ parameter.</p>
</dd>
<dt>PRNGKey<span class="classifier">int</span></dt><dd><p>The initial seed key for random parameters
assigned to Beta’s and any other randomly
initialized parameters.</p>
</dd>
<dt>n_percep_units<span class="classifier">int or None</span></dt><dd><p>If using a perceptron global epistasis
model, this is the number of hidden units
used in the transform.</p>
</dd>
</dl>
</section>
<section id="id4">
<h2>Example<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h2>
<p>To create a <code class="docutils literal notranslate"><span class="pre">MultiDmsModel</span></code> object, all you need is
the respective <code class="docutils literal notranslate"><span class="pre">MultiDmsData</span></code> object for parameter fitting,
and to specify the post-latent model options for the model.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span> <span class="o">=</span> <span class="n">multidms</span><span class="o">.</span><span class="n">MultiDmsModel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">data</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">epistatic_model</span><span class="o">=</span><span class="n">multidms</span><span class="o">.</span><span class="n">biophysical</span><span class="o">.</span><span class="n">sigmoidal_global_epistasis</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">output_activation</span><span class="o">=</span><span class="n">multidms</span><span class="o">.</span><span class="n">biophysical</span><span class="o">.</span><span class="n">softplus_activation</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>Upon initialization, you will now have access to the underlying data
and parameters.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">mutations</span>
<span class="go">(&#39;M1E&#39;, &#39;M1W&#39;, &#39;G3P&#39;, &#39;G3R&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">conditions</span>
<span class="go">(&#39;1&#39;, &#39;2&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">reference</span>
<span class="go">&#39;1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">condition_colors</span>
<span class="go">{&#39;1&#39;: &#39;#0072B2&#39;, &#39;2&#39;: &#39;#009E73&#39;}</span>
</pre></div>
</div>
<p>The mutations_df and variants_df may of course also be accessed.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">mutations_df</span>
<span class="go">  mutation wts  sites muts  times_seen_1  times_seen_2</span>
<span class="go">0      M1E   M      1    E             1           3.0</span>
<span class="go">1      M1W   M      1    W             1           0.0</span>
<span class="go">2      G3P   G      3    P             1           1.0</span>
<span class="go">3      G3R   G      3    R             1           2.0</span>
</pre></div>
</div>
<p>However, if accessed directly through the Model object, you will
get the same information, along with model/parameter specific
features included. These are automatically updated each time you
request the property.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">mutations_df</span>
<span class="go">  mutation wts  sites muts  times_seen_1  ...         β  S_1       F_1  S_2       F_2</span>
<span class="go">0      M1E   M      1    E             1  ...  0.080868  0.0 -0.030881  0.0 -0.030881</span>
<span class="go">1      M1W   M      1    W             1  ... -0.386247  0.0 -0.049086  0.0 -0.049086</span>
<span class="go">2      G3P   G      3    P             1  ... -0.375656  0.0 -0.048574  0.0 -0.048574</span>
<span class="go">3      G3R   G      3    R             1  ...  1.668974  0.0 -0.006340  0.0 -0.006340</span>
</pre></div>
</div>
<p>Notice the respective single mutation effects (“β”), conditional shifts (S_d),
and predicted functional score (F_d) of each mutation in the model are now
easily accessible. Similarly, we can take a look at the variants_df for the model,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">variants_df</span>
<span class="go">  condition aa_substitutions  ...  predicted_func_score  corrected_func_score</span>
<span class="go">0         1              G3P  ...             -0.048574                  -0.5</span>
<span class="go">1         1              G3R  ...             -0.006340                  -7.0</span>
<span class="go">2         1              M1E  ...             -0.030881                   2.0</span>
<span class="go">3         1              M1W  ...             -0.049086                   2.3</span>
<span class="go">4         2              M1E  ...             -0.044834                   1.0</span>
<span class="go">5         2          M1E P3G  ...             -0.030881                   2.7</span>
<span class="go">6         2          M1E P3R  ...             -0.005848                  -2.7</span>
<span class="go">8         2              P3G  ...             -0.033464                   0.4</span>
<span class="go">9         2              P3R  ...             -0.006340                  -5.0</span>
</pre></div>
</div>
<p>We now have access to the predicted (and gamma corrected) functional scores
as predicted by the models current parameters.</p>
<p>So far, these parameters and predictions results from them have not been tuned
to the dataset. Let’s take a look at the loss on the training dataset
given our initialized parameters</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">loss</span>
<span class="go">DeviceArray(4.40537408, dtype=float64)</span>
</pre></div>
</div>
<p>Next, we fit the model with some chosen hyperparameters.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">maxiter</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">lasso_shift</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">model</span><span class="o">.</span><span class="n">loss</span>
<span class="go">DeviceArray(1.01582782, dtype=float64)</span>
</pre></div>
</div>
<p>The model tunes it’s parameters in place, and the subsequent call to retrieve
the loss reflects our models loss given it’s updated parameters.</p>
<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.params">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">params</span></span><a class="headerlink" href="#model.MultiDmsModel.params" title="Permalink to this definition"></a></dt>
<dd><p>All current model parameters in a dictionary.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">data</span></span><a class="headerlink" href="#model.MultiDmsModel.data" title="Permalink to this definition"></a></dt>
<dd><p>multidms.MultiDmsData Object this model references for fitting
it’s parameters.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.model_components">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">model_components</span></span><a class="headerlink" href="#model.MultiDmsModel.model_components" title="Permalink to this definition"></a></dt>
<dd><p>A frozendict which hold the individual components of the model
as well as the objective and forward functions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.variants_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variants_df</span></span><a class="headerlink" href="#model.MultiDmsModel.variants_df" title="Permalink to this definition"></a></dt>
<dd><p>Get all functional score attributes from self._data
updated with all model predictions</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.mutations_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mutations_df</span></span><a class="headerlink" href="#model.MultiDmsModel.mutations_df" title="Permalink to this definition"></a></dt>
<dd><p>Get all single mutational attributes from self._data
updated with all model specific attributes.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.wildtype_df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wildtype_df</span></span><a class="headerlink" href="#model.MultiDmsModel.wildtype_df" title="Permalink to this definition"></a></dt>
<dd><p>Get a dataframe indexed by condition wildtype
containing the prediction features for each.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="model.MultiDmsModel.loss">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">loss</span></span><a class="headerlink" href="#model.MultiDmsModel.loss" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.mutation_site_summary_df">
<span class="sig-name descname"><span class="pre">mutation_site_summary_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">agg_func=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.mutation_site_summary_df" title="Permalink to this definition"></a></dt>
<dd><p>Get all single mutational attributes from self._data
updated with all model specific attributes, then aggregate
all numerical columns by “sites” using
<code class="docutils literal notranslate"><span class="pre">agg</span></code> function. The mean values are given by default.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.get_condition_params">
<span class="sig-name descname"><span class="pre">get_condition_params</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.get_condition_params" title="Permalink to this definition"></a></dt>
<dd><p>get the relent parameters for a model prediction</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.f">
<span class="sig-name descname"><span class="pre">f</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.f" title="Permalink to this definition"></a></dt>
<dd><p>condition specific prediction on X using the biophysical model
given current model parameters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.predicted_latent">
<span class="sig-name descname"><span class="pre">predicted_latent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.predicted_latent" title="Permalink to this definition"></a></dt>
<dd><p>condition specific prediction on X using the biophysical model
given current model parameters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.fit_reference_beta">
<span class="sig-name descname"><span class="pre">fit_reference_beta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.fit_reference_beta" title="Permalink to this definition"></a></dt>
<dd><p>Fit the Model β’s to the reference data</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lasso_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lock_params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.fit" title="Permalink to this definition"></a></dt>
<dd><p>Use jaxopt.ProximalGradiant to optimize parameters</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_pred_accuracy">
<span class="sig-name descname"><span class="pre">plot_pred_accuracy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_corr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_pred_accuracy" title="Permalink to this definition"></a></dt>
<dd><p>Create a figure which visualizes the correlation
between model predicted functional score of all
variants in the training with ground truth measurements.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_epistasis">
<span class="sig-name descname"><span class="pre">plot_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_epistasis" title="Permalink to this definition"></a></dt>
<dd><p>Plot latent predictions against
gamma corrected ground truth measurements
of all samples in the training set.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_param_hist">
<span class="sig-name descname"><span class="pre">plot_param_hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_param_hist" title="Permalink to this definition"></a></dt>
<dd><p>Plot the histogram of a parameter.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_param_heatmap">
<span class="sig-name descname"><span class="pre">plot_param_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_param_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>plot the heatmap of a parameters associated with specific sites and substitutions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_shifts_by_site">
<span class="sig-name descname"><span class="pre">plot_shifts_by_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">times_seen_threshold=3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_func=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_shifts_by_site" title="Permalink to this definition"></a></dt>
<dd><p>summarize shift parameter values by associated sites and conditions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.plot_fit_param_comp_scatter">
<span class="sig-name descname"><span class="pre">plot_fit_param_comp_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'β'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'β'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5,</span> <span class="pre">4]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">saveas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_agg_func</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.plot_fit_param_comp_scatter" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="model.MultiDmsModel.mut_shift_plot">
<span class="sig-name descname"><span class="pre">mut_shift_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#model.MultiDmsModel.mut_shift_plot" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper method for visualizing the shift plot.
see <cite>multidms.plot.mut_shift_plot()</cite> for more</p>
</dd></dl>

</section>
</dd></dl>

<span class="target" id="module-biophysical"></span><section id="biophysical-model">
<h2>biophysical_model<a class="headerlink" href="#biophysical-model" title="Permalink to this heading"></a></h2>
<p>Defines functions for
modeling global epistasis biophysical models
and their respective objective functions
in <cite>multidms.MultiDmsModel</cite> Objects.</p>
</section>
<dl class="py function">
<dt class="sig sig-object py" id="biophysical.identity_activation">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">identity_activation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">act</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.identity_activation" title="Permalink to this definition"></a></dt>
<dd><p>identity function <span class="math notranslate nohighlight">\(f(x)=x\)</span></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.softplus_activation">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">softplus_activation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">act</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lower_bound</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-3.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hinge_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.softplus_activation" title="Permalink to this definition"></a></dt>
<dd><p>A modified softplus that hinges at ‘lower_bound’.
The rate of change at the hinge is defined by ‘hinge_scale’. This is derived from
<a class="reference external" href="https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softplus.html">https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softplus.html</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.additive_model">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">additive_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.additive_model" title="Permalink to this definition"></a></dt>
<dd><p>Model for predicting latent space with
shift parameters.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.sigmoidal_global_epistasis">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">sigmoidal_global_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">α</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.sigmoidal_global_epistasis" title="Permalink to this definition"></a></dt>
<dd><p>A flexible sigmoid function for
modeling global epistasis.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.softplus_global_epistasis">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">softplus_global_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">α</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.softplus_global_epistasis" title="Permalink to this definition"></a></dt>
<dd><p>A flexible sigmoid function for
modeling global epistasis.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.perceptron_global_epistasis">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">perceptron_global_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">α</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.perceptron_global_epistasis" title="Permalink to this definition"></a></dt>
<dd><p>A flexible sigmoid function for
modeling global epistasis.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.abstract_from_latent">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">abstract_from_latent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CompiledFunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CompiledFunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.abstract_from_latent" title="Permalink to this definition"></a></dt>
<dd><p>Biophysical model - compiled for optimization
until model functions additive_model, g, and t are updated.
Model may be composed &amp; compiled
with the required functions fixed
using ‘jax.tree_util.Partial.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.abstract_epistasis">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">abstract_epistasis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">additive_model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CompiledFunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CompiledFunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">t</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">CompiledFunction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d_params</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_h</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.abstract_epistasis" title="Permalink to this definition"></a></dt>
<dd><p>Biophysical model - compiled for optimization
until model functions additive_model, g, and t are updated.
Model may be composed &amp; compiled
with the required functions fixed
using ‘jax.tree_util.Partial’.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.lasso_lock_prox">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">lasso_lock_prox</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hyperparams_prox</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'lasso_params':</span> <span class="pre">None,</span> <span class="pre">'lock_params':</span> <span class="pre">None}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scaling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.lasso_lock_prox" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="biophysical.gamma_corrected_cost_smooth">
<span class="sig-prename descclassname"><span class="pre">biophysical.</span></span><span class="sig-name descname"><span class="pre">gamma_corrected_cost_smooth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">δ</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">λ_ridge_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">λ_ridge_beta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#biophysical.gamma_corrected_cost_smooth" title="Permalink to this definition"></a></dt>
<dd><p>Cost (Objective) function summed across all conditions</p>
</dd></dl>

</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/2020-logo-1000px-transparent.png" alt="Logo"/>
    
    <h1 class="logo logo-name">multidms</h1>
    
  </a>
</p>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=matsengrp&repo=multidms&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">multidms documentation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="biophysical_model.html">Biophysical model</a></li>
<li class="toctree-l1"><a class="reference internal" href="fit_delta_BA1_example.html"><code class="docutils literal notranslate"><span class="pre">multidms</span></code> fitting pipeline</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgments</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="fit_delta_BA1_example.html" title="previous chapter"><code class="docutils literal notranslate"><span class="pre">multidms</span></code> fitting pipeline</a></li>
      <li>Next: <a href="acknowledgments.html" title="next chapter">Acknowledgments</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2023, Jared Galloway, Hugh Haddox.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 6.1.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/autodoc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/matsengrp/multidms" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>